---
title: "LNCaP Analysis"
author: "Jordan S. Kesner"
date: "9/23/2019"
output: pdf_document
editor_options: 
  chunk_output_type: console
---

Install libraries
```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install("MotifDb")
BiocManager::install("BSgenome.Hsapiens.UCSC.hg38")
BiocManager::install("Rsamtools")
BiocManager::install("GenomicAlignments")
BiocManager::install("Biostrings")
BiocManager::install("seqbias")
BiocManager::install("GenomicRanges")
BiocManager::install("genomation")
BiocManager::install("ChIPseeker")
BiocManager::install("TxDb.Hsapiens.UCSC.hg38.knownGene")
BiocManager::install("org.Hs.eg.db")
BiocManager::install("org.Mm.eg.db")
BiocManager::install("biomaRt")
BiocManager::install("limma")
BiocManager::install("aracne.networks")
BiocManager::install("viper")
BiocManager::install("phantasus")
BiocManager::install("AnnotationDbi")
BiocManager::install("mygene")
BiocManager::install("mixtools")
BiocManager::install("bcellViper")
BiocManager::install("EnsDb.Hsapiens.v86")
BiocManager::install("tximport")
BiocManager::install("pathfindR")
BiocManager::install("diggit")
BiocManager::install("DESeq2")
BiocManager::install("GenomeInfoDb")
n
install.packages("dplyr")
install.packages("ggplot2")
install.packages("pheatmap")
install.packages("RIdeogram")
install.packages("enrichR")
```

Set the working directory and library path
```{r}
setwd("C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data")
#.libPaths("C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\lib")
.libPaths("C:\\Users\\Jordan\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\lib")
```

Load libraries
```{r}
library(MotifDb)
library(BSgenome.Hsapiens.UCSC.hg38)
library(Rsamtools)
library(GenomicAlignments)
library(Biostrings)
library(seqbias)
library(GenomicRanges)
library(genomation)
library(ChIPseeker)
library(GenomicRanges)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(org.Hs.eg.db)
library(org.Mm.eg.db)
library(biomaRt)
library(limma)
library(dplyr)
library(aracne.networks)
library(viper)
library(pheatmap)
library(RIdeogram)
library(phantasus)
library(enrichR)
library(AnnotationDbi)
library(mygene)
library(mixtools)
library(bcellViper)
library(tximport)
library(EnsDb.Hsapiens.v86)
library(viper)
library(pathfindR)
library(diggit)
library(DESeq2)
library(ggplot2)
genome <- Hsapiens
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
```

Source functions
```{r}
## My functions
source("C:\\Users\\jsk33\\OneDrive\\Git\\atacPipelineMaster\\snakeResources\\scripts\\atacFunctions.R")
## Alessandro functions
#source("C:\\Users\\jsk33\\OneDrive\\Git\\vaxtools\\R\\utils.R")
```

Local functions
```{r}

entrez2gene.regulon <- function( regulon , organism = "hg" , entrez2gene = TRUE ){
    if (entrez2gene)
    {
      regulon <- lapply( regulon , function(x) { names(x[[1]]) <- unname( entrez2geneSymbol.old( names(x[[1]]) , organism ) ) ; x } )
      names(regulon) <- unname( entrez2geneSymbol.old(names(regulon) , organism) )
    } else {

      regulon <- lapply( regulon , function(x) { names(x[[1]]) <- unname( entrez2geneSymbol.old( names(x[[1]]) , organism ) ) ; x } )
      names(regulon) <- unname( entrez2geneSymbol.old(names(regulon) , organism) )
    }
    return(regulon)
}

entrez2geneSymbol.old <- function( entrez , organism = "hg" ){
    organism.to.query = switch ( organism ,
      "hg" = org.Hs.eg.db ,
      "mm" = org.Mm.eg.db
    )

    symbol <- mapIds( organism.to.query ,
                      keys = entrez ,
                      column = "SYMBOL",
                      keytype = "ENTREZID" ,
                      multiVals = "first" )
    return(symbol)
  }

transformPeak <- function(inputGR, quantThreshold){
  plotGR <- unique(inputGR)
  score <- plotGR@elementMetadata@listData[["score"]]
  log2score <- log2(score)
  plotGR@elementMetadata@listData[["score"]] <- log2score
  quant <- quantile(plotGR@elementMetadata@listData[["score"]], probs = c(as.numeric(quantThreshold)))
  plotGR <- plotGR[which(plotGR@elementMetadata@listData[["score"]] > quant[[1]])]}

printPeakScores <- function(inputGR){
  scores <- inputGR@elementMetadata@listData[["score"]]
  minScore <- min(scores)
  maxScore <- max(scores)
  cat("Min score:", minScore, "\n")
  cat("Max score:", maxScore, "\n")}

generatePeakIdeogram <- function(inputGR, outPath){
  printPeakScores(inputGR)
  data(human_karyotype, package="RIdeogram")
  chrNames <- c(as.character(seqnames(inputGR)))
  chrNames <- gsub("chr", "", chrNames)
  peakDF <- data.frame(
    Chr = chrNames,
    Start = inputGR@ranges@start,
    End = inputGR@ranges@start + inputGR@ranges@width,
    Value = inputGR@elementMetadata@listData[["score"]])
  ideogram(
    karyotype = human_karyotype,
    overlaid = peakDF,
    output = outPath)
  convertSVG(outPath, device = "png")}

getPeakDifferential <- function(controlPeak, subjectPeak){
  overlap <- findOverlaps(controlPeak, subjectPeak, maxgap = 100, ignore.strand = TRUE)
  idxoverlap <- overlap@to
  diffpeaks <- subjectPeak[-idxoverlap]
  return(diffpeaks)}

writeEnrichedGenesToFile <- function(inputGR, outPath){
  anno <- annotatePeak(inputGR, tssRegion=c(-1000, 1000), TxDb=txdb, annoDb="org.Hs.eg.db")
  scores <- anno@anno@elementMetadata@listData[["score"]]
  genes <- anno@anno@elementMetadata@listData[["SYMBOL"]]
  maxscore <- max(scores)
  normscore <- scores/maxscore
  new <- paste0(genes,",",normscore)
  write(new, file = outPath)
}

getSymbolFromGR <- function(inputGR){
  anno <- annotatePeak(inputGR, tssRegion=c(-500, 500), TxDb=txdb, annoDb="org.Hs.eg.db")
  genes <- anno@anno@elementMetadata@listData[["SYMBOL"]]
  return(genes)
}

entrez2gene.regulon <- function( regulon , organism = "hg" , entrez2gene = TRUE ){
    if (entrez2gene)
    {
      regulon <- lapply( regulon , function(x) { names(x[[1]]) <- unname( entrez2geneSymbol.old( names(x[[1]]) , organism ) ) ; x } )
      names(regulon) <- unname( entrez2geneSymbol.old(names(regulon) , organism) )
    } else {

      regulon <- lapply( regulon , function(x) { names(x[[1]]) <- unname( entrez2geneSymbol.old( names(x[[1]]) , organism ) ) ; x } )
      names(regulon) <- unname( entrez2geneSymbol.old(names(regulon) , organism) )
    }
    return(regulon)
}

entrez2geneSymbol.old <- function( entrez , organism = "hg" ){
    organism.to.query = switch ( organism ,
      "hg" = org.Hs.eg.db ,
      "mm" = org.Mm.eg.db
    )

    symbol <- mapIds( organism.to.query ,
                      keys = entrez ,
                      column = "SYMBOL",
                      keytype = "ENTREZID" ,
                      multiVals = "first" )
    return(symbol)
}

```

Setup Functions
```{r}

setupPeakPaths <- function(){

# p-value 0.0001 narrow peak
path.CR01.p0001.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p0001\\LNCaP-CR-01.rep1.refhg38_globalnorm_p0001_peaks.narrowPeak"
path.CR02.p0001.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p0001\\LNCaP-CR-02.rep1.refhg38_globalnorm_p0001_peaks.narrowPeak"
path.CR04.p0001.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p0001\\LNCaP-CR-04.rep1.refhg38_globalnorm_p0001_peaks.narrowPeak"
path.CR05.p0001.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p0001\\LNCaP-CR-05.rep1.refhg38_globalnorm_p0001_peaks.narrowPeak"
path.CR07.p0001.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p0001\\LNCaP-CR-07.rep1.refhg38_globalnorm_p0001_peaks.narrowPeak"
path.CR08.p0001.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p0001\\LNCaP-CR-08.rep1.refhg38_globalnorm_p0001_peaks.narrowPeak"
path.WT01.p0001.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p0001\\LNCaP-WT-01.rep1.refhg38_globalnorm_p0001_peaks.narrowPeak"
path.WT02.p0001.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p0001\\LNCaP-WT-02.rep1.refhg38_globalnorm_p0001_peaks.narrowPeak"

# p-value 0.001 narrow peak
path.CR01.p001.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p001\\LNCaP-CR-01.rep1.refhg38_globalnorm_p001_peaks.narrowPeak"
path.CR02.p001.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p001\\LNCaP-CR-02.rep1.refhg38_globalnorm_p001_peaks.narrowPeak"
path.CR04.p001.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p001\\LNCaP-CR-04.rep1.refhg38_globalnorm_p001_peaks.narrowPeak"
path.CR05.p001.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p001\\LNCaP-CR-05.rep1.refhg38_globalnorm_p001_peaks.narrowPeak"
path.CR07.p001.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p001\\LNCaP-CR-07.rep1.refhg38_globalnorm_p001_peaks.narrowPeak"
path.CR08.p001.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p001\\LNCaP-CR-08.rep1.refhg38_globalnorm_p001_peaks.narrowPeak"
path.WT01.p001.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p001\\LNCaP-WT-01.rep1.refhg38_globalnorm_p001_peaks.narrowPeak"
path.WT02.p001.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p001\\LNCaP-WT-02.rep1.refhg38_globalnorm_p001_peaks.narrowPeak"

# p-value 0.01 narrow peak
path.CR01.p01.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p01\\LNCaP-CR-01.rep1.refhg38_globalnorm_p01_peaks.narrowPeak"
path.CR02.p01.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p01\\LNCaP-CR-02.rep1.refhg38_globalnorm_p01_peaks.narrowPeak"
path.CR04.p01.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p01\\LNCaP-CR-04.rep1.refhg38_globalnorm_p01_peaks.narrowPeak"
path.CR05.p01.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p01\\LNCaP-CR-05.rep1.refhg38_globalnorm_p01_peaks.narrowPeak"
path.CR07.p01.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p01\\LNCaP-CR-07.rep1.refhg38_globalnorm_p01_peaks.narrowPeak"
path.CR08.p01.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p01\\LNCaP-CR-08.rep1.refhg38_globalnorm_p01_peaks.narrowPeak"
path.WT01.p01.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p01\\LNCaP-WT-01.rep1.refhg38_globalnorm_p01_peaks.narrowPeak"
path.WT02.p01.narrowPeak <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p01\\LNCaP-WT-02.rep1.refhg38_globalnorm_p01_peaks.narrowPeak"

# p-value 0.0001 summit
path.CR01.p0001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p0001\\LNCaP-CR-01.rep1.refhg38_globalnorm_p0001_summits.bed"
path.CR02.p0001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p0001\\LNCaP-CR-02.rep1.refhg38_globalnorm_p0001_summits.bed"
path.CR04.p0001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p0001\\LNCaP-CR-04.rep1.refhg38_globalnorm_p0001_summits.bed"
path.CR05.p0001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p0001\\LNCaP-CR-05.rep1.refhg38_globalnorm_p0001_summits.bed"
path.CR07.p0001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p0001\\LNCaP-CR-07.rep1.refhg38_globalnorm_p0001_summits.bed"
path.CR08.p0001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p0001\\LNCaP-CR-08.rep1.refhg38_globalnorm_p0001_summits.bed"
path.WT01.p0001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p0001\\LNCaP-WT-01.rep1.refhg38_globalnorm_p0001_summits.bed"
path.WT02.p0001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p0001\\LNCaP-WT-02.rep1.refhg38_globalnorm_p0001_summits.bed"

# p-value 0.001 summit
path.CR01.p001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p001\\LNCaP-CR-01.rep1.refhg38_globalnorm_p001_summits.bed"
path.CR02.p001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p001\\LNCaP-CR-02.rep1.refhg38_globalnorm_p001_summits.bed"
path.CR04.p001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p001\\LNCaP-CR-04.rep1.refhg38_globalnorm_p001_summits.bed"
path.CR05.p001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p001\\LNCaP-CR-05.rep1.refhg38_globalnorm_p001_summits.bed"
path.CR07.p001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p001\\LNCaP-CR-07.rep1.refhg38_globalnorm_p001_summits.bed"
path.CR08.p001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p001\\LNCaP-CR-08.rep1.refhg38_globalnorm_p001_summits.bed"
path.WT01.p001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p001\\LNCaP-WT-01.rep1.refhg38_globalnorm_p001_summits.bed"
path.WT02.p001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p001\\LNCaP-WT-02.rep1.refhg38_globalnorm_p001_summits.bed"

# p-value 0.01 summit
path.CR01.p01.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p01\\LNCaP-CR-01.rep1.refhg38_globalnorm_p01_summits.bed"
path.CR02.p01.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p01\\LNCaP-CR-02.rep1.refhg38_globalnorm_p01_summits.bed"
path.CR04.p01.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p01\\LNCaP-CR-04.rep1.refhg38_globalnorm_p01_summits.bed"
path.CR05.p01.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p01\\LNCaP-CR-05.rep1.refhg38_globalnorm_p01_summits.bed"
path.CR07.p01.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p01\\LNCaP-CR-07.rep1.refhg38_globalnorm_p01_summits.bed"
path.CR08.p01.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p01\\LNCaP-CR-08.rep1.refhg38_globalnorm_p01_summits.bed"
path.WT01.p01.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p01\\LNCaP-WT-01.rep1.refhg38_globalnorm_p01_summits.bed"
path.WT02.p01.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\global_p01\\LNCaP-WT-02.rep1.refhg38_globalnorm_p01_summits.bed"
  
  
  
}

runPeaksSetup <- function(){

#### Ideogram output paths
outPath.ideogram.CR01.p0001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\figures\\ideograms\\ideogram.CR01.p0001.summit.svg"
outPath.ideogram.CR02.p0001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\figures\\ideograms\\ideogram.CR02.p0001.summit.svg"
outPath.ideogram.CR04.p0001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\figures\\ideograms\\ideogram.CR04.p0001.summit.svg"
outPath.ideogram.CR05.p0001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\figures\\ideograms\\ideogram.CR05.p0001.summit.svg"
outPath.ideogram.CR07.p0001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\figures\\ideograms\\ideogram.CR07.p0001.summit.svg"
outPath.ideogram.CR08.p0001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\figures\\ideograms\\ideogram.CR08.p0001.summit.svg"
outPath.ideogram.WT01.p0001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\figures\\ideograms\\ideogram.WT01.p0001.summit.svg"
outPath.ideogram.WT02.p0001.summit <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\figures\\ideograms\\ideogram.WT02.p0001.summit.svg"

#### Ideogram output paths peak differentials
outPath.ideogram.CR01.p0001.diff <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\figures\\ideograms\\ideogram.CR01.p0001.diff.svg"
outPath.ideogram.CR02.p0001.diff <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\figures\\ideograms\\ideogram.CR02.p0001.diff.svg"
outPath.ideogram.CR04.p0001.diff <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\figures\\ideograms\\ideogram.CR04.p0001.diff.svg"
outPath.ideogram.CR05.p0001.diff <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\figures\\ideograms\\ideogram.CR05.p0001.diff.svg"
outPath.ideogram.CR07.p0001.diff <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\figures\\ideograms\\ideogram.CR07.p0001.diff.svg"
outPath.ideogram.CR08.p0001.diff <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\figures\\ideograms\\ideogram.CR08.p0001.diff.svg"

#### Load the peaks
CR01.p0001.summit <<- importBED(path.CR01.p0001.summit)
CR02.p0001.summit <<- importBED(path.CR02.p0001.summit)
CR04.p0001.summit <<- importBED(path.CR04.p0001.summit)
CR05.p0001.summit <<- importBED(path.CR05.p0001.summit)
CR07.p0001.summit <<- importBED(path.CR07.p0001.summit)
CR08.p0001.summit <<- importBED(path.CR08.p0001.summit)
WT01.p0001.summit <<- importBED(path.WT01.p0001.summit)
WT02.p0001.summit <<- importBED(path.WT02.p0001.summit)

#### Transform the peaks
CR01.p0001.summit.log2.q9 <<- transformPeak(CR01.p0001.summit, 0.9)
CR02.p0001.summit.log2.q9 <<- transformPeak(CR02.p0001.summit, 0.9)
CR04.p0001.summit.log2.q9 <<- transformPeak(CR04.p0001.summit, 0.9)
CR05.p0001.summit.log2.q9 <<- transformPeak(CR05.p0001.summit, 0.9)
CR07.p0001.summit.log2.q9 <<- transformPeak(CR07.p0001.summit, 0.9)
CR08.p0001.summit.log2.q9 <<- transformPeak(CR08.p0001.summit, 0.9)
WT01.p0001.summit.log2.q9 <<- transformPeak(WT01.p0001.summit, 0.9)
WT02.p0001.summit.log2.q9 <<- transformPeak(WT02.p0001.summit, 0.9)

#### Peak differentials
WT.p0001.merged <<- subsetByOverlaps(WT01.p0001.summit, WT02.p0001.summit, maxgap = 100)
##
diffpeaks.CR01 <<- getPeakDifferential(WT.p0001.merged, CR01.p0001.summit)
diffpeaks.CR02 <<- getPeakDifferential(WT.p0001.merged, CR02.p0001.summit)
diffpeaks.CR04 <<- getPeakDifferential(WT.p0001.merged, CR04.p0001.summit)
diffpeaks.CR05 <<- getPeakDifferential(WT.p0001.merged, CR05.p0001.summit)
diffpeaks.CR07 <<- getPeakDifferential(WT.p0001.merged, CR07.p0001.summit)
diffpeaks.CR08 <<- getPeakDifferential(WT.p0001.merged, CR08.p0001.summit)

#### Transform the peaks
diffpeaks.CR01.log2.q9 <<- transformPeak(diffpeaks.CR01, 0.9)
diffpeaks.CR02.log2.q9 <<- transformPeak(diffpeaks.CR02, 0.9)
diffpeaks.CR04.log2.q9 <<- transformPeak(diffpeaks.CR04, 0.9)
diffpeaks.CR05.log2.q9 <<- transformPeak(diffpeaks.CR05, 0.9)
diffpeaks.CR07.log2.q9 <<- transformPeak(diffpeaks.CR07, 0.9)
diffpeaks.CR08.log2.q9 <<- transformPeak(diffpeaks.CR08, 0.9)

}

setupBamPaths <- function(){
path.CR01.bam <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bam\\LNCaP-CR-01.rep1.refhg38.bam"
path.CR02.bam <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bam\\LNCaP-CR-02.rep1.refhg38.bam"
path.CR04.bam <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bam\\LNCaP-CR-04.rep1.refhg38.bam"
path.CR05.bam <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bam\\LNCaP-CR-05.rep1.refhg38.bam"
path.CR07.bam <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bam\\LNCaP-CR-07.rep1.refhg38.bam"
path.CR08.bam <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bam\\LNCaP-CR-08.rep1.refhg38.bam"
path.WT01.bam <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bam\\LNCaP-WT-01.rep1.refhg38.bam"
path.WT02.bam <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bam\\LNCaP-WT-02.rep1.refhg38.bam"
}

setupBwPaths <- function(){
path.CR01.bigwig <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bigiwg\\LNCaP-CR-01.rep1.refhg38.bw"
path.CR02.bigwig <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bigiwg\\LNCaP-CR-02.rep1.refhg38.bw"
path.CR04.bigwig <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bigiwg\\LNCaP-CR-04.rep1.refhg38.bw"
path.CR05.bigwig <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bigiwg\\LNCaP-CR-05.rep1.refhg38.bw"
path.CR07.bigwig <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bigiwg\\LNCaP-CR-07.rep1.refhg38.bw"
path.CR08.bigwig <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bigiwg\\LNCaP-CR-08.rep1.refhg38.bw"
path.WT01.bigwig <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bigiwg\\LNCaP-WT-01.rep1.refhg38.bw"
path.WT02.bigwig <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bigiwg\\LNCaP-WT-02.rep1.refhg38.bw"
}

```

Run setup functions
```{r}
setupPeakPaths()
runPeaksSetup()
setupBamPaths()
setupBwPaths()

```

Generate Genomic Ideograms
```{r}

## Setup
setupPeakPaths()
runPeaksSetup()

#### Ideograms
generatePeakIdeogram(CR01.p0001.summit.log2.q9, outPath.ideogram.CR01.p0001.summit)
generatePeakIdeogram(CR02.p0001.summit.log2.q9, outPath.ideogram.CR02.p0001.summit)
generatePeakIdeogram(CR04.p0001.summit.log2.q9, outPath.ideogram.CR04.p0001.summit)
generatePeakIdeogram(CR05.p0001.summit.log2.q9, outPath.ideogram.CR05.p0001.summit)
generatePeakIdeogram(CR07.p0001.summit.log2.q9, outPath.ideogram.CR07.p0001.summit)
generatePeakIdeogram(CR08.p0001.summit.log2.q9, outPath.ideogram.CR08.p0001.summit)
generatePeakIdeogram(WT01.p0001.summit.log2.q9, outPath.ideogram.WT01.p0001.summit)
generatePeakIdeogram(WT02.p0001.summit.log2.q9, outPath.ideogram.WT02.p0001.summit)

#### Ideograms
generatePeakIdeogram(diffpeaks.CR01.log2.q9, outPath.ideogram.CR01.p0001.diff)
generatePeakIdeogram(diffpeaks.CR02.log2.q9, outPath.ideogram.CR02.p0001.diff)
generatePeakIdeogram(diffpeaks.CR04.log2.q9, outPath.ideogram.CR04.p0001.diff)
generatePeakIdeogram(diffpeaks.CR05.log2.q9, outPath.ideogram.CR05.p0001.diff)
generatePeakIdeogram(diffpeaks.CR07.log2.q9, outPath.ideogram.CR07.p0001.diff)
generatePeakIdeogram(diffpeaks.CR08.log2.q9, outPath.ideogram.CR08.p0001.diff)

####
WT01.merged.p0001.summit.log2 <- transformPeak(WT.p0001.merged, 0.5)
outPath <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\figures\\ideograms\\ideogram.WT.merged.p0001.log2.svg"
generatePeakIdeogram(WT01.merged.p0001.summit.log2, outPath)

```

EnrichR analysis for peaks (use the web interface, much easier)
```{r}

## might want to try by using only unqique genes also, need to write code

####
CR01.outpath <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\cr01.txt"
CR02.outpath <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\cr02.txt"
CR04.outpath <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\cr04.txt"
CR05.outpath <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\cr05.txt"
CR07.outpath <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\cr07.txt"
CR08.outpath <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\cr08.txt"

####


####
writeEnrichedGenesToFile(diffpeaks.CR01.log2.q9, CR01.outpath)
writeEnrichedGenesToFile(diffpeaks.CR02.log2.q9, CR02.outpath)
writeEnrichedGenesToFile(diffpeaks.CR04.log2.q9, CR04.outpath)
writeEnrichedGenesToFile(diffpeaks.CR05.log2.q9, CR05.outpath)
writeEnrichedGenesToFile(diffpeaks.CR07.log2.q9, CR07.outpath)
writeEnrichedGenesToFile(diffpeaks.CR08.log2.q9, CR08.outpath)

## list the databases
dbs <- listEnrichrDbs()

## set database
dbs <- c("GO_Molecular_Function_2018", "GO_Cellular_Component_2018", "GO_Biological_Process_2018", "Cancer_Cell_Line_Encyclopedia", "ARCHS4_Tissues", "Encode_Histone_Modifications_2015", "ENCODE_TF_ChIP-seq_2015", "KEGG_2019_Human")


## CR01 diff q9
anno <- annotatePeak(diffpeaks.CR01.log2.q9, tssRegion=c(-500, 500), TxDb=txdb, annoDb="org.Hs.eg.db")
# filter for only promoters
idx <- which(anno@anno@elementMetadata@listData[["annotation"]] == "Promoter")
#scores <- anno@anno@elementMetadata@listData[["score"]]
genes <- anno@anno@elementMetadata@listData[["SYMBOL"]][idx]
genes <- unique(genes)
enriched.CR01.diffpeaks.q9 <- enrichr(genes, dbs)
#head(enriched[["GO_Molecular_Function_2018"]][["Term"]])

x <- enriched.CR01.diffpeaks.q9[["GO_Biological_Process_2018"]]

### reformat the df for putting on slides
new.df <- data.frame(term = x[,1], overlap = x[,2], pvalue = x[,3], genes = x[,9])

## CR01 diff q9
anno <- annotatePeak(diffpeaks.CR01.log2.q9, tssRegion=c(-500, 500), TxDb=txdb, annoDb="org.Hs.eg.db")
idx <- which(anno@anno@elementMetadata@listData[["annotation"]] == "Promoter")
genes <- anno@anno@elementMetadata@listData[["SYMBOL"]]
genes = genes[idx]
genes <- unique(genes)
enriched <- enrichr(genes, dbs)
x <- enriched[["GO_Biological_Process_2018"]]
new.df <- data.frame(term = x[,1], overlap = x[,2], pvalue = x[,3], adjustedpvalue = x[,4], genes = x[,9])

## CR01 diff q9
anno <- annotatePeak(diffpeaks.CR02.log2.q9, tssRegion=c(-500, 500), TxDb=txdb, annoDb="org.Hs.eg.db")
idx <- which(anno@anno@elementMetadata@listData[["annotation"]] == "Promoter")
genes <- anno@anno@elementMetadata@listData[["SYMBOL"]]
genes = genes[idx]
genes <- unique(genes)
enriched <- enrichr(genes, dbs)
x <- enriched[["GO_Biological_Process_2018"]]
new.df <- data.frame(term = x[,1], overlap = x[,2], pvalue = x[,3], adjustedpvalue = x[,4], genes = x[,9])

## CR01 diff q9
anno <- annotatePeak(diffpeaks.CR04.log2.q9, tssRegion=c(-500, 500), TxDb=txdb, annoDb="org.Hs.eg.db")
idx <- which(anno@anno@elementMetadata@listData[["annotation"]] == "Promoter")
genes <- anno@anno@elementMetadata@listData[["SYMBOL"]]
genes = genes[idx]
genes <- unique(genes)
enriched <- enrichr(genes, dbs)
x <- enriched[["GO_Biological_Process_2018"]]
new.df <- data.frame(term = x[,1], overlap = x[,2], pvalue = x[,3], adjustedpvalue = x[,4], genes = x[,9])

## CR01 diff q9
anno <- annotatePeak(diffpeaks.CR05.log2.q9, tssRegion=c(-500, 500), TxDb=txdb, annoDb="org.Hs.eg.db")
idx <- which(anno@anno@elementMetadata@listData[["annotation"]] == "Promoter")
genes <- anno@anno@elementMetadata@listData[["SYMBOL"]]
genes = genes[idx]
genes <- unique(genes)
enriched <- enrichr(genes, dbs)
x <- enriched[["GO_Biological_Process_2018"]]
new.df <- data.frame(term = x[,1], overlap = x[,2], pvalue = x[,3], adjustedpvalue = x[,4], genes = x[,9])

## CR01 diff q9
anno <- annotatePeak(diffpeaks.CR07.log2.q9, tssRegion=c(-500, 500), TxDb=txdb, annoDb="org.Hs.eg.db")
idx <- which(anno@anno@elementMetadata@listData[["annotation"]] == "Promoter")
genes <- anno@anno@elementMetadata@listData[["SYMBOL"]]
genes = genes[idx]
genes <- unique(genes)
enriched <- enrichr(genes, dbs)
x <- enriched[["GO_Biological_Process_2018"]]
new.df <- data.frame(term = x[,1], overlap = x[,2], pvalue = x[,3], adjustedpvalue = x[,4], genes = x[,9])

## CR01 diff q9
anno <- annotatePeak(diffpeaks.CR08.log2.q9, tssRegion=c(-500, 500), TxDb=txdb, annoDb="org.Hs.eg.db")
idx <- which(anno@anno@elementMetadata@listData[["annotation"]] == "Promoter")
genes <- anno@anno@elementMetadata@listData[["SYMBOL"]]
genes = genes[idx]
genes <- unique(genes)
enriched <- enrichr(genes, dbs)
x <- enriched[["GO_Biological_Process_2018"]]
new.df <- data.frame(term = x[,1], overlap = x[,2], pvalue = x[,3], adjustedpvalue = x[,4], genes = x[,9])




```

Correct Tn5 insertion bias
```{r}

refBEDpath <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\footprints\\hg38.chr1.bed"
refFastaPath <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\footprints\\chr1.fa"
bamPath <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bam\\LNCaP-WT-01.rep1.refhg38.bam"
biasedPlotPath <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\footprints\\wt01.biased.plot.svg"
correctedPlotPath <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\footprints\\wt01.corrected.plot.svg"
biasModelPath <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\footprints\\wt01.tn5model.yml"
  
generateTn5BiasCorrectionModel(refFastaPath, bamPath, refBEDpath, biasedPlotPath, correctedPlotPath, biasModelPath)

```

Use peaks to modify ARACNe networks
```{r}
#### Load the regulon ####
data(regulonprad)
regulonPRADentrez <- regulonprad
regulonPRADsymbol <- entrez2gene.regulon(regulonPRADentrez)

#### Load the counts ####
ccleCountsPath <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\counts\\cclecounts.rda"
load(ccleCountsPath)
a1 <- cclecounts[[1]]
a2 <- cclecounts[[2]]
a3 <- cclecounts[[3]]
a4 <- cclecounts[[4]]
a5 <- cclecounts[[5]]
a6 <- cclecounts[[6]]
a7 <- cclecounts[[7]]
a8 <- cclecounts[[8]]
a9 <- cclecounts[[9]]
a10 <- cclecounts[[10]]
a11 <- cclecounts[[11]]
a12 <- cclecounts[[12]]
a13 <- cclecounts[[13]]
a14 <- cclecounts[[14]]
a15 <- cclecounts[[15]]
a16 <- cclecounts[[16]]
a17 <- cclecounts[[17]]
a18 <- cclecounts[[18]]
a19 <- cclecounts[[19]]
a20 <- cclecounts[[20]]
a21 <- cclecounts[[21]]
a22 <- cclecounts[[22]]
a23 <- cclecounts[[23]]
a24 <- cclecounts[[24]]
a25 <- cclecounts[[25]]
x <- cbind(a1,a2,a3,a4,a5,a6,a7,a8,a9,a10,a11,a12,a13,a14,a15,a16,a17,a18,a19,a20,a21,a22,a23,a24,a25)
### normalize
vstCountsCCLE <- vst(x)
##  convert to symbols
vstCountsCCLESymbol <- vstCountsCCLE
entrez <- rownames(vstCountsCCLE)
symb <- entrez2geneSymbol.old(entrez)
rownames(vstCountsCCLESymbol) <- symb

```

Method 2
```{r}
#### Load the regulon ####
data(regulonprad)
regulonPRADentrez <- regulonprad
regulonPRADsymbol <- entrez2gene.regulon(regulonPRADentrez)

#### Load the counts and convert to symbols ####
ccleCountsPath <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\counts\\cclecounts.rda"
load(ccleCountsPath)
pradCountsEntrez <- cclecounts[["prostate_bat1"]]
entrez <- rownames(pradCountsEntrez)
symb <- entrez2geneSymbol.old(entrez)
pradCountsSymbol <- pradCountsEntrez
rownames(pradCountsSymbol) <- symb

#### run viper ####
pradVIPER <- viper(pradCountsEntrez, regulonPRADsymbol)
idx <- which(rownames(pradVIPER) == "AR")
AR <- pradVIPER[idx,6]
   
#### Modify with peaks, then rerun ####
#### Get the gene symbol annotations of an input peak list
WTmergedq9 <- transformPeak(WT.p0001.merged, 0.9)
anno <- annotatePeak(WT01.p0001.summit, tssRegion=c(-500, 500), TxDb=txdb, annoDb="org.Hs.eg.db")
idx <- which(anno@anno@elementMetadata@listData[["annotation"]] == "Promoter")
#scores <- anno@anno@elementMetadata@listData[["score"]]
genes <- anno@anno@elementMetadata@listData[["SYMBOL"]][idx]
genes <- unique(genes)

#### Get all targets of all regulators in the interactome ####
regulatorTargets <- c(names(regulonPRADsymbol), unlist(lapply(regulonPRADsymbol, function(x) names(x$tfmode)), use.names = FALSE))
regulatorTargets <- unique(regulatorTargets)

#### Determine which targets are represented in the peaks set
idx <- which(genes %in% regulatorTargets)

## Get a list of all genes not included in peaks
keepGenes <- peakGenes[idx]
removeGenes <- peakGenes[-idx]

## remove expression data for all genes not in peaks
testCounts <- pradCountsSymbol[rownames(pradCountsSymbol) %in% unique(keepGenes),]

## rerun viper with new counts matrix
pradVIPERpeaks <- viper(testCounts, regulonPRADsymbol)
idx <- which(rownames(pradVIPERpeaks) == "AR")
AR <- pradVIPERpeaks[idx,6]



# filter for only promoters
idx <- which(anno@anno@elementMetadata@listData[["annotation"]] == "Promoter")
#scores <- anno@anno@elementMetadata@listData[["score"]]
genes <- anno@anno@elementMetadata@listData[["SYMBOL"]][idx]
genes <- unique(genes)
enriched.CR01.diffpeaks.q9 <- enrichr(genes, dbs)
#head(enriched[["GO_Molecular_Function_2018"]][["Term"]])

```

Run viper with unmodified data
```{r}
## entrez
pradVIPER <- viper(vstCountsCCLE, regulonPRADentrez)
dev.new()
plot(pradVIPER, cex=.7)
dev.off()

x <- colnames(pradVIPER)
lncapidx <- 750
y <- pradVIPER[,750]

## symbol
entrez <- names(y)
symb <- entrez2geneSymbol.old(entrez)
ySymbol <- y
names(ySymbol) <- symb

ccleCountsPath <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\counts\\cclecounts.rda"
load(ccleCountsPath)
idx <- c(739,734,735,736)
lncapMatrix <- ccle[,idx]
refMatrix <- ccle[,-idx]
```

Find peak overlap with ARACNe targets
```{r}

#### Get the gene symbol annotations of an input peak list
peakGenes <- getSymbolFromGR(WT01.p0001.summit.log2.q9)
peakGenes <- unique(peakGenes)

#### Get all targets of all regulators in the interactome
regulatorTargets <- c(names(regulonPRADsymbol), unlist(lapply(regulonPRADsymbol, function(x) names(x$tfmode)), use.names = FALSE))
regulatorTargets <- unique(regulatorTargets)

#### Determine which targets are represented in the peaks set
idx <- which(peakGenes %in% regulatorTargets)

## Get a list of all genes not included in peaks
keepGenes <- peakGenes[idx]
removeGenes <- peakGenes[-idx]

## remove expression data for all genes not in peaks
testCounts <- vstCountsCCLESymbol[rownames(vstCountsCCLESymbol) %in% unique(keepGenes),]

## rerun viper with new counts matrix
pradVIPER <- viper(testCounts, regulonPRADsymbol)
t <- pradVIPER[,750]

```

Run VIPER again with modified regulons
```{r}

## Test viper before and after
testVIPER <- viper(testCounts, regulonPRADsymbol)
dev.new()
plot(testVIPER, cex=.7)
dev.off()

id <- which(rownames(testVIPER) == "AR")

testVIPER[266,6]
pradVIPER[266,6]

```

Find ARACNe footprint overlap
```{r}

##
#if (!requireNamespace("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")
#BiocManager::install("aracne.networks", version = "3.8")
#BiocManager::install("org.Hs.eg.db")
#BiocManager::install("annotate")
#BiocManager::install("viper")
#BiocManager::install("AnnotationDbi")
#BiocManager::install("mygene")
#install.packages("rlist")

##
cat("Loading libraries...", "\n")
suppressMessages(library(aracne.networks))
suppressMessages(library(org.Hs.eg.db))
suppressMessages(library(annotate))
suppressMessages(library(rlist))
suppressMessages(library(viper))
suppressMessages(library(GenomicRanges))
suppressMessages(library(mygene))

##
cat("Setting snakemake vars...", "\n")
inputfile <- snakemake@input[[1]]
outputfile <- snakemake@output[[1]]
entrezid <- snakemake@wildcards[["entrez"]]

## load file
load(inputfile)

## get binding sites of merged signals
sites <- mergedMotifs[["sites"]]

## get the ARACNe interactome
coad_interactome <- aracne.networks::reguloncoad

## get the targets
com <- paste0("targets <- names(coad_interactome[['", entrezid, "']][['tfmode']])")
eval(parse(text = com))


## Retrieve info for the network targets
target_list <- list()
for (a in 1:length(targets)){
  target_list[a] <- mygene::getGene(geneid = targets[a], fields = "all")}


## count the number of gene locations we have
loc_count <- 0
for (a in 1:length(targets)){
  if (is.null(target_list[[a]][["genomic_pos"]])){next} else {
    if (is.list(target_list[[a]][["genomic_pos"]][[1]])){
      loc_count <- (loc_count + length(target_list[[a]][["genomic_pos"]]))
    } else {
      loc_count <- (loc_count + 1)}}}


## Retrieve the genomic coordinates of all network targets
## retrieve all genomic coords here, can later trim non standard ones more easily with GRanges functions
target_locations <- matrix(data = NA, nrow = loc_count, ncol = 4)
colnames(target_locations) <- c("gene", "chr", "start", "end")
idx <- 1
#
for (a in 1:length(target_list)){
  
  if (is.null(target_list[[a]][["genomic_pos"]])){next} else {
    
    if (is.list(target_list[[a]][["genomic_pos"]][[1]])){
      
      for (b in 1:length(target_list[[a]][["genomic_pos"]])){
        target_locations[idx,1] <- target_list[[a]][["symbol"]]
        target_locations[idx,2] <- paste0("chr", target_list[[a]][["genomic_pos"]][[b]][["chr"]])
        target_locations[idx,3] <- target_list[[a]][["genomic_pos"]][[b]][["start"]]
        target_locations[idx,4] <- target_list[[a]][["genomic_pos"]][[b]][["end"]]
        idx <- (idx+1)
      } # end for (b in 1:length(target_list[[a]][["genomic_pos"]]))
      
    } else {
      
      target_locations[idx,1] <- target_list[[a]][["symbol"]]
      target_locations[idx,2] <- paste0("chr", target_list[[a]][["genomic_pos"]][["chr"]])
      target_locations[idx,3] <- target_list[[a]][["genomic_pos"]][["start"]]
      target_locations[idx,4] <- target_list[[a]][["genomic_pos"]][["end"]]
      idx <- (idx+1)}}}


## Convert the genomic locations of the targets into GRanges
gr <- GRanges(
  seqnames = target_locations[,2],
  ranges = IRanges(start = as.numeric(target_locations[,3]), end = as.numeric(target_locations[,4])))
#strand = c(rep("+", times = num_names)),
#seqinfo = Seqinfo(genome="hg38"),
#score = c(rep(1, times = num_names)))

## prune to standard xsomes
gr <- keepStandardChromosomes(gr, pruning.mode="coarse")

gr1 <- gr
gr2 <- gr
#
start(gr1) <- (start(gr1) - 2000)
width(gr1) <- (width(gr1) + 2000)
#
start(gr2) <- (start(gr2) - 50000)
width(gr2) <- (width(gr2) + 50000)


##
#intersection <- intersect(gr, wg_sites)

## Find overlaps
## YOU WILL WANT TO FIND THE OVERLAPS FOR A RANGE OF EXTENDED VALUES PAST THE TARGETS, GRAPH IT
overlaps1 <- findOverlaps(gr1, sites)
overlaps2 <- findOverlaps(gr2, sites)

info <- list()
info$overlap1 <- overlaps1
info$overlap2 <- overlaps2

save(info, file = outputfile)






```

Integrate footprints
```{r}


ggplot(MRdata, aes(peak.log2Depth, peak.log2Flank, color = viperNES)) + 
  geom_point() + 
  scale_color_gradient(low="blue", high="red") +
  ggtitle(paste0("COAD MR, Depth vs Flank, VIPER NES ", cellName)) +
  geom_label_repel(aes(label = Gene),
                   box.padding   = 0.35, 
                   point.padding = 0.5,
                   segment.color = 'black',
                   label.size = NA)

#### PROMOTERS
ggplot(MRdata, aes(promoterPeak.log2Depth, promoterPeak.log2Flank, color = log2(RNAexp))) + 
  geom_point() + 
  scale_color_gradient(low="blue", high="red") +
  ggtitle(paste0("Promoters COAD MR, Depth vs Flank, RNA exp ", cellName)) +
  geom_label_repel(aes(label = Gene),
                   box.padding   = 0.35, 
                   point.padding = 0.5,
                   segment.color = 'black',
                   label.size = NA)

ggplot(MRdata, aes(promoterPeak.log2Depth, promoterPeak.log2Flank, color = viperNES)) + 
  geom_point() + 
  scale_color_gradient(low="blue", high="red") +
  ggtitle(paste0("Promoters COAD MR, Depth vs Flank, VIPER NES ", cellName)) +
  geom_label_repel(aes(label = Gene),
                   box.padding   = 0.35, 
                   point.padding = 0.5,
                   segment.color = 'black',
                   label.size = NA)


### Bound sites
ggplot(MRdata, aes(bound.log2Depth, bound.log2Flank, color = log2(RNAexp))) + 
  geom_point() + 
  scale_color_gradient(low="blue", high="red") +
  ggtitle(paste0("Bound COAD MR, Depth vs Flank, RNA exp ", cellName)) +
  geom_label_repel(aes(label = Gene),
                   box.padding   = 0.35, 
                   point.padding = 0.5,
                   segment.color = 'black',
                   label.size = NA)

ggplot(MRdata, aes(bound.log2Depth, bound.log2Flank, color = viperNES)) + 
  geom_point() + 
  scale_color_gradient(low="blue", high="red") +
  ggtitle(paste0("Bound COAD MR, Depth vs Flank, VIPER NES ", cellName)) +
  geom_label_repel(aes(label = Gene),
                   box.padding   = 0.35, 
                   point.padding = 0.5,
                   segment.color = 'black',
                   label.size = NA)

## RNA exp vs Flank
ggplot(footprintData, aes(peak.log2Flank, log2(RNAexp))) + 
  geom_point() +
  stat_cor(method = "spearman") +
  ggtitle(paste0("RNA exp vs Flank ", cellName)) +
  geom_smooth(method ="lm", se = FALSE) # se controls the confidence band


## VIPER vs Depth
ggplot(footprintData, aes(peak.log2Depth, viperNES)) + 
  geom_point() +
  stat_cor(method = "spearman") +
  ggtitle(paste0("VIPER vs Depth ", cellName)) +
  geom_smooth(method ="lm", se = FALSE) # se controls the confidence band


## VIPER vs Flank
ggplot(footprintData, aes(peak.log2Flank, viperNES)) + 
  geom_point() +
  stat_cor(method = "spearman") +
  ggtitle(paste0("VIPER vs Flank ", cellName)) +
  geom_smooth(method ="lm", se = FALSE) # se controls the confidence band


#### PROMOTERS AND DISTAL ####

## Promoter sites, RNA exp
ggplot(footprintData, aes(promoterPeak.log2Depth, promoterPeak.log2Flank, color = log2(RNAexp))) + 
  geom_point() + 
  scale_color_gradient(low="blue", high="red") +
  ggtitle(paste0("Promoter sites, Depth vs Flank, RNA exp ", cellName)) +
  geom_smooth(method ="lm", se = FALSE) # se controls the confidence band

## Promoter sites, VIPER
ggplot(footprintData, aes(promoterPeak.log2Depth, promoterPeak.log2Flank, color = viperNES)) + 
  geom_point() + 
  scale_color_gradient(low="blue", high="red") +
  ggtitle(paste0("Promoter sites, Depth vs Flank, VIPER NES ", cellName)) +
  geom_smooth(method ="lm", se = FALSE) # se controls the confidence band


## Promoter sites, RNA exp vs Depth
ggplot(footprintData, aes(promoterPeak.log2Depth, log2(RNAexp))) + 
  geom_point() +
  stat_cor(method = "spearman") +
  ggtitle(paste0("Promoter sites, RNA exp vs Depth ", cellName)) +
  geom_smooth(method ="lm", se = FALSE) # se controls the confidence band


## Promoter sites, RNA exp vs Flank
ggplot(footprintData, aes(promoterPeak.log2Flank, log2(RNAexp))) + 
  geom_point() +
  stat_cor(method = "spearman") +
  ggtitle(paste0("Promoter sites, RNA exp vs Flank ", cellName)) +
  geom_smooth(method ="lm", se = FALSE) # se controls the confidence band


## Promoter sites, VIPER vs Depth
ggplot(footprintData, aes(promoterPeak.log2Depth, viperNES)) + 
  geom_point() +
  stat_cor(method = "spearman") +
  ggtitle(paste0("Promoter sites, VIPER NES vs Depth ", cellName)) +
  geom_smooth(method ="lm", se = FALSE) # se controls the confidence band


## Promoter sites, VIPER vs Flank
ggplot(footprintData, aes(promoterPeak.log2Flank, viperNES)) + 
  geom_point() +
  stat_cor(method = "spearman") +
  ggtitle(paste0("Promoter sites, VIPER NES vs Flank ", cellName)) +
  geom_smooth(method ="lm", se = FALSE) # se controls the confidence band


## Distal
## Distal sites, RNA exp
ggplot(footprintData, aes(distalPeak.log2Depth, distalPeak.log2Flank, color = log2(RNAexp))) + 
  geom_point() + 
  scale_color_gradient(low="blue", high="red") +
  ggtitle(paste0("Distal sites, Depth vs Flank, RNA exp ", cellName)) +
  geom_smooth(method ="lm", se = FALSE) # se controls the confidence band


## Distal sites, VIPER
ggplot(footprintData, aes(distalPeak.log2Depth, distalPeak.log2Flank, color = viperNES)) + 
  geom_point() + 
  scale_color_gradient(low="blue", high="red") +
  ggtitle(paste0("Distal sites, Depth vs Flank, VIPER NES ", cellName)) +
  geom_smooth(method ="lm", se = FALSE) # se controls the confidence band


## Distal sites, RNA exp vs Depth
ggplot(footprintData, aes(distalPeak.log2Depth, log2(RNAexp))) + 
  geom_point() +
  stat_cor(method = "spearman") +
  ggtitle(paste0("Distal sites, RNA exp vs Depth ", cellName)) +
  geom_smooth(method ="lm", se = FALSE) # se controls the confidence band


## Distal sites, RNA exp vs Flank
ggplot(footprintData, aes(distalPeak.log2Flank, log2(RNAexp))) + 
  geom_point() +
  stat_cor(method = "spearman") +
  ggtitle(paste0("Distal sites, RNA exp vs Flank ", cellName)) +
  geom_smooth(method ="lm", se = FALSE) # se controls the confidence band


## Distal sites, VIPER vs Depth
ggplot(footprintData, aes(distalPeak.log2Depth, viperNES)) + 
  geom_point() +
  stat_cor(method = "spearman") +
  ggtitle(paste0("Distal sites, VIPER NES vs Depth ", cellName)) +
  geom_smooth(method ="lm", se = FALSE) # se controls the confidence band


## Distal sites, VIPER vs Flank
ggplot(footprintData, aes(distalPeak.log2Flank, viperNES)) + 
  geom_point() +
  stat_cor(method = "spearman") +
  ggtitle(paste0("Distal sites, VIPER NES vs Flank ", cellName)) +
  geom_smooth(method ="lm", se = FALSE) # se controls the confidence band


## Bound promoter
## Bound Promoter sites, RNA exp
ggplot(footprintData, aes(promoterBound.log2Depth, promoterBound.log2Flank, color = log2(RNAexp))) + 
  geom_point() + 
  scale_color_gradient(low="blue", high="red") +
  stat_cor(method = "spearman") +
  ggtitle(paste0("Bound Promoter sites, Depth vs Flank, RNA exp ", cellName)) +
  geom_smooth(method ="lm", se = FALSE) # se controls the confidence band


## Bound Promoter sites, VIPER
ggplot(footprintData, aes(promoterBound.log2Depth, promoterBound.log2Flank, color = viperNES)) + 
  geom_point() + 
  scale_color_gradient(low="blue", high="red") +
  stat_cor(method = "spearman") +
  ggtitle(paste0("Bound Promoter sites, Depth vs Flank, VIPER NES ", cellName)) +
  geom_smooth(method ="lm", se = FALSE) # se controls the confidence band


## Bound Distal
## Bound Distal sites, RNA exp
ggplot(footprintData, aes(distalBound.log2Depth, distalBound.log2Flank, color = log2(RNAexp))) + 
  geom_point() + 
  scale_color_gradient(low="blue", high="red") +
  stat_cor(method = "spearman") +
  ggtitle(paste0("Bound Distal sites, Depth vs Flank, RNA exp ", cellName)) +
  geom_smooth(method ="lm", se = FALSE) # se controls the confidence band

## Bound Distal sites, VIPER
ggplot(footprintData, aes(distalBound.log2Depth, distalBound.log2Flank, color = viperNES)) + 
  geom_point() + 
  scale_color_gradient(low="blue", high="red") +
  stat_cor(method = "spearman") +
  ggtitle(paste0("Bound Distal sites, Depth vs Flank, VIPER NES ", cellName)) +
  geom_smooth(method ="lm", se = FALSE) # se controls the confidence band


####### PLOT COAD MRs, ALL THREE CELL LINES IN SAME PLOT ########

coadmr <- c("CDX2", "TCF7", "MNX1", "POU5F1B", "ESRRA", "HNF4A", "GMEB2", "HOXA3",
            "OVOL1", "ASCL2", "ZSWIM1", "CBFA2T2", "PAX6", "ADNP", "TAF4", "ZMYND8", "ZNF696")

mrIdx <- which(footprintData[,1] %in% coadmr)

## Create new dataframe
MRdata <- data.frame(matrix(vector(), 0, 10,
                            dimnames=list(c(), c(
                              "Gene", "Type", "peak.log2Depth", "peak.log2Flank", "RNAexp", "viperNES",
                              "promoterPeak.log2Depth", "promoterPeak.log2Flank", "bound.log2Depth", "bound.log2Flank"))),
                     stringsAsFactors=F)

## Transfer info to new dataframe
numPoints <- length(mrIdx)
cellLine <- "LS1034"
cellName <- "LS1034"

## Xfer data
for (d in 1:numPoints){
  
  Idx <- mrIdx[d]
  
  MRdata[d,1] <- footprintData[Idx,1]
  MRdata[d,2] <- cellLine
  MRdata[d,3] <- footprintData[Idx,"peak.log2Depth"]
  MRdata[d,4] <- footprintData[Idx,"peak.log2Flank"]
  MRdata[d,5] <- footprintData[Idx,"RNAexp"]
  MRdata[d,6] <- footprintData[Idx,"viperNES"]
  MRdata[d,7] <- footprintData[Idx,"promoterPeak.log2Depth"]
  MRdata[d,8] <- footprintData[Idx,"promoterPeak.log2Flank"]
  MRdata[d,9] <- footprintData[Idx,"bound.log2Depth"]
  MRdata[d,10] <- footprintData[Idx,"bound.log2Flank"]
}


#### COAD MR PLOTS ##########################################################################################

## All sites, VIPER
ggplot(footprintData, aes(peak.log2Depth, peak.log2Flank, color = viperNES)) + 
  geom_point() + 
  scale_color_gradient(low="blue", high="red") +
  ggtitle(paste0("All sites, Depth vs Flank, VIPER NES ", cellName))


## Bound ratio vs Exp
ggplot(footprintData, aes(boundRatio, log2(RNAexp))) + 
  geom_point() +
  stat_cor(method = "spearman") +
  ggtitle(paste0("Percent bound sites, RNA exp ", cellName))


## Bound ratio vs VIPER
ggplot(footprintData, aes(boundRatio, viperNES)) + 
  geom_point() +
  stat_cor(method = "spearman") +
  ggtitle(paste0("Percent bound sites, VIPER NES ", cellName))



#### ADD RNA EXPRESSION VALUES TO THE DF ####
## Load the RNA expression data from ccle
load("~/git/atacPipelineMaster/atacVIPER/expressionData/ccle/cclecounts.rda")
coadCounts <- cclecounts[["large_intestine_bat1"]]

## Set expression for current cell line
snu61Exp <- coadCounts[,"SNU-61"]
h508Exp <- coadCounts[,"NCI-H508"]
mdst8Exp <- coadCounts[,"MDST8"]
ls1034Exp <- coadCounts[,"LS1034"]

##
curExp <- ls1034Exp
geneList <- names(curExp)
mappings <- queryMany(geneList, scopes="entrezgene", fields="symbol", species="human")
genemaps <- mappings@listData[["symbol"]]

## Find index of RNA expression values for TFs in this dataset
geneIdx <- which(genemaps %in% footprintData[,"Gene"])

## Add the RNA expression value of each TF to the dataframe
addExp <- c()
for (a in 1:numGenes){
  ## Get the mapping index corresponding to current gene
  curMap <- which(genemaps %in% footprintData[a,1])
  ## If it doesnt exist, set RNAexp to 0
  if (length(curMap) == 0){
    footprintData[a,52] <- 0
  } else {
    footprintData[a,52] <- curExp[[curMap]]
  }} # end for (a in 1:numGenes)


#### ADD VIPER NES SCORE VALUES TO THE DF ####
load("~/git/atacPipelineMaster/atacVIPER/regulons/coad-tcga-regulon.rda")
curVIPER <- msviper(curExp, regul)
curNES <- curVIPER[["es"]][["nes"]]
viperNames <- names(curNES)
viperMap <- queryMany(viperNames, scopes="entrezgene", fields="symbol", species="human")
viperSym <- viperMap@listData[["symbol"]]
viperIdx <- which(viperSym %in% footprintData[,"Gene"])

## Add the VIEPR NES value of each TF to the dataframe
addExp <- c()
for (b in 1:numGenes){
  ## Get the mapping index corresponding to current gene
  curMap <- which(viperSym %in% footprintData[b,1])
  ## If it doesnt exist, set RNAexp to 0
  if (length(curMap) == 0){
    footprintData[b,53] <- 0
  } else {
    footprintData[b,53] <- curNES[[curMap]]
  }} # end for (b in 1:numGenes)


## Transfer to new
LS1034_data <- footprintData
save(LS1034_data, file = "C:\\Users\\jsk33\\Desktop\\LS1034_data.Rdata")

## Set cell line name
cellName <- "LS1034"
footprintData <- LS1034_data

#### GENERATE PLOTS #########################################################################################

## Load libraries and setup
suppressMessages(library(ggplot2))
suppressMessages(library(ggsci))
suppressMessages(library(mygene))
suppressMessages(library(viper))
suppressMessages(library(ggrepel))
suppressMessages(library(ggpubr))
suppressMessages(library(TxDb.Hsapiens.UCSC.hg38.knownGene))
suppressMessages(library(aracne.networks))
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
##
options(warn = -1)
options(scipen = 999)

#### GET THE DATA ####
load("C://Users//jsk33//Desktop//aggregate data//H508A-WT-02.aggregated.Rdata") # for testing only
footprintData <- aggregateData[["aggregateFootprintData"]]
footprintMetrics <- aggregateData[["aggregateFootprintMetricsData"]]
numGenes <- length(footprintData[,1])
## Load the regulon
load("~/git/atacPipelineMaster/atacVIPER/regulons/coad-tcga-regulon.rda")
coadInteractome <- regul
## Get symbols of regulator
coadRegulators <- names(coadInteractome)
coadMappings <- queryMany(coadRegulators, scopes="entrezgene", fields="symbol", species="human")
coadMaps <- coadMappings@listData[["symbol"]]
## Which of our TFs with footprints are in the interactome?
matchedRegulatorIdx <- which(coadMaps %in% names(footprintMetrics))
numMatched <- length(matchedRegulatorIdx)

#### CODE TESTING - PROMOTER/DISTAL groups ####
## Pull promoters from txdb, define promoter region as -1000/+100 in accordance with TCGA paper
## This modified what metadata is returned from the Txdb database regarding the promoters
#cols <- c("gene_id", "tx_id", "tx_name", "tx_chrom", "tx_strand")
cols <- c("gene_id")
promoters <- promoters(txdb, upstream = 1000, downstream = 500, use.names = TRUE, columns = cols)
## Trim the GRanges object to keep standard entries only
scope <- paste0("chr", c(1:22, "X", "Y"))
promoters <- keepStandardChromosomes(promoters, pruning.mode="coarse")
promoters <- keepSeqlevels(promoters, scope, pruning.mode="coarse")
#### "unlisting the returned geneids" 
## might need to look at this more closely at some point
## use this code to get a flattened list of the Entrez IDs
#za <- as.matrix(a@listData[["gene_id"]])
promoterIDs <- drop(promoters@elementMetadata@listData[["gene_id"]])


#### Setup a loop to iterate through all regulators also in our ATAC dataset
overlapData <- matrix(data = NA, ncol = 5, nrow = numMatched)
colnames(overlapData) <- c("Total targets", "Peak overlaps", "Bound site overlaps", "Unbound site overlaps", "Percent total overlap")

for (a in 1:numMatched){
  
  tryCatch({
    
    ## Get the name of the current matched regulator
    curName <- coadMaps[matchedRegulatorIdx[a]]
    
    ## Translate that to Entrez ID
    curNameIdx <- which(coadMaps == curName)
    curEntrezID <- coadRegulators[curNameIdx]
    
    ## Pull the targets of the current gene using the Entez ID
    ## For each regulator that has a cooresponding TF footprint,
    ## calculate what percent of the targets are captured by the motif binding sites
    ## Will need to grab the promoter region - extend TSS from -1000/+500 for each gene
    com <- paste0("curTargets <- names(coadInteractome[['", curEntrezID, "']][['tfmode']])")
    eval(parse(text = com))
    
    ## Get the Granges from the peaks of TF footprints for current gene
    com <- paste0("curFootprintsPeaks <- footprintMetrics[['", curName, "']][['peakSites']]")
    eval(parse(text = com))
    ## Get bound ranges
    com <- paste0("curFootprintsBound <- footprintMetrics[['", curName, "']][['boundSites']]")
    eval(parse(text = com))
    ## Get unbound ranges
    com <- paste0("curFootprintsUnbound <- footprintMetrics[['", curName, "']][['unboundSites']]")
    eval(parse(text = com))
    
    ## Which of the returned promoter ranges match to the current target genes? 
    targetPromoterIdx <- which(promoterIDs %in% curTargets)
    ## Subset that Granges for target promoters
    targetPromoters <- promoters[targetPromoterIdx]
    
    ## Find which of the ATAC peak motifs overlap with promoters of any targets for current gene
    targetPeakOverlaps <- findOverlaps(targetPromoters, curFootprintsPeaks, ignore.strand = TRUE)
    targetBoundOverlaps <- findOverlaps(targetPromoters, curFootprintsBound, ignore.strand = TRUE)
    targetUnboundOverlaps <- findOverlaps(targetPromoters, curFootprintsUnbound, ignore.strand = TRUE)
    
    ## Record the extent of the overlap
    overlapData[a,1] <- length(targetPromoters)
    overlapData[a,2] <- length(unique(targetPeakOverlaps@to))
    overlapData[a,3] <- length(unique(targetBoundOverlaps@to))
    overlapData[a,4] <- length(unique(targetUnboundOverlaps@to))
    overlapData[a,5] <- (overlapData[a,2] / overlapData[a,1])*100
    
  }, # end try
  error=function(cond){
    message(cond)
    return(NA)
  },
  finally={})
}

## Plot the overlap data
#plot(density(overlapData[,3]))
#plot(hist(overlapData[,3]))

df <- data.frame(overlapData[,5])
#stip plot
p <- ggplot(df, aes(x=1, y=df[,1])) + geom_jitter(position=position_jitter(0.3)) + theme_bw()
p

p <- ggplot(df, aes(x=1, y=df[,1])) + geom_violin(trim=FALSE) + geom_boxplot(width=0.1) + theme_classic()
p

## Modify the COAD interactome according to Footprint data
data("reguloncoad")
write.regulon(regulonblca,file="C:\Users\jsk33\Desktop\aggregate data")


#### MODIFY THE COAD REGULON ####
#### THE FOLLOWING CODE IS ADAPTED FROM CODE ORIGINALLY WRITTEN BY AJAY NAIR ####


# save(dset,regulon_motifMI, file = "dset_motifMIregulon.RData")

load(file = "dset_motifMIregulon.RData")#get the dset and regulon_motifMI
geneId <- "4089" #smad4
interactome_tinker <- read.table(file = "network_coad_tfCotf_3col_sorted.txt", sep = "\t")
regulon_req <- interactome_tinker[which(interactome_tinker[,1]==geneId),]#getting req regulon to tinker
interactome_tinker <- interactome_tinker[-(which(interactome_tinker[,1]==geneId)),]# removing the req regulon from interactome
colnames(regulon_motifMI) <- colnames(regulon_req)
regulon_req <- rbind(regulon_req,regulon_motifMI)
regulon_req <- regulon_req[order(regulon_req[,3], decreasing = TRUE),]
write.table(interactome_tinker, file = "network_coad_tfCotf_3col_tinkered.txt", sep = "\t", append = FALSE, row.names = FALSE, col.names = FALSE)
write.table(regulon_req, file = "network_coad_tfCotf_3col_tinkered.txt", sep = "\t", append = TRUE, row.names = FALSE, col.names = FALSE)
regul <- aracne2regulon("network_coad_tfCotf_3col_tinkered.txt", dset, format = "3col", verbose = TRUE)#regulon
print(regul)
coadViper <- viper(dset, regul, method = "scale") #the viper algorithm
dim(coadViper)


#### THE ABOVE CODE IS ADAPTED FROM CODE ORIGINALLY WRITTEN BY AJAY NAIR ####

```

Classify Gene Accessibility
```{r}

## Install libraries if necessary
#source("https://bioconductor.org/biocLite.R")
#biocLite("stringi", suppressUpdates = TRUE)
#biocLite("GenomicRanges", suppressUpdates = TRUE)
#biocLite("ensembldb", suppressUpdates = TRUE)
#biocLite("EnsDb.Hsapiens.v86", suppressUpdates = TRUE)
#biocLite("S4Vectors", suppressUpdates = TRUE)
#biocLite("Repitools", suppressUpdates = TRUE)
#biocLite("TxDb.Hsapiens.UCSC.hg38.knownGene", suppressUpdates = TRUE)
#biocLite("genomation", suppressUpdates = TRUE)
#install.packages("ggplot2")

## Load libraries
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(genomation)
library(ensembldb)
library(EnsDb.Hsapiens.v86)
library(GenomicRanges)
library(S4Vectors)
library(Repitools)
library(ggplot2)
library(crayon)
library(Rsamtools)


## Shorten variable name for TxDb database
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene

## Define the scope
scope <- paste0("chr", c(1:22, "X", "Y"))


#### The input files

## First file is provided by Jeremy, contains RNAseq data
inputFile <- "C:\\Users\\jsk33\\Documents\\lab\\atac\\snu61_tfc"
expData <- read.table(inputFile, header=TRUE, sep = ',')

## Second file is called peaks from ATACseq data, in .bed format
bedFile <- "C:\\Users\\jsk33\\Documents\\lab\\atac\\atac\\snu61\\wt01\\peaks\\macs2\\merged\\SNU61-WT-01-merged_global_normalization_peaks.narrowPeak"
## Read in the .bed peaks file
snu61Peaks <- readBed(bedFile, track.line = FALSE, remove.unusual = FALSE, zero.based = TRUE)
## Keep standard ranges only
snu61Peaks <- keepStandardChromosomes(snu61Peaks, pruning.mode="coarse")
snu61Peaks <- keepSeqlevels(snu61Peaks, scope, pruning.mode="coarse")

## The input bam files, for calculating raw signals
inputBam <- "C:\\Users\\jsk33\\Documents\\lab\\atac\\atac\\snu61\\wt01\\preprocessing\\11repmerged\\SNU61-WT-01-repmerged.bam"


#### Retrieve gene information from TxDb

## Make a gene ID based key for retrieving data for genes of interest from expression matrix
geneKey <- c(as.character(expData[,3]))

## Use the select method to get mapping between tx_name (UCSC) and gene_id (ENTREZ)
annotData <- select(txdb, keys = geneKey, columns = "TXNAME", keytype = "GENEID")

## Make a vector with the tx_names
txNames <- c(annotData[,2])

## Get promoter regions, -200 bp from TSS (can be adjusted) for all transcripts from TxDb
promoters <- promoters(txdb, upstream = 200, downstream = 0)
## Trim the GRanges object to keep standard entries only
promoters <- keepStandardChromosomes(promoters, pruning.mode="coarse")
promoters <- keepSeqlevels(promoters, scope, pruning.mode="coarse")

## Subset the promoters GRanges object using the generated index
## Note that with multiple transcript variants, this number will be much higher than the gene_id list
promoterData <- promoters[promoters$tx_name %in% txNames]


#### Each hit may refer to multiple transcript variants. Subset based on the first hit for a given gene only
newGeneID <- c()
newTxName <- promoterData@elementMetadata@listData[["tx_name"]]
for (a in 1:length(newTxName)){
  idx <- which(newTxName[a] == annotData[,2])
  newGeneID[a] <- annotData[idx,1]}
## Check how many unique genes there are now
length(unique(newGeneID))
## Add the GeneIDs to the promoters GRanges
promoterData@elementMetadata@listData[["gene_id"]] <- c(newGeneID)


#### There may be duplicate entries for gene IDs in the promoterData Granges, so lets keep only one for each
unIdx <- c()
for (b in 1:length(geneKey)){
  tempIdx <- c(which(promoterData@elementMetadata@listData[["gene_id"]] == geneKey[b]))
  unIdx[b] <- tempIdx[1]}

## As a sanity check
length(unique(unIdx))
## Remove NAs
unIdx <- unIdx[which(is.na(unIdx) == FALSE)]
## Remove duplicate IDS
unIdx <- unique(unIdx)
## As a sanity check
length(unique(unIdx))


## Subset for unique only
uniquePromoters <- promoterData[unIdx]
## Check number of entries
length(unique(uniquePromoters@elementMetadata@listData[["gene_id"]]))


#### Find which gene promoters are accessible (have a peak) using the ATACseq data
overlaps <- findOverlaps(snu61Peaks, uniquePromoters)
## How many of the overlaps referend unique annotated promoters?
idxuq <- unique(overlaps@to)
## Get a GRanges of the genes that have a peak
genePeaks <- uniquePromoters[idxuq]
## How many of these are unique GeneIDs? 
length(unique(genePeaks@elementMetadata@listData[["gene_id"]]))
## Get the list of ENTREZ IDs for genes in this set with active promoters
activeGenePromoters <- unique(genePeaks@elementMetadata@listData[["gene_id"]])


#### Annotate all unique genes as having a peak ("RED") or not ("BLACK")
idx2 <- which(expData[,3] %in% activeGenePromoters)
expData$peak <- "BLACK"
expData$peak[idx2] <- "RED"
## Change -inf values to -11
idx3 <- which(expData[,6] == "-Inf")
expData[idx3,6] <- -11


#### Sort and plot the data
sortedExpData <- expData[order(expData$SNU61_LARGE_INTESTINE_log2, decreasing = FALSE),]
##
plot(sortedExpData[,6], col = sortedExpData$peak)
##
hist(sortedExpData[,6], breaks = 40)


#### Make bins of the genes based on log2 expression
## Plot a histogram/dotplot overlay of the percentage of genes in that bin with/without peak
binData <- list()
binData$"-11"$data <- sortedExpData[which(sortedExpData[,6] <= -10),]
binData$"-10"$data <- sortedExpData[which(sortedExpData[,6] <= -9 & sortedExpData[,6] > -10),]
binData$"-9"$data <- sortedExpData[which(sortedExpData[,6] <= -8 & sortedExpData[,6] > -9),]
binData$"-8"$data <- sortedExpData[which(sortedExpData[,6] <= -7 & sortedExpData[,6] > -8),]
binData$"-7"$data <- sortedExpData[which(sortedExpData[,6] <= -6 & sortedExpData[,6] > -7),]
binData$"-6"$data <- sortedExpData[which(sortedExpData[,6] <= -5 & sortedExpData[,6] > -6),]
binData$"-5"$data <- sortedExpData[which(sortedExpData[,6] <= -4 & sortedExpData[,6] > -5),]
binData$"-4"$data <- sortedExpData[which(sortedExpData[,6] <= -3 & sortedExpData[,6] > -4),]
binData$"-3"$data <- sortedExpData[which(sortedExpData[,6] <= -2 & sortedExpData[,6] > -3),]
binData$"-2"$data <- sortedExpData[which(sortedExpData[,6] <= -1 & sortedExpData[,6] > -2),]
binData$"-1"$data <- sortedExpData[which(sortedExpData[,6] <= 0 & sortedExpData[,6] > -1),]
binData$"0"$data <- sortedExpData[which(sortedExpData[,6] <= 1 & sortedExpData[,6] > 0),]
binData$"1"$data <- sortedExpData[which(sortedExpData[,6] <= 2 & sortedExpData[,6] > 1),]
binData$"2"$data <- sortedExpData[which(sortedExpData[,6] <= 3 & sortedExpData[,6] > 2),]
binData$"3"$data <- sortedExpData[which(sortedExpData[,6] <= 4 & sortedExpData[,6] > 3),]
binData$"4"$data <- sortedExpData[which(sortedExpData[,6] <= 5 & sortedExpData[,6] > 4),]
binData$"5"$data <- sortedExpData[which(sortedExpData[,6] <= 6 & sortedExpData[,6] > 5),]
binData$"6"$data <- sortedExpData[which(sortedExpData[,6] <= 7 & sortedExpData[,6] > 6),]
binData$"7"$data <- sortedExpData[which(sortedExpData[,6] <= 8 & sortedExpData[,6] > 7),]
binData$"8"$data <- sortedExpData[which(sortedExpData[,6] <= 9 & sortedExpData[,6] > 8),]
binData$"9"$data <- sortedExpData[which(sortedExpData[,6] <= 10 & sortedExpData[,6] > 9),]
## Calculate the ratio
binData$"-11"$ratio <- (length(which(binData[["-11"]][["data"]][["peak"]] == "RED")) / length(binData[["-11"]][["data"]][["X"]]))*100
binData$"-10"$ratio <- (length(which(binData[["-10"]][["data"]][["peak"]] == "RED")) / length(binData[["-10"]][["data"]][["X"]]))*100
binData$"-9"$ratio <- (length(which(binData[["-9"]][["data"]][["peak"]] == "RED")) / length(binData[["-9"]][["data"]][["X"]]))*100
binData$"-8"$ratio <- (length(which(binData[["-8"]][["data"]][["peak"]] == "RED")) / length(binData[["-8"]][["data"]][["X"]]))*100
binData$"-7"$ratio <- (length(which(binData[["-7"]][["data"]][["peak"]] == "RED")) / length(binData[["-7"]][["data"]][["X"]]))*100
binData$"-6"$ratio <- (length(which(binData[["-6"]][["data"]][["peak"]] == "RED")) / length(binData[["-6"]][["data"]][["X"]]))*100
binData$"-5"$ratio <- (length(which(binData[["-5"]][["data"]][["peak"]] == "RED")) / length(binData[["-5"]][["data"]][["X"]]))*100
binData$"-4"$ratio <- (length(which(binData[["-4"]][["data"]][["peak"]] == "RED")) / length(binData[["-4"]][["data"]][["X"]]))*100
binData$"-3"$ratio <- (length(which(binData[["-3"]][["data"]][["peak"]] == "RED")) / length(binData[["-3"]][["data"]][["X"]]))*100
binData$"-2"$ratio <- (length(which(binData[["-2"]][["data"]][["peak"]] == "RED")) / length(binData[["-2"]][["data"]][["X"]]))*100
binData$"-1"$ratio <- (length(which(binData[["-1"]][["data"]][["peak"]] == "RED")) / length(binData[["-1"]][["data"]][["X"]]))*100
binData$"0"$ratio <- (length(which(binData[["0"]][["data"]][["peak"]] == "RED")) / length(binData[["0"]][["data"]][["X"]]))*100
binData$"1"$ratio <- (length(which(binData[["1"]][["data"]][["peak"]] == "RED")) / length(binData[["1"]][["data"]][["X"]]))*100
binData$"2"$ratio <- (length(which(binData[["2"]][["data"]][["peak"]] == "RED")) / length(binData[["2"]][["data"]][["X"]]))*100
binData$"3"$ratio <- (length(which(binData[["3"]][["data"]][["peak"]] == "RED")) / length(binData[["3"]][["data"]][["X"]]))*100
binData$"4"$ratio <- (length(which(binData[["4"]][["data"]][["peak"]] == "RED")) / length(binData[["4"]][["data"]][["X"]]))*100
binData$"5"$ratio <- (length(which(binData[["5"]][["data"]][["peak"]] == "RED")) / length(binData[["5"]][["data"]][["X"]]))*100
binData$"6"$ratio <- (length(which(binData[["6"]][["data"]][["peak"]] == "RED")) / length(binData[["6"]][["data"]][["X"]]))*100
binData$"7"$ratio <- (length(which(binData[["7"]][["data"]][["peak"]] == "RED")) / length(binData[["7"]][["data"]][["X"]]))*100
binData$"8"$ratio <- (length(which(binData[["8"]][["data"]][["peak"]] == "RED")) / length(binData[["8"]][["data"]][["X"]]))*100
binData$"9"$ratio <- (length(which(binData[["9"]][["data"]][["peak"]] == "RED")) / length(binData[["9"]][["data"]][["X"]]))*100
##
ratios <- c(binData$"-11"$ratio, binData$"-10"$ratio, binData$"-9"$ratio, binData$"-8"$ratio, binData$"-7"$ratio, binData$"-6"$ratio,
            binData$"-5"$ratio, binData$"-4"$ratio, binData$"-3"$ratio, binData$"-2"$ratio, binData$"-1"$ratio, binData$"0"$ratio,
            binData$"1"$ratio, binData$"2"$ratio, binData$"3"$ratio, binData$"4"$ratio, binData$"5"$ratio, binData$"6"$ratio,
            binData$"7"$ratio, binData$"8"$ratio, binData$"9"$ratio)

##
plot(ratios)


#### Generate data to plot signal intensity (area under peak) against log2 gene expression
## Create the required annotation file to run Rsubread
uniquePromoters@elementMetadata$id <- uniquePromoters@elementMetadata@listData[["gene_id"]]
annotUP <- createAnnotationFile(uniquePromoters)
## Get raw counts for each range
uniquePromoterCounts <- featureCounts(files = inputBam, nthreads = 20, annot.ext = annotUP)


#### FIX THIS LATER ####
expData <- expData[-1114,]
expData <- expData[-1882,]
expData <- expData[-1909,]
expData <- expData[-2478,]
#### FIX THIS LATER ####


## Pull the required log2exp values
log2Matrix <- matrix(data = NA, nrow = length(uniquePromoters@elementMetadata@listData[["gene_id"]]), ncol = 2)
colnames(log2Matrix) <- c("ID", "log2exp")
##
for (x in 1:length(uniquePromoters@elementMetadata@listData[["gene_id"]])){
  
  genetemp <- uniquePromoters@elementMetadata@listData[["gene_id"]][[x]]
  ##
  log2Matrix[x,1] <- genetemp
  log2Matrix[x,2] <- expData[idxtemp,6]}


## Create a matrix to hold the log2 expression and ATACseq signal of the genes
## Rownames are geneID
plotMatrix <- matrix(data = NA, nrow = length(uniquePromoters@ranges@start), ncol = 2)
colnames(plotMatrix) <- c("log2exp", "accessibility")
##
plotMatrix[,1] <- log2Matrix[,2]
## Check that two matrices are in same order and aligned
for (l in 1:2435){
  if (log2Matrix[l,1] != uniquePromoterCounts[["annotation"]][["GeneID"]][[l]]){cat("MISMATCH")}}

## Add the ATACseq signals
plotMatrix[,2] <- uniquePromoterCounts[["counts"]]

#### Sort and plot the data
sortedplotMatrix <- plotMatrix[order(plotMatrix[,1], decreasing = FALSE),]

##
plot(sortedplotMatrix[,1], sortedplotMatrix[,2])
plot(sortedplotMatrix[,2], sortedplotMatrix[,1])

################ above is code i wrote for Jeremy ##########################################################################################################



# This script references: https://biodatascience.github.io/compbio/bioc/ranges.html

##
#source("https://bioconductor.org/biocLite.R")
#biocLite("stringi", suppressUpdates = TRUE)
#biocLite("GenomicRanges", suppressUpdates = TRUE)
#biocLite("ensembldb", suppressUpdates = TRUE)
#biocLite("EnsDb.Hsapiens.v86", suppressUpdates = TRUE)
#biocLite("S4Vectors", suppressUpdates = TRUE)
#biocLite("Rsubread", suppressUpdates = TRUE)
#biocLite("Repitools", suppressUpdates = TRUE)
#install.packages("ggplot2")
#devtools::install_github("r-lib/crayon")

##
library(ensembldb)
library(EnsDb.Hsapiens.v86)
library(GenomicRanges)
library(S4Vectors)
library(Rsubread)
library(Repitools)
library(ggplot2)
library(crayon)
library(Rsamtools)


# From unmodified genes list, normalize data and compare

# specify input files
input_files <- list()
input_files$snu61_bam1 <- "/home/ubuntu1/atac/snu61/wt01/dp_bam/SNU61-WT-01-S1_S6.lanemerge.dp.bam"
input_files$snu61_bam2 <- "/home/ubuntu1/atac/snu61/wt01/dp_bam/SNU61-WT-01-S2_S4.lanemerge.dp.bam"
input_files$snu61_bam3 <- "/home/ubuntu1/atac/snu61/wt01/dp_bam/SNU61-WT-01-S3_S1.lanemerge.dp.bam"
input_files$ls1034_bam1 <- "/home/ubuntu1/atac/ls1034/wt01/dp_bam/LS1034-WT-01-S2_S3.lanemerge.dp.bam"
input_files$ls1034_bam2 <- "/home/ubuntu1/atac/ls1034/wt01/dp_bam/LS1034-WT-01-S1_S5.lanemerge.dp.bam"
input_files$ls1034_bam3 <- "/home/ubuntu1/atac/ls1034/wt01/dp_bam/LS1034-WT-01-S3_S2.lanemerge.dp.bam"
input_files$h508_bam1 <- "/home/ubuntu1/atac/h508/wt01/dp_bam/H508-2_S2.lanemerge.dp.bam"
input_files$h508_bam2 <- "/home/ubuntu1/atac/h508/wt01/dp_bam/H508-3_S1.lanemerge.dp.bam"
input_files$h508_bam3 <- "/home/ubuntu1/atac/h508/wt01/dp_bam/H508-1_S3.lanemerge.dp.bam"

# find the total reads in all bam files
total_reads <- list()
total_reads$snu61_r1 <- countBam(input_files$snu61_bam1)
total_reads$snu61_r2 <- countBam(input_files$snu61_bam2)
total_reads$snu61_r3 <- countBam(input_files$snu61_bam3)
total_reads$ls1034_r1 <- countBam(input_files$ls1034_bam1)
total_reads$ls1034_r2 <- countBam(input_files$ls1034_bam2)
total_reads$ls1034_r3 <- countBam(input_files$ls1034_bam3)
total_reads$h508_r1 <- countBam(input_files$h508_bam1)
total_reads$h508_r2 <- countBam(input_files$h508_bam2)
total_reads$h508_r3 <- countBam(input_files$h508_bam3)

# calculate normalization factors
replicate_normalization <- matrix(data=NA, ncol=3, nrow=9)
colnames(replicate_normalization) <- c("total reads", "normalization factor", "normalized total reads")
rownames(replicate_normalization) <- c("snu61_r1","snu61_r2","snu61_r3","ls1034_r1","ls1034_r2","ls1034_r3","h508_r1","h508_r2","h508_r3")
#
replicate_normalization[1,1] <- total_reads[["snu61_r1"]][["records"]]
replicate_normalization[2,1] <- total_reads[["snu61_r2"]][["records"]]
replicate_normalization[3,1] <- total_reads[["snu61_r3"]][["records"]]
replicate_normalization[4,1] <- total_reads[["ls1034_r1"]][["records"]]
replicate_normalization[5,1] <- total_reads[["ls1034_r2"]][["records"]]
replicate_normalization[6,1] <- total_reads[["ls1034_r3"]][["records"]]
replicate_normalization[7,1] <- total_reads[["h508_r1"]][["records"]]
replicate_normalization[8,1] <- total_reads[["h508_r2"]][["records"]]
replicate_normalization[9,1] <- total_reads[["h508_r3"]][["records"]]
#
minimums <- list()
minimums$snu61_min <- min(replicate_normalization[1,1],replicate_normalization[2,1],replicate_normalization[3,1]) 
minimums$ls1034_min <- min(replicate_normalization[4,1],replicate_normalization[5,1],replicate_normalization[6,1])
minimums$h508_min <- min(replicate_normalization[7,1],replicate_normalization[8,1],replicate_normalization[9,1])
#
replicate_normalization[1,2] <- replicate_normalization[1,1] / minimums$snu61_min
replicate_normalization[2,2] <- replicate_normalization[2,1] / minimums$snu61_min
replicate_normalization[3,2] <- replicate_normalization[3,1] / minimums$snu61_min
replicate_normalization[4,2] <- replicate_normalization[4,1] / minimums$ls1034_min
replicate_normalization[5,2] <- replicate_normalization[5,1] / minimums$ls1034_min
replicate_normalization[6,2] <- replicate_normalization[6,1] / minimums$ls1034_min
replicate_normalization[7,2] <- replicate_normalization[7,1] / minimums$h508_min
replicate_normalization[8,2] <- replicate_normalization[8,1] / minimums$h508_min
replicate_normalization[9,2] <- replicate_normalization[9,1] / minimums$h508_min
#
replicate_normalization[1,3] <- replicate_normalization[1,1] / replicate_normalization[1,2]
replicate_normalization[2,3] <- replicate_normalization[2,1] / replicate_normalization[2,2]
replicate_normalization[3,3] <- replicate_normalization[3,1] / replicate_normalization[3,2]
replicate_normalization[4,3] <- replicate_normalization[4,1] / replicate_normalization[4,2]
replicate_normalization[5,3] <- replicate_normalization[5,1] / replicate_normalization[5,2]
replicate_normalization[6,3] <- replicate_normalization[6,1] / replicate_normalization[6,2]
replicate_normalization[7,3] <- replicate_normalization[7,1] / replicate_normalization[7,2]
replicate_normalization[8,3] <- replicate_normalization[8,1] / replicate_normalization[8,2]
replicate_normalization[9,3] <- replicate_normalization[9,1] / replicate_normalization[9,2]
#
edb <- EnsDb.Hsapiens.v86
norm_genes <- genes(edb)
rm(edb)
norm_genes <- keepStandardChromosomes(norm_genes, pruning.mode="coarse") # subset to only the standard chromosomes
norm_genes@elementMetadata$id <- norm_genes@elementMetadata@listData[["symbol"]]
annot_norm_gene <- createAnnotationFile(norm_genes)
# get the raw counts for each sample / replicate
gene_count_list <- list()
gene_count_list$snu61_r1 <- featureCounts(files = input_files$snu61_bam1, nthreads = 20, annot.ext = annot_norm_gene)
gene_count_list$snu61_r2 <- featureCounts(files = input_files$snu61_bam2, nthreads = 20, annot.ext = annot_norm_gene)
gene_count_list$snu61_r3 <- featureCounts(files = input_files$snu61_bam3, nthreads = 20, annot.ext = annot_norm_gene)
gene_count_list$ls1034_r1 <- featureCounts(files = input_files$ls1034_bam1, nthreads = 20, annot.ext = annot_norm_gene)
gene_count_list$ls1034_r2 <- featureCounts(files = input_files$ls1034_bam2, nthreads = 20, annot.ext = annot_norm_gene)
gene_count_list$ls1034_r3 <- featureCounts(files = input_files$ls1034_bam3, nthreads = 20, annot.ext = annot_norm_gene)
gene_count_list$h508_r1 <- featureCounts(files = input_files$h508_bam1, nthreads = 20, annot.ext = annot_norm_gene)
gene_count_list$h508_r2 <- featureCounts(files = input_files$h508_bam2, nthreads = 20, annot.ext = annot_norm_gene)
gene_count_list$h508_r3 <- featureCounts(files = input_files$h508_bam3, nthreads = 20, annot.ext = annot_norm_gene)
#
num_genes <- length(gene_count_list[["snu61_r1"]][["counts"]])
gene_names <- c(gene_count_list[["snu61_r1"]][["annotation"]][["GeneID"]])
#
normalized_counts <- list()
normalized_counts$snu61 <- matrix(data=NA, ncol=5, nrow=num_genes)
colnames(normalized_counts$snu61) <- c("rep1","rep2","rep3","avg","total")
rownames(normalized_counts$snu61) <- c(gene_names)
normalized_counts$ls1034 <- matrix(data=NA, ncol=5, nrow=num_genes)
colnames(normalized_counts$ls1034) <- c("rep1","rep2","rep3","avg","total")
rownames(normalized_counts$ls1034) <- c(gene_names)
normalized_counts$h508 <- matrix(data=NA, ncol=5, nrow=num_genes)
colnames(normalized_counts$h508) <- c("rep1","rep2","rep3","avg","total")
rownames(normalized_counts$h508) <- c(gene_names)
#
for (a in 1:num_genes){
  #
  normalized_counts$snu61[a,1] <- as.integer(gene_count_list[["snu61_r1"]][["counts"]][[a]]/replicate_normalization[1,2])
  normalized_counts$snu61[a,2] <- as.integer(gene_count_list[["snu61_r2"]][["counts"]][[a]]/replicate_normalization[2,2])
  normalized_counts$snu61[a,3] <- as.integer(gene_count_list[["snu61_r3"]][["counts"]][[a]]/replicate_normalization[3,2])
  normalized_counts$snu61[a,4] <- ((normalized_counts$snu61[a,1]+normalized_counts$snu61[a,2]+normalized_counts$snu61[a,3])/3)
  normalized_counts$snu61[a,5] <- normalized_counts$snu61[a,1]+normalized_counts$snu61[a,2]+normalized_counts$snu61[a,3]
  #
  normalized_counts$ls1034[a,1] <- as.integer(gene_count_list[["ls1034_r1"]][["counts"]][[a]]/replicate_normalization[4,2])
  normalized_counts$ls1034[a,2] <- as.integer(gene_count_list[["ls1034_r2"]][["counts"]][[a]]/replicate_normalization[5,2])
  normalized_counts$ls1034[a,3] <- as.integer(gene_count_list[["ls1034_r3"]][["counts"]][[a]]/replicate_normalization[6,2])
  normalized_counts$ls1034[a,4] <- ((normalized_counts$ls1034[a,1]+normalized_counts$ls1034[a,2]+normalized_counts$ls1034[a,3])/3)
  normalized_counts$ls1034[a,5] <- normalized_counts$ls1034[a,1]+normalized_counts$ls1034[a,2]+normalized_counts$ls1034[a,3]
  #
  normalized_counts$h508[a,1] <- as.integer(gene_count_list[["h508_r1"]][["counts"]][[a]]/replicate_normalization[7,2])
  normalized_counts$h508[a,2] <- as.integer(gene_count_list[["h508_r2"]][["counts"]][[a]]/replicate_normalization[8,2])
  normalized_counts$h508[a,3] <- as.integer(gene_count_list[["h508_r3"]][["counts"]][[a]]/replicate_normalization[9,2])
  normalized_counts$h508[a,4] <- ((normalized_counts$h508[a,1]+normalized_counts$h508[a,2]+normalized_counts$h508[a,3])/3)
  normalized_counts$h508[a,5] <- normalized_counts$h508[a,1]+normalized_counts$h508[a,2]+normalized_counts$h508[a,3]}
#
sample_normalization <- matrix(data=NA, ncol=3, nrow=3)
colnames(sample_normalization) <- c("total reads", "normalization factor", "normalized total reads")
rownames(sample_normalization) <- c("snu61", "ls1034", "h508")
#
sample_normalization[1,1] <- replicate_normalization[1,3]+replicate_normalization[2,3]+replicate_normalization[3,3]
sample_normalization[2,1] <- replicate_normalization[4,3]+replicate_normalization[5,3]+replicate_normalization[6,3]
sample_normalization[3,1] <- replicate_normalization[7,3]+replicate_normalization[8,3]+replicate_normalization[9,3]
#
minimums$sample_min <- min(sample_normalization[1,1],sample_normalization[2,1],sample_normalization[3,1])
#
sample_normalization[1,2] <- sample_normalization[1,1] / minimums$sample_min
sample_normalization[2,2] <- sample_normalization[2,1] / minimums$sample_min
sample_normalization[3,2] <- sample_normalization[3,1] / minimums$sample_min
#
sample_normalization[1,3] <- sample_normalization[1,1] /sample_normalization[1,2]
sample_normalization[2,3] <- sample_normalization[2,1] /sample_normalization[2,2]
sample_normalization[3,3] <- sample_normalization[3,1] /sample_normalization[3,2]
#
signal_matrix <- matrix(data=NA,nrow=num_genes,ncol=5)
rownames(signal_matrix) <- c(gene_names)
colnames(signal_matrix) <- c("snu61","ls1034","h508","var","sd")
#
for (b in 1:num_genes){
  #
  signal_matrix[b,1] <- as.integer((normalized_counts[["snu61"]][b,4]/sample_normalization[1,2]))
  signal_matrix[b,2] <- as.integer((normalized_counts[["ls1034"]][b,4]/sample_normalization[2,2]))
  signal_matrix[b,3] <- as.integer((normalized_counts[["h508"]][b,4]/sample_normalization[3,2]))
  signal_matrix[b,4] <- stats::var(signal_matrix[b,1:3])
  signal_matrix[b,5] <- stats::sd(signal_matrix[b,1:3])
}
#
coad_mr <- c("TCF7","MNX1","POU5F1B","ESRRA","CDX2","HNF4A","GMEB2","HOXA3","OVOL1","ASCL2","ZSWIM1","CBFA2T2","TARBP1","KLHL31","GTF2IRD1","ZNF696","NUFIP1","SUPT20H","KAT8","KAT2A","TAF4","RBM39","ZMYND8","L3MBTL1","PLAGL2","NCOA5","ZSWIM3","ADNP","ASXL1")
ind <- which(rownames(signal_matrix) %in% coad_mr)
mr_counts <- signal_matrix[ind,]
#



# Visualizations
# Decide on colors for cell lines here
# Violin plots

# Violin plots
# prep data
violin_data <- data.frame("Signal" = c(mr_counts[,1],mr_counts[,2],mr_counts[,3]),
                          "Sample" = c(rep("snu61", times=29), rep("ls1034",times=29),rep("h508",times=29)))
#
violin_plot <- ggplot(violin_data, aes(x=Sample, y=Signal, fill=Sample)) + 
  geom_violin()
violin_plot


# Change violin plot colors by groups
p<-ggplot(ToothGrowth, aes(x=dose, y=len, fill=dose)) +
  geom_violin(trim=FALSE)
p



# Violin for all genes


# Violin plots
# prep data
violin_data_all <- data.frame("Signal" = c(signal_matrix[,1],signal_matrix[,2],signal_matrix[,3]),
                              "Sample" = c(rep("snu61", times=num_genes), rep("ls1034",times=num_genes),rep("h508",times=num_genes)))
#
violin_plot <- ggplot(violin_data_all, aes(x=Sample, y=Signal, fill=Sample)) + 
  geom_violin()
violin_plot


# Heatmaps of the master regulators

mr_names <- rownames(mr_counts)
heatmap_data <- data.frame("Signal" = c(mr_counts[,1],mr_counts[,2],mr_counts[,3]),
                           "Sample" = c(rep("snu61", times=29), rep("ls1034",times=29),rep("h508",times=29)),
                           "Gene" = rep(mr_names))

heatmap.plot <- ggplot(data = heatmap_data, aes(x = Sample, y = Gene)) +
  geom_tile(aes(fill = Signal)) +
  scale_fill_gradient2(low="darkblue", high="red", guide="colorbar") +
  theme(axis.text.y = element_text(size = 6))

# Preview the heatmap
print(heatmap.plot)


# Heatmaps for individual replicates


#
mr_all_reps <- matrix(data=NA, nrow=29, ncol=9)
rownames(mr_all_reps) <- mr_names
colnames(mr_all_reps) <- c("snu61_r1","snu61_r2","snu61_r3","ls1034_r1","ls1034_r2","ls1034_r3","h508_r1","h508_r2","h508_r3")
#
mr_all_reps[,1] <- normalized_counts[["snu61"]][ind,1]
mr_all_reps[,2] <- normalized_counts[["snu61"]][ind,2]
mr_all_reps[,3] <- normalized_counts[["snu61"]][ind,3]
mr_all_reps[,4] <- normalized_counts[["ls1034"]][ind,1]
mr_all_reps[,5] <- normalized_counts[["ls1034"]][ind,2]
mr_all_reps[,6] <- normalized_counts[["ls1034"]][ind,3]
mr_all_reps[,7] <- normalized_counts[["h508"]][ind,1]
mr_all_reps[,8] <- normalized_counts[["h508"]][ind,2]
mr_all_reps[,9] <- normalized_counts[["h508"]][ind,3]
#
heatmap_data_reps <- data.frame(
  "Signal" = c(mr_all_reps[,1],mr_all_reps[,2],mr_all_reps[,3],mr_all_reps[,4],mr_all_reps[,5],mr_all_reps[,6],mr_all_reps[,7],mr_all_reps[,8],mr_all_reps[,9]),
  "Sample" = c(rep("snu61_r1", times=29), rep("snu61_r2", times=29), rep("snu61_r3", times=29), rep("ls1034_r1",times=29), rep("ls1034_r2",times=29), rep("ls1034_r3",times=29), rep("h508_r1",times=29), rep("h508_r2",times=29), rep("h508_r3",times=29)),
  "Gene" = rep(mr_names))
#
heatmap.plot.reps <- ggplot(data = heatmap_data_reps, aes(x = Sample, y = Gene)) +
  geom_tile(aes(fill = Signal)) +
  scale_fill_gradient2(low="darkblue", high="red", guide="colorbar") +
  theme(axis.text.y = element_text(size = 6))

# Preview the heatmap
print(heatmap.plot.reps)



#### Copy gene info from ensembl database and convert to Granges

edb <- EnsDb.Hsapiens.v86
genes <- genes(edb)
rm(edb)

# subset to only the standard chromosomes
genes <- keepStandardChromosomes(genes, pruning.mode="coarse")

## Create the modified subsets of annotated genes

#### initialize the modified granges objects ####
granges_list <- list()
genes_200up <- genes
genes_2000up <- genes
genes_200up_gene <- genes
genes_2000up_gene <- genes

#### make the modified granges objects ####
for (a in 1:length(genes)){
  
  ### change positive strand entries
  if (genes[a]@strand@values == "+"){
    
    ### 200 bp upstream only
    genes_200up[a]@ranges@start <- as.integer(genes_200up[a]@ranges@start - 200)
    genes_200up[a]@ranges@width <- as.integer(200)
    ### 2000 bp upstream only
    genes_2000up[a]@ranges@start <- as.integer(genes_2000up[a]@ranges@start - 2000)
    genes_2000up[a]@ranges@width <- as.integer(2000)
    ### 200 bp upstream and gene
    genes_200up_gene[a]@ranges@start <- as.integer(genes_200up_gene[a]@ranges@start - 200)
    genes_200up_gene[a]@ranges@width <- as.integer(genes_200up_gene[a]@ranges@width + 200)
    ### 2000 bp upstream and gene
    genes_2000up_gene[a]@ranges@start <- as.integer(genes_2000up_gene[a]@ranges@start - 2000)
    genes_2000up_gene[a]@ranges@width <- as.integer(genes_2000up_gene[a]@ranges@width + 2000)}
  
  ### change negative strand entries
  if (genes[a]@strand@values == "-"){
    
    ### 200 bp upstream only
    genes_200up[a]@ranges@start <- as.integer(genes_200up[a]@ranges@start + genes_200up[a]@ranges@width)
    genes_200up[a]@ranges@width <- as.integer(200)
    ### 2000 bp upstream only
    genes_2000up[a]@ranges@start <- as.integer(genes_2000up[a]@ranges@start + genes_2000up[a]@ranges@width)
    genes_2000up[a]@ranges@width <- as.integer(2000)
    ### 200 bp upstream and gene
    genes_200up_gene[a]@ranges@width <- as.integer(genes_200up_gene[a]@ranges@width + 200)
    ### 2000 bp upstream and gene
    genes_2000up_gene[a]@ranges@width <- as.integer(genes_2000up_gene[a]@ranges@width + 2000)
  }
}

granges_list$genes_200up <- genes_200up
granges_list$genes_2000up <- genes_2000up
granges_list$genes_200up_gene <- genes_200up_gene
granges_list$genes_2000up_gene <- genes_2000up_gene
rm(genes_200up, genes_2000up, genes_200up_gene, genes_2000up_gene)


#### Positive strand ####
# Get the first instance of a positive and negative strand gene from the original list
firstPlus <- which(genes@strand@values == "+")[1]
cat(black$bold(bgWhite("First plus strand gene found at index:",                (blue$bold(firstPlus)),"\n")))
firstName <- genes@elementMetadata@listData[["gene_name"]][firstPlus]
cat(black$bold(bgWhite("This genes name is:",(blue$bold(firstName)),"\n")))
firstStart <- genes@ranges@start[firstPlus]
cat(black$bold(bgWhite("It starts at position:",(blue$bold(firstStart)),"\n")))
firstWidth <- genes@ranges@width[firstPlus]
cat(black$bold(bgWhite("It has a width of:",(blue$bold(firstWidth)),"\n")))

# 200 bp upstream
name200 <- genes_200up@elementMetadata@listData[["gene_name"]][firstPlus]
start200 <- genes_200up@ranges@start[firstPlus]
width200 <- genes_200up@ranges@width[firstPlus]
cat(red$bold(bgWhite("Checking against:",(blue$bold("200 bp upstream")),"\n")))
cat(black$bold(bgWhite("This genes name is:",(blue$bold(name200)),"\n")))
cat(black$bold(bgWhite("It starts at position:",(blue$bold(start200)),"\n")))
cat(black$bold(bgWhite("It has a width of:",(blue$bold(width200)),"\n")))
if(firstName == name200){sanityChecks[1] = TRUE}else{sanityChecks[1] = FALSE}
if(firstStart == (start200+200)){sanityChecks[2] = TRUE}else{sanityChecks[2] = FALSE}
if(width200 == 200){sanityChecks[3] = TRUE}else{sanityChecks[3] = FALSE}

# 2000 bp upstream
name2000 <- genes_2000up@elementMetadata@listData[["gene_name"]][firstPlus]
start2000 <- genes_2000up@ranges@start[firstPlus]
width2000 <- genes_2000up@ranges@width[firstPlus]
cat(red$bold(bgWhite("Checking against:",(blue$bold("2000 bp upstream")),"\n")))
cat(black$bold(bgWhite("This genes name is:",(blue$bold(name2000)),"\n")))
cat(black$bold(bgWhite("It starts at position:",(blue$bold(start2000)),"\n")))
cat(black$bold(bgWhite("It has a width of:",(blue$bold(width2000)),"\n")))
if(firstName == name2000){sanityChecks[4] = TRUE}else{sanityChecks[4] = FALSE}
if(firstStart == (start2000+2000)){sanityChecks[5] = TRUE}else{sanityChecks[5] = FALSE}
if(width2000 == 2000){sanityChecks[6] = TRUE}else{sanityChecks[6] = FALSE}

# 200 bp upstream and gene
name200gene <- genes_200up_gene@elementMetadata@listData[["gene_name"]][firstPlus]
start200gene <- genes_200up_gene@ranges@start[firstPlus]
width200gene <- genes_200up_gene@ranges@width[firstPlus]
cat(red$bold(bgWhite("Checking against:",(blue$bold("200 bp upstream and gene")),"\n")))
cat(black$bold(bgWhite("This genes name is:",(blue$bold(name200gene)),"\n")))
cat(black$bold(bgWhite("It starts at position:",(blue$bold(start200gene)),"\n")))
cat(black$bold(bgWhite("It has a width of:",(blue$bold(width200gene)),"\n")))
if(firstName == name200gene){sanityChecks[7] = TRUE}else{sanityChecks[7] = FALSE}
if(firstStart == (start200gene+200)){sanityChecks[8] = TRUE}else{sanityChecks[8] = FALSE}
if(width200gene == (200+firstWidth)){sanityChecks[9] = TRUE}else{sanityChecks[9] = FALSE}

# 2000 bp upstream and gene
name2000gene <- genes_2000up_gene@elementMetadata@listData[["gene_name"]][firstPlus]
start2000gene <- genes_2000up_gene@ranges@start[firstPlus]
width2000gene <- genes_2000up_gene@ranges@width[firstPlus]
cat(red$bold(bgWhite("Checking against:",(blue$bold("2000 bp upstream and gene")),"\n")))
cat(black$bold(bgWhite("This genes name is:",(blue$bold(name2000gene)),"\n")))
cat(black$bold(bgWhite("It starts at position:",(blue$bold(start2000gene)),"\n")))
cat(black$bold(bgWhite("It has a width of:",(blue$bold(width2000gene)),"\n")))
if(firstName == name2000gene){sanityChecks[10] = TRUE}else{sanityChecks[10] = FALSE}
if(firstStart == (start2000gene+2000)){sanityChecks[11] = TRUE}else{sanityChecks[11] = FALSE}
if(width2000gene == (2000+firstWidth)){sanityChecks[12] = TRUE}else{sanityChecks[12] = FALSE}

#### Negative strand ####
firstNeg <- which(genes@strand@values == "-")[1]
cat(black$bold(bgWhite("First minus strand gene found at index:",                (blue$bold(firstNeg)),"\n")))
firstName <- genes@elementMetadata@listData[["gene_name"]][firstNeg]
cat(black$bold(bgWhite("This genes name is:",(blue$bold(firstName)),"\n")))
firstStart <- genes@ranges@start[firstNeg]
cat(black$bold(bgWhite("It starts at position:",(blue$bold(firstStart)),"\n")))
firstWidth <- genes@ranges@width[firstNeg]
cat(black$bold(bgWhite("It has a width of:",(blue$bold(firstWidth)),"\n")))

# 200 bp upstream
name200 <- genes_200up@elementMetadata@listData[["gene_name"]][firstNeg]
start200 <- genes_200up@ranges@start[firstNeg]
width200 <- genes_200up@ranges@width[firstNeg]
cat(red$bold(bgWhite("Checking against:",(blue$bold("200 bp upstream")),"\n")))
cat(black$bold(bgWhite("This genes name is:",(blue$bold(name200)),"\n")))
cat(black$bold(bgWhite("It starts at position:",(blue$bold(start200)),"\n")))
cat(black$bold(bgWhite("It has a width of:",(blue$bold(width200)),"\n")))
if(firstName == name200){sanityChecks[13] = TRUE}else{sanityChecks[13] = FALSE}
if(firstStart == (start200-firstWidth)){sanityChecks[14] = TRUE}else{sanityChecks[14] = FALSE}
if(width200 == 200){sanityChecks[15] = TRUE}else{sanityChecks[15] = FALSE}

# 2000 bp upstream
name2000 <- genes_2000up@elementMetadata@listData[["gene_name"]][firstNeg]
start2000 <- genes_2000up@ranges@start[firstNeg]
width2000 <- genes_2000up@ranges@width[firstNeg]
cat(red$bold(bgWhite("Checking against:",(blue$bold("2000 bp upstream")),"\n")))
cat(black$bold(bgWhite("This genes name is:",(blue$bold(name2000)),"\n")))
cat(black$bold(bgWhite("It starts at position:",(blue$bold(start2000)),"\n")))
cat(black$bold(bgWhite("It has a width of:",(blue$bold(width2000)),"\n")))
if(firstName == name2000){sanityChecks[16] = TRUE}else{sanityChecks[16] = FALSE}
if(firstStart == (start2000-firstWidth)){sanityChecks[17] = TRUE}else{sanityChecks[17] = FALSE}
if(width2000 == 2000){sanityChecks[18] = TRUE}else{sanityChecks[18] = FALSE}

# 200 bp upstream and gene
name200gene <- genes_200up_gene@elementMetadata@listData[["gene_name"]][firstNeg]
start200gene <- genes_200up_gene@ranges@start[firstNeg]
width200gene <- genes_200up_gene@ranges@width[firstNeg]
cat(red$bold(bgWhite("Checking against:",(blue$bold("200 bp upstream and gene")),"\n")))
cat(black$bold(bgWhite("This genes name is:",(blue$bold(name200gene)),"\n")))
cat(black$bold(bgWhite("It starts at position:",(blue$bold(start200gene)),"\n")))
cat(black$bold(bgWhite("It has a width of:",(blue$bold(width200gene)),"\n")))
if(firstName == name200gene){sanityChecks[19] = TRUE}else{sanityChecks[19] = FALSE}
if(firstStart == start200gene){sanityChecks[20] = TRUE}else{sanityChecks[20] = FALSE}
if(width200gene == (200+firstWidth)){sanityChecks[21] = TRUE}else{sanityChecks[21] = FALSE}

# 2000 bp upstream and gene
name2000gene <- genes_2000up_gene@elementMetadata@listData[["gene_name"]][firstNeg]
start2000gene <- genes_2000up_gene@ranges@start[firstNeg]
width2000gene <- genes_2000up_gene@ranges@width[firstNeg]
cat(red$bold(bgWhite("Checking against:",(blue$bold("2000 bp upstream and gene")),"\n")))
cat(black$bold(bgWhite("This genes name is:",(blue$bold(name2000gene)),"\n")))
cat(black$bold(bgWhite("It starts at position:",(blue$bold(start2000gene)),"\n")))
cat(black$bold(bgWhite("It has a width of:",(blue$bold(width2000gene)),"\n")))
if(firstName == name2000gene){sanityChecks[22] = TRUE}else{sanityChecks[22] = FALSE}
if(firstStart == start2000gene){sanityChecks[23] = TRUE}else{sanityChecks[23] = FALSE}
if(width2000gene == (2000+firstWidth)){sanityChecks[24] = TRUE}else{sanityChecks[24] = FALSE}


# Trim out of bounds ranges


idx <- GenomicRanges:::get_out_of_bound_index(genes)
if (length(idx) != 0L)
  genes <- genes[-idx]

idx <- GenomicRanges:::get_out_of_bound_index(genes_200up)
if (length(idx) != 0L)
  genes_200up <- genes_200up[-idx]

idx <- GenomicRanges:::get_out_of_bound_index(genes_2000up)
if (length(idx) != 0L)
  genes_2000up <- genes_2000up[-idx]

idx <- GenomicRanges:::get_out_of_bound_index(genes_200up_gene)
if (length(idx) != 0L)
  genes_200up_gene <- genes_200up_gene[-idx]

idx <- GenomicRanges:::get_out_of_bound_index(genes_2000up_gene)
if (length(idx) != 0L)
  genes_2000up_gene <- genes_2000up_gene[-idx]



# Granges must contain gene id in metadata column to create annotation files

genes@elementMetadata$id <- genes@elementMetadata@listData[["symbol"]]
genes_200up@elementMetadata$id <- genes_200up@elementMetadata@listData[["symbol"]]
genes_2000up@elementMetadata$id <- genes_2000up@elementMetadata@listData[["symbol"]]
genes_200up_gene@elementMetadata$id <- genes_200up_gene@elementMetadata@listData[["symbol"]]
genes_2000up_gene@elementMetadata$id <- genes_2000up_gene@elementMetadata@listData[["symbol"]]


## Create annotation file from Granges object for counting reads

annot_gene <- createAnnotationFile(genes)
annot_200up <- createAnnotationFile(genes_200up)
annot_2000up <- createAnnotationFile(genes_2000up)
annot_gene_200up <- createAnnotationFile(genes_200up_gene)
annot_gene_2000up <- createAnnotationFile(genes_2000up_gene)


# Make sure all annotation files are the same size

# use annot 2000 up as the base index

idx <- c(annot_2000up[["GeneID"]])

idx2 <- which(annot_200up[["GeneID"]] %in% idx)

annot_200up <- annot_200up[idx2,]
annot_gene <- annot_gene[idx2,]
annot_gene_200up <- annot_gene_200up[idx2,]



# run featureCounts to count up Tn5 insertions in each interval, for all replicates

snu_bam1 <- "/home/ubuntu1/atac/snu61/wt01/dp_bam/SNU61-WT-01-S1_S6.lanemerge.dp.bam"
snu_bam2 <- "/home/ubuntu1/atac/snu61/wt01/dp_bam/SNU61-WT-01-S2_S4.lanemerge.dp.bam"
snu_bam3 <- "/home/ubuntu1/atac/snu61/wt01/dp_bam/SNU61-WT-01-S3_S1.lanemerge.dp.bam"

## Gene ##
snu_r1_counts_gene <- featureCounts(
  files=snu_bam1,
  nthreads=20,
  annot.ext=annot_gene)

snu_r2_counts_gene <- featureCounts(
  files=snu_bam2,
  nthreads=20,
  annot.ext=annot_gene)

snu_r3_counts_gene <- featureCounts(
  files=snu_bam3,
  nthreads=20,
  annot.ext=annot_gene)

## 200 bp upstream
snu_r1_counts_200up <- featureCounts(
  files=snu_bam1,
  isPairedEnd=TRUE,
  nthreads=20,
  annot.ext=annot_200up)

snu_r2_counts_200up <- featureCounts(
  files=snu_bam2,
  nthreads=20,
  annot.ext=annot_200up)

snu_r3_counts_200up <- featureCounts(
  files=snu_bam3,
  nthreads=20,
  annot.ext=annot_200up)

## 2000 bp upstream
snu_r1_counts_2000up <- featureCounts(
  files=snu_bam1,
  isPairedEnd=TRUE,
  nthreads=20,
  annot.ext=annot_2000up)

snu_r2_counts_2000up <- featureCounts(
  files=snu_bam2,
  nthreads=20,
  annot.ext=annot_2000up)

snu_r3_counts_2000up <- featureCounts(
  files=snu_bam3,
  nthreads=20,
  annot.ext=annot_2000up)

## 200 up plus gene
snu_r1_counts_gene_200up <- featureCounts(
  files=snu_bam1,
  nthreads=20,
  annot.ext=annot_gene_200up)

snu_r2_counts_gene_200up <- featureCounts(
  files=snu_bam2,
  nthreads=20,
  annot.ext=annot_gene_200up)

snu_r3_counts_gene_200up <- featureCounts(
  files=snu_bam3,
  nthreads=20,
  annot.ext=annot_gene_200up)

## 2000 up and gene
snu_r1_counts_gene_2000up <- featureCounts(
  files=snu_bam1,
  nthreads=20,
  annot.ext=annot_gene_2000up)

snu_r2_counts_gene_2000up <- featureCounts(
  files=snu_bam2,
  nthreads=20,
  annot.ext=annot_gene_2000up)

snu_r3_counts_gene_2000up <- featureCounts(
  files=snu_bam3,
  nthreads=20,
  annot.ext=annot_gene_2000up)


##### LS1034 WT 01
ls1034_bam1 <- "/home/ubuntu1/atac/ls1034/wt01/dp_bam/LS1034-WT-01-S2_S3.lanemerge.dp.bam"
ls1034_bam2 <- "/home/ubuntu1/atac/ls1034/wt01/dp_bam/LS1034-WT-01-S1_S5.lanemerge.dp.bam"
ls1034_bam3 <- "/home/ubuntu1/atac/ls1034/wt01/dp_bam/LS1034-WT-01-S3_S2.lanemerge.dp.bam"

## Gene ##
ls1034_r1_counts_gene <- featureCounts(
  files=ls1034_bam1,
  nthreads=20,
  annot.ext=annot_gene)

ls1034_r2_counts_gene <- featureCounts(
  files=ls1034_bam2,
  nthreads=20,
  annot.ext=annot_gene)

ls1034_r3_counts_gene <- featureCounts(
  files=ls1034_bam3,
  nthreads=20,
  annot.ext=annot_gene)

## 200 bp upstream
ls1034_r1_counts_200up <- featureCounts(
  files=ls1034_bam1,
  nthreads=20,
  annot.ext=annot_200up)

ls1034_r2_counts_200up <- featureCounts(
  files=ls1034_bam2,
  nthreads=20,
  annot.ext=annot_200up)

ls1034_r3_counts_200up <- featureCounts(
  files=ls1034_bam3,
  nthreads=20,
  annot.ext=annot_200up)

## 2000 bp upstream
ls1034_r1_counts_2000up <- featureCounts(
  files=ls1034_bam1,
  nthreads=20,
  annot.ext=annot_2000up)

ls1034_r2_counts_2000up <- featureCounts(
  files=ls1034_bam2,
  nthreads=20,
  annot.ext=annot_2000up)

ls1034_r3_counts_2000up <- featureCounts(
  files=ls1034_bam3,
  nthreads=20,
  annot.ext=annot_2000up)

## 200 up plus gene
ls1034_r1_counts_gene_200up <- featureCounts(
  files=ls1034_bam1,
  nthreads=20,
  annot.ext=annot_gene_200up)

ls1034_r2_counts_gene_200up <- featureCounts(
  files=ls1034_bam2,
  nthreads=20,
  annot.ext=annot_gene_200up)

ls1034_r3_counts_gene_200up <- featureCounts(
  files=ls1034_bam3,
  nthreads=20,
  annot.ext=annot_gene_200up)

## 2000 up and gene
ls1034_r1_counts_gene_2000up <- featureCounts(
  files=ls1034_bam1,
  nthreads=20,
  annot.ext=annot_gene_2000up)

ls1034_r2_counts_gene_2000up <- featureCounts(
  files=ls1034_bam2,
  nthreads=20,
  annot.ext=annot_gene_2000up)

ls1034_r3_counts_gene_2000up <- featureCounts(
  files=ls1034_bam3,
  nthreads=20,
  annot.ext=annot_gene_2000up)

#### H508 WT 01
h508_bam1 <- "/home/ubuntu1/atac/h508/wt01/dp_bam/H508-2_S2.lanemerge.dp.bam"
h508_bam2 <- "/home/ubuntu1/atac/h508/wt01/dp_bam/H508-3_S1.lanemerge.dp.bam"
h508_bam3 <- "/home/ubuntu1/atac/h508/wt01/dp_bam/H508-1_S3.lanemerge.dp.bam"

## Gene ##
h508_r1_counts_gene <- featureCounts(
  files=h508_bam1,
  nthreads=20,
  annot.ext=annot_gene)

h508_r2_counts_gene <- featureCounts(
  files=h508_bam2,
  nthreads=20,
  annot.ext=annot_gene)

h508_r3_counts_gene <- featureCounts(
  files=h508_bam3,
  nthreads=20,
  annot.ext=annot_gene)

## 200 bp upstream
h508_r1_counts_200up <- featureCounts(
  files=h508_bam1,
  nthreads=20,
  annot.ext=annot_200up)

h508_r2_counts_200up <- featureCounts(
  files=h508_bam2,
  nthreads=20,
  annot.ext=annot_200up)

h508_r3_counts_200up <- featureCounts(
  files=h508_bam3,
  nthreads=20,
  annot.ext=annot_200up)

## 2000 bp upstream
h508_r1_counts_2000up <- featureCounts(
  files=h508_bam1,
  nthreads=20,
  annot.ext=annot_2000up)

h508_r2_counts_2000up <- featureCounts(
  files=h508_bam2,
  nthreads=20,
  annot.ext=annot_2000up)

h508_r3_counts_2000up <- featureCounts(
  files=h508_bam3,
  nthreads=20,
  annot.ext=annot_2000up)

## 200 up plus gene
h508_r1_counts_gene_200up <- featureCounts(
  files=h508_bam1,
  nthreads=20,
  annot.ext=annot_gene_200up)

h508_r2_counts_gene_200up <- featureCounts(
  files=h508_bam2,
  nthreads=20,
  annot.ext=annot_gene_200up)

h508_r3_counts_gene_200up <- featureCounts(
  files=h508_bam3,
  nthreads=20,
  annot.ext=annot_gene_200up)

## 2000 up and gene
h508_r1_counts_gene_2000up <- featureCounts(
  files=h508_bam1,
  nthreads=20,
  annot.ext=annot_gene_2000up)

h508_r2_counts_gene_2000up <- featureCounts(
  files=h508_bam2,
  nthreads=20,
  annot.ext=annot_gene_2000up)

h508_r3_counts_gene_2000up <- featureCounts(
  files=h508_bam3,
  nthreads=20,
  annot.ext=annot_gene_2000up)


## Copy the data to a matrix


one <- paste0("snu61",c("gene_r1","gene_r2","gene_r3","200up_r1","200up_r2","200up_r3","2000up_r1","2000up_r2","2000up_r3","gene+200up_r1","gene+200up_r2","gene+200up_r3","gene+2000up_r1","gene+2000up_r2","gene+2000up_r3"))
two <- paste0("ls1034",c("gene_r1","gene_r2","gene_r3","200up_r1","200up_r2","200up_r3","2000up_r1","2000up_r2","2000up_r3","gene+200up_r1","gene+200up_r2","gene+200up_r3","gene+2000up_r1","gene+2000up_r2","gene+2000up_r3"))
three <- paste0("h508",c("gene_r1","gene_r2","gene_r3","200up_r1","200up_r2","200up_r3","2000up_r1","2000up_r2","2000up_r3","gene+200up_r1","gene+200up_r2","gene+200up_r3","gene+2000up_r1","gene+2000up_r2","gene+2000up_r3"))

four <- paste0(c("snu61_", "ls1034_", "h508_"), "gene_avg")
five <- paste0(c("snu61_", "ls1034_", "h508_"), "200bp_avg")
six <- paste0(c("snu61_", "ls1034_", "h508_"), "2000bp_avg")
seven <- paste0(c("snu61_", "ls1034_", "h508_"), "gene+200_avg")
eight <- paste0(c("snu61_", "ls1034_", "h508_"), "gene+2000_avg")

nine <- c("gene_diff_h508_ls1034","gene_diff_h508_snu61","gene_diff_snu61_ls1034")
ten <- c("200_diff_h508_ls1034","200_diff_h508_snu61","200_diff_snu61_ls1034")
eleven <- c("2000_diff_h508_ls1034","2000_diff_h508_snu61","2000_diff_snu61_ls1034")
twelve <- c("gene+200_diff_h508_ls1034","gene+200_diff_h508_snu61","gene+200_diff_snu61_ls1034")
thirteen <- c("gene+2000_diff_h508_ls1034","gene+2000_diff_h508_snu61","gene+2000_diff_snu61_ls1034")

colnames <- c(one,two,three,four,five,six,seven,eight,nine,ten,eleven,twelve,thirteen)
rm(one,two,three,four,five,six,seven,eight,nine,ten,eleven,twelve,thirteen)

# make the matrix
items <- length(annot_gene[,1])
#
signal_matrix <- matrix(data = NA, items, 75)
#
colnames(signal_matrix) <- colnames
# apply rownames to matrix
rownames(signal_matrix) <- c(annot_gene[["GeneID"]])


# Copy signals to matrix

for (c in 1:items){
  
  ## SNU61 raw data
  signal_matrix[[c,1]] <- as.integer(snu_r1_counts_gene[["counts"]][[c]])
  signal_matrix[[c,2]] <- as.integer(snu_r2_counts_gene[["counts"]][[c]])
  signal_matrix[[c,3]] <- as.integer(snu_r3_counts_gene[["counts"]][[c]])
  signal_matrix[[c,4]] <- as.integer(snu_r1_counts_200up[["counts"]][[c]])
  signal_matrix[[c,5]] <- as.integer(snu_r2_counts_200up[["counts"]][[c]])
  signal_matrix[[c,6]] <- as.integer(snu_r3_counts_200up[["counts"]][[c]])
  signal_matrix[[c,7]] <- as.integer(snu_r1_counts_2000up[["counts"]][[c]])
  signal_matrix[[c,8]] <- as.integer(snu_r2_counts_2000up[["counts"]][[c]])
  signal_matrix[[c,9]] <- as.integer(snu_r3_counts_2000up[["counts"]][[c]])
  signal_matrix[[c,10]] <- as.integer(snu_r1_counts_gene_200up[["counts"]][[c]])
  signal_matrix[[c,11]] <- as.integer(snu_r2_counts_gene_200up[["counts"]][[c]])
  signal_matrix[[c,12]] <- as.integer(snu_r3_counts_gene_200up[["counts"]][[c]])
  signal_matrix[[c,13]] <- as.integer(snu_r1_counts_gene_2000up[["counts"]][[c]])
  signal_matrix[[c,14]] <- as.integer(snu_r2_counts_gene_2000up[["counts"]][[c]])
  signal_matrix[[c,15]] <- as.integer(snu_r3_counts_gene_2000up[["counts"]][[c]])
  
  ## LS1034 raw data
  signal_matrix[[c,16]] <- as.integer(ls1034_r1_counts_gene[["counts"]][[c]])
  signal_matrix[[c,17]] <- as.integer(ls1034_r2_counts_gene[["counts"]][[c]])
  signal_matrix[[c,18]] <- as.integer(ls1034_r3_counts_gene[["counts"]][[c]])
  signal_matrix[[c,19]] <- as.integer(ls1034_r1_counts_200up[["counts"]][[c]])
  signal_matrix[[c,20]] <- as.integer(ls1034_r2_counts_200up[["counts"]][[c]])
  signal_matrix[[c,21]] <- as.integer(ls1034_r3_counts_200up[["counts"]][[c]])
  signal_matrix[[c,22]] <- as.integer(ls1034_r1_counts_2000up[["counts"]][[c]])
  signal_matrix[[c,23]] <- as.integer(ls1034_r2_counts_2000up[["counts"]][[c]])
  signal_matrix[[c,24]] <- as.integer(ls1034_r3_counts_2000up[["counts"]][[c]])
  signal_matrix[[c,25]] <- as.integer(ls1034_r1_counts_gene_200up[["counts"]][[c]])
  signal_matrix[[c,26]] <- as.integer(ls1034_r2_counts_gene_200up[["counts"]][[c]])
  signal_matrix[[c,27]] <- as.integer(ls1034_r3_counts_gene_200up[["counts"]][[c]])
  signal_matrix[[c,28]] <- as.integer(ls1034_r1_counts_gene_2000up[["counts"]][[c]])
  signal_matrix[[c,29]] <- as.integer(ls1034_r2_counts_gene_2000up[["counts"]][[c]])
  signal_matrix[[c,30]] <- as.integer(ls1034_r3_counts_gene_2000up[["counts"]][[c]])
  
  ## H508 raw data
  signal_matrix[[c,31]] <- as.integer(h508_r1_counts_gene[["counts"]][[c]])
  signal_matrix[[c,32]] <- as.integer(h508_r2_counts_gene[["counts"]][[c]])
  signal_matrix[[c,33]] <- as.integer(h508_r3_counts_gene[["counts"]][[c]])
  signal_matrix[[c,34]] <- as.integer(h508_r1_counts_200up[["counts"]][[c]])
  signal_matrix[[c,35]] <- as.integer(h508_r2_counts_200up[["counts"]][[c]])
  signal_matrix[[c,36]] <- as.integer(h508_r3_counts_200up[["counts"]][[c]])
  signal_matrix[[c,37]] <- as.integer(h508_r1_counts_2000up[["counts"]][[c]])
  signal_matrix[[c,38]] <- as.integer(h508_r2_counts_2000up[["counts"]][[c]])
  signal_matrix[[c,39]] <- as.integer(h508_r3_counts_2000up[["counts"]][[c]])
  signal_matrix[[c,40]] <- as.integer(h508_r1_counts_gene_200up[["counts"]][[c]])
  signal_matrix[[c,41]] <- as.integer(h508_r2_counts_gene_200up[["counts"]][[c]])
  signal_matrix[[c,42]] <- as.integer(h508_r3_counts_gene_200up[["counts"]][[c]])
  signal_matrix[[c,43]] <- as.integer(h508_r1_counts_gene_2000up[["counts"]][[c]])
  signal_matrix[[c,44]] <- as.integer(h508_r2_counts_gene_2000up[["counts"]][[c]])
  signal_matrix[[c,45]] <- as.integer(h508_r3_counts_gene_2000up[["counts"]][[c]])
  
  ## Gene signal averages (snu61, ls1034, h508)
  signal_matrix[[c,46]] <- as.integer(mean(signal_matrix[c,1],signal_matrix[c,2],signal_matrix[c,3]))
  signal_matrix[[c,47]] <- as.integer(mean(signal_matrix[c,16],signal_matrix[c,17],signal_matrix[c,18]))
  signal_matrix[[c,48]] <- as.integer(mean(signal_matrix[c,31],signal_matrix[c,32],signal_matrix[c,33]))
  
  ## 200 bp up signal averages (snu61, ls1034, h508)
  signal_matrix[[c,49]] <- as.integer(mean(signal_matrix[c,4],signal_matrix[c,5],signal_matrix[c,6]))
  signal_matrix[[c,50]] <- as.integer(mean(signal_matrix[c,19],signal_matrix[c,20],signal_matrix[c,21]))
  signal_matrix[[c,51]] <- as.integer(mean(signal_matrix[c,34],signal_matrix[c,35],signal_matrix[c,36]))
  
  ## 2000 bp up signal averages (snu61, ls1034, h508)
  signal_matrix[[c,52]] <- as.integer(mean(signal_matrix[c,7],signal_matrix[c,8],signal_matrix[c,9]))
  signal_matrix[[c,53]] <- as.integer(mean(signal_matrix[c,22],signal_matrix[c,23],signal_matrix[c,24]))
  signal_matrix[[c,54]] <- as.integer(mean(signal_matrix[c,37],signal_matrix[c,38],signal_matrix[c,39]))
  
  ## 200 bp up and gene signal averages (snu61, ls1034, h508)
  signal_matrix[[c,55]] <- as.integer(mean(signal_matrix[c,10],signal_matrix[c,11],signal_matrix[c,12]))
  signal_matrix[[c,56]] <- as.integer(mean(signal_matrix[c,25],signal_matrix[c,26],signal_matrix[c,27]))
  signal_matrix[[c,57]] <- as.integer(mean(signal_matrix[c,40],signal_matrix[c,41],signal_matrix[c,42]))
  
  ## 2000 bp and up gene signal averages (snu61, ls1034, h508)
  signal_matrix[[c,58]] <- as.integer(mean(signal_matrix[c,13],signal_matrix[c,14],signal_matrix[c,15]))
  signal_matrix[[c,59]] <- as.integer(mean(signal_matrix[c,28],signal_matrix[c,29],signal_matrix[c,30]))
  signal_matrix[[c,60]] <- as.integer(mean(signal_matrix[c,43],signal_matrix[c,44],signal_matrix[c,45]))
  
  #### Pairwise differences
  
  ## gene (h508/ls1034, h508/snu61, snu61/ls1034)
  signal_matrix[[c,61]] <- abs(signal_matrix[[c,48]]-signal_matrix[[c,47]])
  signal_matrix[[c,62]] <- abs(signal_matrix[[c,48]]-signal_matrix[[c,46]])
  signal_matrix[[c,63]] <- abs(signal_matrix[[c,47]]-signal_matrix[[c,46]])
  
  ## 200 bp upstream (h508/ls1034, h508/snu61, snu61/ls1034)
  signal_matrix[[c,64]] <- abs(signal_matrix[[c,51]]-signal_matrix[[c,50]])
  signal_matrix[[c,65]] <- abs(signal_matrix[[c,51]]-signal_matrix[[c,49]])
  signal_matrix[[c,66]] <- abs(signal_matrix[[c,50]]-signal_matrix[[c,49]])
  
  ## 2000 bp upstream (h508/ls1034, h508/snu61, snu61/ls1034)
  signal_matrix[[c,67]] <- abs(signal_matrix[[c,54]]-signal_matrix[[c,53]])
  signal_matrix[[c,68]] <- abs(signal_matrix[[c,54]]-signal_matrix[[c,52]])
  signal_matrix[[c,69]] <- abs(signal_matrix[[c,53]]-signal_matrix[[c,52]])
  
  ## 200 bp upstream and gene (h508/ls1034, h508/snu61, snu61/ls1034)
  signal_matrix[[c,70]] <- abs(signal_matrix[[c,57]]-signal_matrix[[c,56]])
  signal_matrix[[c,71]] <- abs(signal_matrix[[c,57]]-signal_matrix[[c,55]])
  signal_matrix[[c,72]] <- abs(signal_matrix[[c,56]]-signal_matrix[[c,55]])
  
  ## 2000 bp upstream and gene (h508/ls1034, h508/snu61, snu61/ls1034)
  signal_matrix[[c,73]] <- abs(signal_matrix[[c,60]]-signal_matrix[[c,59]])
  signal_matrix[[c,74]] <- abs(signal_matrix[[c,60]]-signal_matrix[[c,58]])
  signal_matrix[[c,75]] <- abs(signal_matrix[[c,59]]-signal_matrix[[c,58]])
}


# Convert values of interest to dataframe to make descriptive plots

# snu61, ls1034, h508
df <- data.frame(
  "snu61_gene"=signal_matrix[,46],
  "ls1034_gene"=signal_matrix[,47],
  "h508_gene"=signal_matrix[,48],
  #
  "snu61_200"=signal_matrix[,49],
  "ls1034_200"=signal_matrix[,50],
  "h508_200"=signal_matrix[,51],
  #
  "snu61_2000"=signal_matrix[,52],
  "ls1034_2000"=signal_matrix[,53],
  "h508_2000"=signal_matrix[,54],
  #
  "snu61_200_gene"=signal_matrix[,55],
  "ls1034_200_gene"=signal_matrix[,56],
  "h508_200_gene"=signal_matrix[,57],
  #
  "snu61_2000_gene"=signal_matrix[,58],
  "ls1034_2000_gene"=signal_matrix[,59],
  "h508_2000_gene"=signal_matrix[,60]
)



# Make descriptive statistics and plots from data

options(scipen=999)  # turn-off scientific notation like 1e+4


## calculate quantiles ###
quantiles <- list()
# SNU61 quantiles
quantiles$snu61_gene <- quantile(df[["snu61_gene"]], na.rm=TRUE)
quantiles$snu61_200 <- quantile(df[["snu61_200"]], na.rm=TRUE)
quantiles$snu61_2000 <- quantile(df[["snu61_2000"]], na.rm=TRUE)
quantiles$snu61_200_gene <- quantile(df[["snu61_200_gene"]], na.rm=TRUE)
quantiles$snu61_2000_gene <- quantile(df[["snu61_2000_gene"]], na.rm=TRUE)
# LS1034 quantiles
quantiles$ls1034_gene <- quantile(df[["ls1034_gene"]], na.rm=TRUE)
quantiles$ls1034_200 <- quantile(df[["ls1034_200"]], na.rm=TRUE)
quantiles$ls1034_2000 <- quantile(df[["ls1034_2000"]], na.rm=TRUE)
quantiles$ls1034_200_gene <- quantile(df[["ls1034_200_gene"]], na.rm=TRUE)
quantiles$ls1034_2000_gene <- quantile(df[["ls1034_2000_gene"]], na.rm=TRUE)
# H508 quantiles
quantiles$h508_gene <- quantile(df[["h508_gene"]], na.rm=TRUE)
quantiles$h508_200 <- quantile(df[["h508_200"]], na.rm=TRUE)
quantiles$h508_2000 <- quantile(df[["h508_2000"]], na.rm=TRUE)
quantiles$h508_200_gene <- quantile(df[["h508_200_gene"]], na.rm=TRUE)
quantiles$h508_2000_gene <- quantile(df[["h508_2000_gene"]], na.rm=TRUE)


### make basic scatter plots to show range of signal values ###
# SNU61
plot.default(x=df[["snu61_gene"]], y=NULL, ylab="READS GENE", xlab="SNU-61")
plot.default(x=df[["snu61_200"]], y=NULL, ylab="READS 200", xlab="SNU-61")
plot.default(x=df[["snu61_2000"]], y=NULL, ylab="READS 2000", xlab="SNU-61")
plot.default(x=df[["snu61_200_gene"]], y=NULL, ylab="READS 200 GENE", xlab="SNU-61")
plot.default(x=df[["snu61_2000_gene"]], y=NULL, ylab="READS 2000 GENE", xlab="SNU-61")
# LS1034
plot.default(x=df[["ls1034_gene"]], y=NULL, ylab="READS GENE", xlab="LS1034")
plot.default(x=df[["ls1034_200"]], y=NULL, ylab="READS 200", xlab="LS1034")
plot.default(x=df[["ls1034_2000"]], y=NULL, ylab="READS 2000", xlab="LS1034")
plot.default(x=df[["ls1034_200_gene"]], y=NULL, ylab="READS 200 GENE", xlab="LS1034")
plot.default(x=df[["ls1034_2000_gene"]], y=NULL, ylab="READS 2000 GENE", xlab="LS1034")
# H508
plot.default(x=df[["h508_gene"]], y=NULL, ylab="READS GENE", xlab="H508")
plot.default(x=df[["h508_200"]], y=NULL, ylab="READS 200", xlab="H508")
plot.default(x=df[["h508_2000"]], y=NULL, ylab="READS 2000", xlab="H508")
plot.default(x=df[["h508_200_gene"]], y=NULL, ylab="READS 200 GENE", xlab="H508")
plot.default(x=df[["h508_2000_gene"]], y=NULL, ylab="READS 2000 GENE", xlab="H508")


### Make scatterplots comparing two columns (categories or cell lines) ###
# Scatterplot
theme_set(theme_bw())  # pre-set the bw theme.
g <- ggplot(df, aes(snu61_gene, snu61_200))
g + geom_jitter(width = .5, size=1) +
  labs(y="200 bp upstream", 
       x="gene", 
       title="Read counts gene v 200 bp upstream")

#** log transforms? **#

### Make histograms ###
hist(df[["snu61_gene"]], breaks = 50, col="blue", main="snu61 gene")
hist(df[["ls1034_gene"]], breaks = 50, col="blue", main="ls1034 gene")
hist(df[["h508_gene"]], breaks = 50, col="blue", main="h508 gene")

hist(df[["snu61_200"]], breaks = 50, col="blue", main="snu61 200")

hist(df[["snu61_2000"]], breaks = 50, col="blue", main="snu61 2000")

hist(df[["snu61_200_gene"]], breaks = 50, col="blue", main="snu61 200 gene")

hist(df[["snu61_2000_gene"]], breaks = 50, col="blue", main="snu61 2000 gene")




## archived code (for now) subset granges for specific targets only (like from a text file) **

textfile <- "/home/rstudio1/atac/h508/smad4_entrez_targets.txt"
con <- file(description=textfile, open="r")

## Copy list of ENTREZ target IDS from text file
smad4_targets <- c()
for (i in 1:7897){
  smad4_targets[i] <- scan(file=con, nlines=1, quiet=TRUE)
}

#### end ####
close.connection(con=con)
#### Subset the granges object for specific targets ####
targets <- c(which(g@elementMetadata@listData[["entrezid"]] %in% smad4_targets))
g_smad4 <- g
# this metadata column is required to make the annot object
g_smad4@elementMetadata$id <- g_smad4@elementMetadata@listData[["gene_name"]]
#### end ####


reg_200up_counts <- list()
#
reg_200up_counts$snu61_r1 <- snu_r1_counts_200up
reg_200up_counts$snu61_r2 <- snu_r2_counts_200up
reg_200up_counts$snu61_r3 <- snu_r3_counts_200up
reg_200up_counts$ls1034_r1 <- ls1034_r1_counts_200up
reg_200up_counts$ls1034_r2 <- ls1034_r2_counts_200up
reg_200up_counts$ls1034_r3 <- ls1034_r3_counts_200up
reg_200up_counts$h508_r1 <- h508_r1_counts_200up
reg_200up_counts$h508_r2 <- h508_r2_counts_200up
reg_200up_counts$h508_r3 <- h508_r3_counts_200up


reg_2000up_counts <- list()
#
reg_2000up_counts$snu61_r1 <- snu_r1_counts_2000up
reg_2000up_counts$snu61_r2 <- snu_r2_counts_2000up
reg_2000up_counts$snu61_r3 <- snu_r3_counts_2000up
reg_2000up_counts$ls1034_r1 <- ls1034_r1_counts_2000up
reg_2000up_counts$ls1034_r2 <- ls1034_r2_counts_2000up
reg_2000up_counts$ls1034_r3 <- ls1034_r3_counts_2000up
reg_2000up_counts$h508_r1 <- h508_r1_counts_2000up
reg_2000up_counts$h508_r2 <- h508_r2_counts_2000up
reg_2000up_counts$h508_r3 <- h508_r3_counts_2000up


num_genes <- length(reg_200up_counts[["snu61_r1"]][["counts"]])
gene_names <- reg_200up_counts[["snu61_r1"]][["annotation"]][["GeneID"]]
#
reg_200_norm_counts <- list()
reg_200_norm_counts$snu61 <- matrix(data=NA, ncol=5, nrow=num_genes)
colnames(reg_200_norm_counts$snu61) <- c("rep1","rep2","rep3","avg","total")
rownames(reg_200_norm_counts$snu61) <- c(gene_names)
reg_200_norm_counts$ls1034 <- matrix(data=NA, ncol=5, nrow=num_genes)
colnames(reg_200_norm_counts$ls1034) <- c("rep1","rep2","rep3","avg","total")
rownames(reg_200_norm_counts$ls1034) <- c(gene_names)
reg_200_norm_counts$h508 <- matrix(data=NA, ncol=5, nrow=num_genes)
colnames(reg_200_norm_counts$h508) <- c("rep1","rep2","rep3","avg","total")
rownames(reg_200_norm_counts$h508) <- c(gene_names)
#
for (a in 1:num_genes){
  #
  reg_200_norm_counts$snu61[a,1] <- as.integer(reg_200up_counts[["snu61_r1"]][["counts"]][[a]]/replicate_normalization[1,2])
  reg_200_norm_counts$snu61[a,2] <- as.integer(reg_200up_counts[["snu61_r2"]][["counts"]][[a]]/replicate_normalization[2,2])
  reg_200_norm_counts$snu61[a,3] <- as.integer(reg_200up_counts[["snu61_r3"]][["counts"]][[a]]/replicate_normalization[3,2])
  reg_200_norm_counts$snu61[a,4] <- ((reg_200_norm_counts$snu61[a,1]+reg_200_norm_counts$snu61[a,2]+reg_200_norm_counts$snu61[a,3])/3)
  reg_200_norm_counts$snu61[a,5] <- reg_200_norm_counts$snu61[a,1]+reg_200_norm_counts$snu61[a,2]+reg_200_norm_counts$snu61[a,3]
  #
  reg_200_norm_counts$ls1034[a,1] <- as.integer(reg_200up_counts[["ls1034_r1"]][["counts"]][[a]]/replicate_normalization[4,2])
  reg_200_norm_counts$ls1034[a,2] <- as.integer(reg_200up_counts[["ls1034_r2"]][["counts"]][[a]]/replicate_normalization[5,2])
  reg_200_norm_counts$ls1034[a,3] <- as.integer(reg_200up_counts[["ls1034_r3"]][["counts"]][[a]]/replicate_normalization[6,2])
  reg_200_norm_counts$ls1034[a,4] <- ((reg_200_norm_counts$ls1034[a,1]+reg_200_norm_counts$ls1034[a,2]+reg_200_norm_counts$ls1034[a,3])/3)
  reg_200_norm_counts$ls1034[a,5] <- reg_200_norm_counts$ls1034[a,1]+reg_200_norm_counts$ls1034[a,2]+reg_200_norm_counts$ls1034[a,3]
  #
  reg_200_norm_counts$h508[a,1] <- as.integer(reg_200up_counts[["h508_r1"]][["counts"]][[a]]/replicate_normalization[7,2])
  reg_200_norm_counts$h508[a,2] <- as.integer(reg_200up_counts[["h508_r2"]][["counts"]][[a]]/replicate_normalization[8,2])
  reg_200_norm_counts$h508[a,3] <- as.integer(reg_200up_counts[["h508_r3"]][["counts"]][[a]]/replicate_normalization[9,2])
  reg_200_norm_counts$h508[a,4] <- ((reg_200_norm_counts$h508[a,1]+reg_200_norm_counts$h508[a,2]+reg_200_norm_counts$h508[a,3])/3)
  reg_200_norm_counts$h508[a,5] <- reg_200_norm_counts$h508[a,1]+reg_200_norm_counts$h508[a,2]+reg_200_norm_counts$h508[a,3]}
#

#
reg200_signal_matrix <- matrix(data=NA,nrow=num_genes,ncol=5)
rownames(reg200_signal_matrix) <- c(gene_names)
colnames(reg200_signal_matrix) <- c("snu61","ls1034","h508","var","sd")
#
for (b in 1:num_genes){
  #
  reg200_signal_matrix[b,1] <- as.integer((reg_200_norm_counts[["snu61"]][b,4]/sample_normalization[1,2]))
  reg200_signal_matrix[b,2] <- as.integer((reg_200_norm_counts[["ls1034"]][b,4]/sample_normalization[2,2]))
  reg200_signal_matrix[b,3] <- as.integer((reg_200_norm_counts[["h508"]][b,4]/sample_normalization[3,2]))
  reg200_signal_matrix[b,4] <- stats::var(reg200_signal_matrix[b,1:3])
  reg200_signal_matrix[b,5] <- stats::sd(reg200_signal_matrix[b,1:3])
}
#
coad_mr <- c("TCF7","MNX1","POU5F1B","ESRRA","CDX2","HNF4A","GMEB2","HOXA3","OVOL1","ASCL2","ZSWIM1","CBFA2T2","TARBP1","KLHL31","GTF2IRD1","ZNF696","NUFIP1","SUPT20H","KAT8","KAT2A","TAF4","RBM39","ZMYND8","L3MBTL1","PLAGL2","NCOA5","ZSWIM3","ADNP","ASXL1")
ind <- which(rownames(reg200_signal_matrix) %in% coad_mr)
reg200_mr_counts <- reg200_signal_matrix[ind,]


# Violin plots
# prep data
violin_data_reg200 <- data.frame("Signal" = c(reg200_mr_counts[,1],reg200_mr_counts[,2],reg200_mr_counts[,3]),
                                 "Sample" = c(rep("snu61", times=29), rep("ls1034",times=29),rep("h508",times=29)))
#
violin_plot_reg200 <- ggplot(violin_data_reg200, aes(x=Sample, y=Signal, fill=Sample)) + 
  geom_violin()
violin_plot_reg200


heatmap_data_reg200 <- data.frame("Signal" = c(reg200_mr_counts[,1],reg200_mr_counts[,2],reg200_mr_counts[,3]),
                                  "Sample" = c(rep("snu61", times=29), rep("ls1034",times=29),rep("h508",times=29)),
                                  "Gene" = rep(mr_names))

heatmap.plot.reg200 <- ggplot(data = heatmap_data_reg200, aes(x = Sample, y = Gene)) +
  geom_tile(aes(fill = Signal)) +
  scale_fill_gradient2(low="darkblue", high="red", guide="colorbar") +
  theme(axis.text.y = element_text(size = 6))

# Preview the heatmap
print(heatmap.plot.reg200)

#
mr_all_reps_reg200 <- matrix(data=NA, nrow=29, ncol=9)
rownames(mr_all_reps_reg200) <- mr_names
colnames(mr_all_reps_reg200) <- c("snu61_r1","snu61_r2","snu61_r3","ls1034_r1","ls1034_r2","ls1034_r3","h508_r1","h508_r2","h508_r3")
#
mr_all_reps_reg200[,1] <- reg_200_norm_counts[["snu61"]][ind,1]
mr_all_reps_reg200[,2] <- reg_200_norm_counts[["snu61"]][ind,2]
mr_all_reps_reg200[,3] <- reg_200_norm_counts[["snu61"]][ind,3]
mr_all_reps_reg200[,4] <- reg_200_norm_counts[["ls1034"]][ind,1]
mr_all_reps_reg200[,5] <- reg_200_norm_counts[["ls1034"]][ind,2]
mr_all_reps_reg200[,6] <- reg_200_norm_counts[["ls1034"]][ind,3]
mr_all_reps_reg200[,7] <- reg_200_norm_counts[["h508"]][ind,1]
mr_all_reps_reg200[,8] <- reg_200_norm_counts[["h508"]][ind,2]
mr_all_reps_reg200[,9] <- reg_200_norm_counts[["h508"]][ind,3]
#
heatmap_data_reps_reg200 <- data.frame(
  "Signal" = c(mr_all_reps_reg200[,1],mr_all_reps_reg200[,2],mr_all_reps_reg200[,3],mr_all_reps_reg200[,4],mr_all_reps_reg200[,5],mr_all_reps_reg200[,6],mr_all_reps_reg200[,7],mr_all_reps_reg200[,8],mr_all_reps_reg200[,9]),
  "Sample" = c(rep("snu61_r1", times=29), rep("snu61_r2", times=29), rep("snu61_r3", times=29), rep("ls1034_r1",times=29), rep("ls1034_r2",times=29), rep("ls1034_r3",times=29), rep("h508_r1",times=29), rep("h508_r2",times=29), rep("h508_r3",times=29)),
  "Gene" = rep(mr_names))
#
heatmap.plot.reps.reg200 <- ggplot(data = heatmap_data_reps_reg200, aes(x = Sample, y = Gene)) +
  geom_tile(aes(fill = Signal)) +
  scale_fill_gradient2(low="darkblue", high="red", guide="colorbar") +
  theme(axis.text.y = element_text(size = 6))

# Preview the heatmap
print(heatmap.plot.reps.reg200)





## load libs
library(ChIPpeakAnno)
library(genomation)
library(org.Hs.eg.db)

## import
snu61_peak <- readBed("/home/ubuntu1/atac/snu61/wt01/preprocessing/13allpeaks/SNU61-WT-01.all_summits.bed")
snu61_peak2 <- readBed("/home/ubuntu1/atac/snu61/wt01/preprocessing/13allpeaks/SNU61-WT-01.all_peaks.narrowPeak")




##
snu61.anno <- annotatePeakInBatch(snu61_peak, AnnotationData=TSS.human.GRCh38)

## Add gene symbols
snu61.anno <- addGeneIDs(annotatedPeak=snu61.anno, 
                         orgAnn="org.Hs.eg.db", 
                         IDs2Add="symbol")

## Find unique
snu61.genes <- unique(snu61.anno@elementMetadata@listData[["symbol"]])


x <- keepStandardChromosomes(snu61_peak)


## Write to file
write.table(snu61_ls1034_genes, file = "/home/ubuntu1/atac/peaks/snu61_ls1034_genes.txt", quote = FALSE, sep = ",", eol = "\n", row.names = FALSE, col.names = FALSE)

## Overlaps
ov <- findOverlaps(snu61_peak2, TSS.human.GRCh38)

######################################################################################

## annotate the peaks with precompiled ensembl annotation
data(TSS.human.GRCh38)
## Do some coercion to convert the TSS.human GRANGEs seqnames to standard format (with chr)
newseqs <- paste0("chr", c(1:22, "X", "Y"))
names(newseqs) <- c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10",
                    "11", "12", "13", "14", "15", "16", "17", "18", "19", "20",
                    "21", "22", "X", "Y")
## Rename the sequences
newTSS <- renameSeqlevels(TSS.human.GRCh38, newseqs)
## Find overlaps
nov <- findOverlaps(snu61_peak, newTSS)
nov2 <- unique(nov@to)
##
genes <- TSS.human.GRCh38@elementMetadata@listData[["description"]][nov2]
genes <- genes[-which(is.null(genes) == FALSE)]
genes <- genes[-which(genes == "")]




```

Run AR footprints
```{r}
install.packages("gridSVG")
library(gridSVG)

library(MotifDb)
library(BSgenome.Hsapiens.UCSC.hg38)
library(Biostrings)
suppressMessages(library(BSgenome.Hsapiens.UCSC.hg38))
suppressMessages(library(Biostrings))
suppressMessages(library(MotifDb))
suppressMessages(library(GenomicRanges))
suppressMessages(library(Rsamtools))
suppressMessages(library(GenomicAlignments))
suppressMessages(library(ChIPseeker))
suppressMessages(library(TxDb.Hsapiens.UCSC.hg38.knownGene))
suppressMessages(library(genomation))
suppressMessages(library(stringr))
genome <- Hsapiens
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene

#### bam paths
path.CR01.bam <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bam\\LNCaP-CR-01.rep1.refhg38.bam"
path.CR02.bam <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bam\\LNCaP-CR-02.rep1.refhg38.bam"
path.CR04.bam <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bam\\LNCaP-CR-04.rep1.refhg38.bam"
path.CR05.bam <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bam\\LNCaP-CR-05.rep1.refhg38.bam"
path.CR07.bam <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bam\\LNCaP-CR-07.rep1.refhg38.bam"
path.CR08.bam <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bam\\LNCaP-CR-08.rep1.refhg38.bam"
path.WT01.bam <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bam\\LNCaP-WT-01.rep1.refhg38.bam"
path.WT02.bam <<- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\bam\\LNCaP-WT-02.rep1.refhg38.bam"

ARsitesPath <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\sites\\AR.bindingSites.Rdata"
load(ARsitesPath)
numSites <- AR.bindingSites[["numSites"]]

sampleMergedPeaksPath <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\peaks\\sample_merged\\LNCaP-CR-01.rep1.refhg38_sample_merged_peaks.narrowPeak"
importBED <- function(bedPath){
  bedin <- readBed(bedPath, track.line = FALSE, remove.unusual = FALSE, zero.based = TRUE)
  bedin <- keepStandardChromosomes(bedin, pruning.mode = "coarse")
  bedin <- trim(bedin)
  return(bedin)
}
sampleMergedPeaks <- importBED(sampleMergedPeaksPath)

getGRoverlapIndex <- function(input1, input2){
  overlap <- findOverlaps(input1, input2)
  index <- unique(overlap@from)
  return(index)
}

bindingSites <- AR.bindingSites[["bindingSites"]]

#### Subset the binding sites with the peaks
cat("Subsetting sites to peak regions", "\n")
peaksGR <- importBED(sampleMergedPeaksPath)
bindingSites <- subsetByOverlaps(bindingSites, peaksGR)
numSites <- length(bindingSites@ranges)
cat("Identified", numSites, "total binding sites", "\n")

#### Determine scope for current binding sites ####
maxWidth <- max(bindingSites@ranges@width)
scope <- paste0("chr", c(1:22, "X", "Y"))
currentScope <- scope[which(scope %in% bindingSites@seqnames@values)]

bamFile <- path.WT01.bam

chr1ins <- generateInsertionMatrixByChr(bamFile, bindingSites, maxWidth, "chr1")

out <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\plots\\plot1.svg"

#### Generate a motif-aligned signal heatmap ####
plotInsertionSiteProbability(chr1ins, 250, 15, "CR01CR01", out)


##################
calculateBasicFootprintStatistics <- function(insMatrix, bindingSites, chrName){
  
  numSites <- length(insMatrix[,1])
  rawSiteBasicStats <- matrix(data = NA, nrow = numSites, ncol = 19)
  
  colnames(rawSiteBasicStats) <- c("Chromosome", "Motif Start", "Motif Width",
                                   "Motif Score", "Motif Score 2",
                                   "Total signal", "Total signal per bp", "Motif signal per bp",
                                   "Flank signal per bp", "Background signal per bp", "Wide flank signal per bp",
                                   "Flank / Background", "Motif / Flank", "Motif / Wide Flank",
                                   "Binding", "p-value", "z-score", "Annotation", "Closest gene")
  
  rawSiteBasicStats[,1] <- chrName
  rawSiteBasicStats[,2] <- as.numeric(bindingSites@ranges@start)
  rawSiteBasicStats[,3] <- as.numeric(bindingSites@ranges@width)
  rawSiteBasicStats[,4] <- as.numeric(bindingSites@elementMetadata@listData[["score"]])
  rawSiteBasicStats[,5] <- as.numeric(bindingSites@elementMetadata@listData[["score"]])
  
  ## Annotate
  annotations <- annotatePeak(
    peak = bindingSites,
    tssRegion = c(-3000, 3000),
    TxDb = TxDb.Hsapiens.UCSC.hg38.knownGene,
    level = "gene",
    assignGenomicAnnotation = TRUE,
    genomicAnnotationPriority = c("Promoter", "5UTR", "3UTR", "Exon", "Intron", "Downstream", "Intergenic"),
    annoDb = "org.Hs.eg.db",
    addFlankGeneInfo = FALSE,
    flankDistance = 5000,
    sameStrand = FALSE,
    ignoreOverlap = FALSE,
    ignoreUpstream = FALSE,
    ignoreDownstream = FALSE,
    overlap = "TSS",
    verbose = TRUE)
  
  rawSiteBasicStats[,18] <- annotations@anno@elementMetadata@listData[["annotation"]]
  rawSiteBasicStats[,19] <- annotations@anno@elementMetadata@listData[["SYMBOL"]]
  
  for (b in 1:numSites){
    tempWidth <- as.numeric(rawSiteBasicStats[b,3])
    tempTotalWidth <- as.numeric(tempWidth + 500)
    tempTotalSignal <- as.numeric(sum(insMatrix[b,1:tempTotalWidth]))
    tempTotalSignalPerBP <- as.numeric(tempTotalSignal/tempTotalWidth)
    tempMotifSignalPerBP <- as.numeric(sum(insMatrix[b,(250:(250 + tempWidth))] / tempWidth))
    tempFlankSignalPerBP <- as.numeric((sum(insMatrix[b,200:250])+sum(insMatrix[b,(250+tempWidth):(300+tempWidth)])) / 100)
    tempBackgroundSignalPerBP <- as.numeric((sum(insMatrix[b,1:50])+sum(insMatrix[b,(500+tempWidth-50):(500+tempWidth)])) / 100)
    tempWideFlankSignalPerBP <- as.numeric((sum(insMatrix[b,1:250])+sum(insMatrix[b,(250+tempWidth):(500+tempWidth)])) / 500)
    ##
    rawSiteBasicStats[b,6] <- as.numeric(tempTotalSignal)
    rawSiteBasicStats[b,7] <- as.numeric(tempTotalSignalPerBP)
    rawSiteBasicStats[b,8] <- as.numeric(tempMotifSignalPerBP)
    rawSiteBasicStats[b,9] <- as.numeric(tempFlankSignalPerBP)
    rawSiteBasicStats[b,10] <- as.numeric(tempBackgroundSignalPerBP)
    rawSiteBasicStats[b,11] <- as.numeric(tempWideFlankSignalPerBP)
    rawSiteBasicStats[b,12] <- as.numeric(tempFlankSignalPerBP / tempBackgroundSignalPerBP)
    rawSiteBasicStats[b,13] <- as.numeric(tempMotifSignalPerBP / tempFlankSignalPerBP)
    rawSiteBasicStats[b,14] <- as.numeric(tempMotifSignalPerBP / tempWideFlankSignalPerBP)
    
    if (tempTotalSignal == 0){
      
    } else {
      averageNullMotifSignal <- generateNullFP(1000, tempTotalSignal, tempTotalWidth, tempWidth)
      motifSignals <- c(insMatrix[b,(250:(250 + tempWidth))])
      
      result = tryCatch({
        ttest <- t.test(motifSignals, mu = averageNullMotifSignal, alternative = "less", conf.level = 0.95)
        pvalue <- ttest$p.value
        rawSiteBasicStats[b,16] <- pvalue
        rawSiteBasicStats[b,17] <- qnorm(pvalue)
        if (pvalue < 0.05){rawSiteBasicStats[b,15] <- 1} else {rawSiteBasicStats[b,15] <- 0}
      }, warning = function(w) {
      }, error = function(e) {
      }, finally = {
      })}
  }
  
  #### Convert to a dataframe before return ####
  ## This allows you to store different data types
  cat("Transferring data to dataframe", "\n")
  rawFootprintStats <- data.frame(
    chr = as.character(rawSiteBasicStats[,1]),
    start = as.numeric(rawSiteBasicStats[,2]),
    width = as.numeric(rawSiteBasicStats[,3]),
    score = as.numeric(rawSiteBasicStats[,4]),
    score2 = as.numeric(rawSiteBasicStats[,5]),
    totalSignal = as.numeric(rawSiteBasicStats[,6]),
    totalSignalPerBP = as.numeric(rawSiteBasicStats[,7]),
    motifSignalPerBP = as.numeric(rawSiteBasicStats[,8]),
    flankSignalPerBP = as.numeric(rawSiteBasicStats[,9]),
    backgroundSignalPerBP = as.numeric(rawSiteBasicStats[,10]),
    wideFlankSignalPerBP = as.numeric(rawSiteBasicStats[,11]),
    flankAccessibility = as.numeric(rawSiteBasicStats[,12]),
    footprintDepth = as.numeric(rawSiteBasicStats[,13]),
    wideAccessibility = as.numeric(rawSiteBasicStats[,14]),
    binding = as.numeric(rawSiteBasicStats[,15]),
    pvalue = as.numeric(rawSiteBasicStats[,16]),
    zscore = as.numeric(rawSiteBasicStats[,17]),
    annotation = as.character(rawSiteBasicStats[,18]),
    symbol = as.character(rawSiteBasicStats[,19])
  )
  cat("Returning dataframe", "\n")
  return(rawFootprintStats)
}

#####
footstats <- calculateBasicFootprintStatistics(chr1ins, bindingSites, "chr1")


```

---
output: html_document
editor_options: 
  chunk_output_type: console
---

Functions
```{r}

#### RUN THIS CODE BEFORE USING FUNCTIONS ####
suppressMessages(library(BSgenome.Hsapiens.UCSC.hg38))
suppressMessages(library(Biostrings))
suppressMessages(library(MotifDb))
suppressMessages(library(GenomicRanges))
suppressMessages(library(Rsamtools))
suppressMessages(library(GenomicAlignments))
suppressMessages(library(genomation))
genome <- Hsapiens

#### Query motifDB to return binding sites of given gene ####
getAllBindingSites <- function(gene, pwmScanScore = "99%"){
  
  cat("querying motifDB", "\n")
  mdbHuman <- query(MotifDb, 'hsapiens')
  cat(length(mdbHuman), "records in database", "\n")
  geneIdx <- which(mdbHuman@elementMetadata@listData[["geneSymbol"]] == gene)
  cat(geneIdx, "\n")
  cat("Found", length(geneIdx), "records matching current gene", "\n")
  cat("retrieving relevant records from mdb", "\n")
  tempMotifs <- list()
  c <- 1
  for (idx in geneIdx){
    tempMotifs[c] <- mdbHuman@listData[idx]
    c <- c+1}
  cat("finding unique motifs", "\n")
  uniqueMotifs <- unique(tempMotifs)
  numUniqueMotifs <- length(uniqueMotifs)
  cat("found", numUniqueMotifs, "unique motifs", "\n")
  if (numUniqueMotifs > 1){
    cat("processing more than one unique motif", "\n")
    PWM <- uniqueMotifs[[1]]
    allSites <- Biostrings::matchPWM(PWM, genome, min.score = pwmScanScore, with.score = TRUE)
    allSites <- keepStandardChromosomes(allSites, pruning.mode = "coarse")
    allSites <- trim(allSites)
    allSites@elementMetadata@listData$score2 <- allSites@elementMetadata@listData[["score"]] / max(allSites@elementMetadata@listData[["score"]])
    for (a in 2:numUniqueMotifs){
      com <- paste0("PWM <- uniqueMotifs[[", a, "]]")
      eval(parse(text = com))
      sitesTemp <- Biostrings::matchPWM(PWM, genome, min.score = pwmScanScore, with.score = TRUE)
      sitesTemp <- keepStandardChromosomes(sitesTemp, pruning.mode = "coarse")
      sitesTemp <- trim(sitesTemp)
      sitesTemp@elementMetadata@listData$score2 <- sitesTemp@elementMetadata@listData[["score"]] / max(sitesTemp@elementMetadata@listData[["score"]])
      allSites <- c(allSites, sitesTemp)}
  } else {
    cat("processing one unique motif", "\n")
    PWM <- uniqueMotifs[[1]]
    allSites <- Biostrings::matchPWM(PWM, genome, min.score = pwmScanScore, with.score = TRUE)
    allSites <- keepStandardChromosomes(allSites, pruning.mode = "coarse")
    allSites <- trim(allSites)
    allSites@elementMetadata@listData$score2 <- allSites@elementMetadata@listData[["score"]] / max(allSites@elementMetadata@listData[["score"]])
  }
  
  ## Remove chrM entries, if present
  cat("Removing mitochondrial binding sites, if present", "\n")
  indexChrM <- which(seqnames(allSites) == "chrM")
  if (length(indexChrM > 0)){
    allSites <- allSites[-indexChrM]
  }
  
  cat("returning allSites", "\n")
  return(allSites)
}

#### Import a BED file as a GRanges object ####
importBED <- function(bedPath){
  bedin <- readBed(bedPath, track.line = FALSE, remove.unusual = FALSE, zero.based = TRUE)
  bedin <- keepStandardChromosomes(bedin, pruning.mode = "coarse")
  bedin <- trim(bedin)
  return(bedin)
}

generateNullFP <- function(iterations, inputSignal, analysisWidth, motifWidth){
  # This script will be used to generate indiviudal null models at predicted motif binding sites across the genome when scanning for TF footprinting from ATAC-seq data. To generate these null models, the current model will need to:
  # Consider the total signal (number of insertions) at each specific ~200 bp locus
  # Use the actul underlying reference sequence of that ~200 bp stretch from the hg38 reference genome
  # Use published or experimentally derived models of Tn5 sequence specific insertion bias
  # For each locus, build a probablistic model of insertion site distributions based on the underlying sequence and Tn5 insertion bias
  # Generate the null model graph by weighted random residstribution of the total observed signal at that site
  # Importantly, the null model must be generated separately for the plus and minus strand, it can then be combined and compared to the combined signal from the reference observed signal at that sequence
  # These null models can then be used for a site-by-site comparison of the null model against the observed data to accept or reject the null hypothesis
  # iterations = number of iterations
  # inputSignals = unique values for total signal
  # analysisWidth = total bp in region of interest (flank + background + motif)
  # motifWidth = motif width
  
  # declare vector of size n to store average motif signal values
  averages <- c()
  
  # generate the null models and calculate motif averages
  for (a in 1:iterations){
    
    # declare the null vector
    null <- c(1:(analysisWidth))
    
    # randomly distribute the total signal
    # size = the number of values to distribute
    # prob = probability of each site
    # length = length of the generated vector
    null <- c(as.vector(rmultinom(1, size=inputSignal, prob=rep(1, length(null)))))
    
    ## Calculate the mean signal in motif region
    motifStart <- ((analysisWidth - motifWidth)/2)
    motifEnd <- (motifStart + motifWidth)
    motifAvg <- (sum(null[motifStart:motifEnd])) / motifWidth
    
    ## Store the average values
    averages[a] <- motifAvg
    
  } # end for (a in 1:n)
  return(averages)
} # end generateNullFP function

analyzeFP <- function(geneName, bindingSites, sampleName, sampleBamPath, sampleBaiPath){

  ##
  inputSites <- bindingSites
  ## ???
  inputSites@elementType <- "GENE"
  
  ##
  cat("Checking number of sites and motif width", "\n")
  motifWidth <- length(bindingSites[["PWM"]][1,])
  numSites <- length(inputSites)
  
  ##
  cat("Extending sites", "\n")
  extSites <- promoters(inputSites, upstream = 250, downstream = (250 + motifWidth), use.names = TRUE)
  extSites <- keepStandardChromosomes(extSites, pruning.mode = "coarse")
  extSites <- trim(extSites)
  
  ##
  cat("Loading reads from bam file", "\n")
  param <- ScanBamParam(which = extSites)
  bamIn <- readGAlignments(sampleBamPath, index = sampleBaiPath, param = param)
  grIn <- granges(bamIn)
  grIn <- keepStandardChromosomes(grIn, pruning.mode="coarse")
  grIn <- trim(grIn)
  
  ##
  cat("Resizing reads and shifting +4/-5", "\n")
  grIn2 <- resize(grIn, width = 1)
  plusIdx <- which(strand(grIn2) == "+")
  minusIdx <- which(strand(grIn2) == "-")
  grPlus <- grIn2[plusIdx]
  grMinus <- grIn2[minusIdx]
  grPlusShifted <- shift(grPlus, shift=4L)
  grMinusShifted <- shift(grMinus, shift=-5L)
  grMerged <- c(grPlusShifted, grMinusShifted)
  shiftedInsertions <- grMerged
  
  ##
  cat("Generating insertion matrix", "\n")
  insRLE <- coverage(grMerged)
  insViews <- Views(insRLE, extSites)
  insMatrix <- as.matrix(insViews)
  
  ##
  cat("Calculating normalization factors", "\n")
  libSize <- length(bamIn)
  coverageSize <- sum(as.numeric(width(reduce(grIn, ignore.strand=TRUE))))
  libFactor <- libSize / coverageSize
  
  ##
  cat("Calculating raw sites basic footprint stats", "\n")
  rawSiteBasicStats <- matrix(data = NA, nrow = numSites, ncol = 10)
  colnames(rawSiteBasicStats) <- c("Site index", "Total signal", "Total signal per bp", "Motif signal per bp",
                                 "Flank signal per bp", "Background signal per bp", "Wide flank signal per bp",
                                 "Flank / Background", "Motif / Flank", "Motif / Wide Flank")
  
  ##
  for (b in 1:numSites){
    rawSiteBasicStats[b,1] <- b # Site index
    rawSiteBasicStats[b,2] <- sum(insMatrix[b,]) # Total signal
    rawSiteBasicStats[b,3] <- rawSiteBasicStats[b,2] / (500 + motifWidth) # Total signal per bp
    rawSiteBasicStats[b,4] <- sum(insMatrix[b,(250:(250 + motifWidth))] / motifWidth) # Motif signal per bp
    rawSiteBasicStats[b,5] <- (sum(insMatrix[b,200:250]) + sum(insMatrix[b,(250 + motifWidth):(300 + motifWidth)])) / 100 # Flank signal per bp
    rawSiteBasicStats[b,6] <- (sum(insMatrix[b,1:50]) + sum(insMatrix[b,(500 + motifWidth-50):(500 + motifWidth)])) / 100 # Background signal per bp
    rawSiteBasicStats[b,7] <- (sum(insMatrix[b,1:250]) + sum(insMatrix[b,(250 + motifWidth):(500 + motifWidth)])) / 500 # Wide flank signal per bp
    rawSiteBasicStats[b,8] <- rawSiteBasicStats[b,5] / rawSiteBasicStats[b,6] # Flank / background
    rawSiteBasicStats[b,9] <- rawSiteBasicStats[b,4] / rawSiteBasicStats[b,5] # Motif / flank
    rawSiteBasicStats[b,10] <- rawSiteBasicStats[b,4] / rawSiteBasicStats[b,7] # Motif / wide flank
  } # end for (b in 1:numSites)
  
  ##
  cat("Analyzing insertion probability", "\n")
  rawInsProb <- c()
  for (c in 1:(500 + motifWidth)){
    rawInsProb[c] <- sum(insMatrix[,c])} # end for (c in 1:(500 + motifWidth))
  
  ##
  cat("Finding raw total signal", "\n")
  rawTotalSignal<- sum(rawInsProb)
  rawInsProb <- rawInsProb / rawTotalSignal
  uniqueTotalSignals <- unique(rawSiteBasicStats[,2])
  siteWidth <- 500 + motifWidth
  
  ##
  cat("Generating null models", "\n")
  nullModels <- matrix(data = NA, ncol = 2, nrow = length(uniqueTotalSignals))
  colnames(nullModels) <- c("Input signal", "Avg motif signal in null model")
  for (c in 1:length(uniqueTotalSignals)){
    nullVec <- generateNullFP(1000, uniqueTotalSignals[c], siteWidth, motifWidth)
    nullModels[c,1] <- uniqueTotalSignals[c]
    nullModels[c,2] <- mean(nullVec)
  } # end for (c in 1:length(uniqueTotalSignals))
  
  ##
  cat("Performing hypothesis testing", "\n")
  ttest <- list()
  pvalue <- c()
  tvalue <- c()
  for (d in 1:numSites){
    currentSignal <- c(rawSiteBasicStats[d,2])
    currentNullModel <- nullModels[which(nullModels[,1]==currentSignal),2]
    ttest[[d]] <- t.test(insMatrix[d,250:(250+motifWidth)], mu=currentNullModel, alternative="less", conf.level = 0.95)
    pvalue[d] <- ttest[[d]][["p.value"]]
    tvalue[d] <- ttest[[d]][["statistic"]][["t"]]} # end for (d in 1:numSites)
  
  ##
  cat("Performing test corrections", "\n")
  idxPvaluePass <- which(pvalue < 0.05)
  pvaluePass <- pvalue[idxPvaluePass]
  ppassNumSites <- length(idxPvaluePass)
  idxBFpass <- which(pvalue < (0.05 / numSites))
  BFpvaluePass <- pvalue[idxBFpass]
  BFpassNumSites <- length(idxBFpass)
  BHpvalue <- p.adjust(pvalue, method = "BH")
  idxBHpass <- which(BHpvalue < 0.05)
  BHpvaluePass <- pvalue[idxBHpass]
  BHpassNumSites <- length(idxBHpass)
  
  ##
  cat("Transferring data to storage list", "\n")
  tempList <- list()
  tempList$geneName <- geneName
  tempList$sampleName <- sampleName
  tempList$sampleBamPath <- sampleBamPath
  tempList$sampleBaiPath <- sampleBaiPath
  tempList$numSites <- numSites
  tempList$motifWidth <- motifWidth
  tempList$PWM <- bindingSites[["PWM"]]
  tempList$minPWMscore <- bindingSites[["minPWMscore"]]
  tempList$maxPWMscore <- bindingSites[["maxPWMscore"]]
  tempList$inputSites <- inputSites
  tempList$extSites <- extSites
  tempList$insMatrix <- insMatrix
  tempList$libSize <- libSize
  tempList$coverageSize <- coverageSize
  tempList$libFactor <- libFactor
  tempList$rawSiteBasicStats <- rawSiteBasicStats
  tempList$rawInsProb <- rawInsProb
  tempList$ttest <- ttest
  tempList$pvalue <- pvalue
  tempList$tvalue <- tvalue
  tempList$idxPvaluePass <- idxPvaluePass
  tempList$pvaluePass <- pvaluePass
  tempList$ppassNumSites <- ppassNumSites
  tempList$idxBFpass <- idxBFpass
  tempList$BFpvaluePass <- BFpvaluePass
  tempList$BFpassNumSites <- BFpassNumSites
  tempList$BHpvalue <- BHpvalue
  tempList$idxBHpass <- idxBHpass
  tempList$BHpvaluePass <- BHpvaluePass
  tempList$BHpassNumSites <- BHpassNumSites
  
  ##
  cat("Returning data", "\n")
  return(tempList)
  
}

importBED <- function(bedPath){
  bedin <- readBed(bedPath, track.line = FALSE, remove.unusual = FALSE, zero.based = TRUE)
  bedin <- keepStandardChromosomes(bedin, pruning.mode="coarse")
  bedin <- trim(bedin)
  return(bedin)
}

getGRoverlapIndex <- function(input1, input2){
  overlap <- findOverlaps(input1, input2)
  index <- unique(overlap@from)
  return(index)
}



```

GENERATE AR BINDING SITES OBJECT
```{r}

##
currentGene <- "AR"
score <- "80%"
genome <- Hsapiens

##
AR.bindingSites.80 <- getAllBindingSites(currentGene, score)
numSites <- length(AR.bindingSites.80)
minPWMscore <- min(AR.bindingSites.80@elementMetadata@listData[["score"]])
maxPWMscore <- max(AR.bindingSites.80@elementMetadata@listData[["score"]])

#### Subset the binding sites with the peaks
cat("Subsetting sites to peak regions", "\n")
peaksGR <- importBED(sampleMergedPeaks)
AR.bindingSites.80 <- subsetByOverlaps(AR.bindingSites.80, peaksGR)
numSites <- length(AR.bindingSites.80@ranges)


```

SETUP LOAD LIBRARIES AND FUNCTIONS
```{r}
#### Disable scientific notation in variables
hg38TotalBP <- 3272116950
suppressMessages(library(BSgenome.Hsapiens.UCSC.hg38))
suppressMessages(library(Biostrings))
suppressMessages(library(GenomicRanges))
suppressMessages(library(Rsamtools))
suppressMessages(library(GenomicAlignments))
suppressMessages(library(stats4))
suppressMessages(library(BiocGenerics))
suppressMessages(library(parallel))
suppressMessages(library(genomation))
suppressMessages(library(seqLogo))
suppressMessages(library(ChIPpeakAnno))
suppressMessages(library(rlist))
suppressMessages(library(TxDb.Hsapiens.UCSC.hg38.knownGene))
suppressMessages(library(pheatmap))
suppressMessages(library(org.Hs.eg.db))
library(pheatmap)
library(dplyr)

## Bam and bai
##
FPdata.LNCaP.AR <- list()

bam.WT01 <- "C:\\Users\\jsk33\\OneDrive\\R\\LNCaPCodeAlessandroAndPrem\\bam\\LNCaP-WT-01-REP1.u.bam"
bai.WT01 <-"C:\\Users\\jsk33\\OneDrive\\R\\LNCaPCodeAlessandroAndPrem\\bam\\LNCaP-WT-01-REP1.u.bam.bai"
bam.WT02 <- "C:\\Users\\jsk33\\OneDrive\\R\\LNCaPCodeAlessandroAndPrem\\bam\\LNCaP-WT-02-REP1.u.bam"
bai.WT02 <-"C:\\Users\\jsk33\\OneDrive\\R\\LNCaPCodeAlessandroAndPrem\\bam\\LNCaP-WT-02-REP1.u.bam.bai"
bam.CR01 <- "C:\\Users\\jsk33\\OneDrive\\R\\LNCaPCodeAlessandroAndPrem\\bam\\LNCaP-CR-01-REP1.u.bam"
bai.CR01 <-"C:\\Users\\jsk33\\OneDrive\\R\\LNCaPCodeAlessandroAndPrem\\bam\\LNCaP-CR-01-REP1.u.bam.bai"
bam.CR02 <- "C:\\Users\\jsk33\\OneDrive\\R\\LNCaPCodeAlessandroAndPrem\\bam\\LNCaP-CR-02-REP1.u.bam"
bai.CR02 <-"C:\\Users\\jsk33\\OneDrive\\R\\LNCaPCodeAlessandroAndPrem\\bam\\LNCaP-CR-02-REP1.u.bam.bai"
bam.CR04 <- "C:\\Users\\jsk33\\OneDrive\\R\\LNCaPCodeAlessandroAndPrem\\bam\\LNCaP-CR-04-REP1.u.bam"
bai.CR04 <-"C:\\Users\\jsk33\\OneDrive\\R\\LNCaPCodeAlessandroAndPrem\\bam\\LNCaP-CR-04-REP1.u.bam.bai"
bam.CR05 <- "C:\\Users\\jsk33\\OneDrive\\R\\LNCaPCodeAlessandroAndPrem\\bam\\LNCaP-CR-05-REP1.u.bam"
bai.CR05 <-"C:\\Users\\jsk33\\OneDrive\\R\\LNCaPCodeAlessandroAndPrem\\bam\\LNCaP-CR-05-REP1.u.bam.bai"
bam.CR07 <- "C:\\Users\\jsk33\\OneDrive\\R\\LNCaPCodeAlessandroAndPrem\\bam\\LNCaP-CR-07-REP1.u.bam"
bai.CR07 <-"C:\\Users\\jsk33\\OneDrive\\R\\LNCaPCodeAlessandroAndPrem\\bam\\LNCaP-CR-07-REP1.u.bam.bai"
bam.CR08 <- "C:\\Users\\jsk33\\OneDrive\\R\\LNCaPCodeAlessandroAndPrem\\bam\\LNCaP-CR-08-REP1.u.bam"
bai.CR08 <-"C:\\Users\\jsk33\\OneDrive\\R\\LNCaPCodeAlessandroAndPrem\\bam\\LNCaP-CR-08-REP1.u.bam.bai"
sampleMergedPeaks <- "C:\\Users\\jsk33\\OneDrive\\R\\LNCaPCodeAlessandroAndPrem\\peaks\\LNCaP-CR-01.rep1.refhg38_sample_merged_peaks.narrowPeak"


##
FPdata.LNCaP.AR$WT01 <- analyzeFP("AR", AR.bindingSites.80, "WT01", bam.WT01, bai.WT01)
FPdata.LNCaP.AR$WT02 <- analyzeFP("AR", AR.bindingSites, "WT02", bam.WT02, bai.WT02)
FPdata.LNCaP.AR$CR01 <- analyzeFP("AR", AR.bindingSites, "CR01", bam.CR01, bai.CR01)
FPdata.LNCaP.AR$CR02 <- analyzeFP("AR", AR.bindingSites, "CR02", bam.CR02, bai.CR02)
FPdata.LNCaP.AR$CR04 <- analyzeFP("AR", AR.bindingSites, "CR04", bam.CR04, bai.CR04)
FPdata.LNCaP.AR$CR05 <- analyzeFP("AR", AR.bindingSites, "CR05", bam.CR05, bai.CR05)
FPdata.LNCaP.AR$CR07 <- analyzeFP("AR", AR.bindingSites, "CR07", bam.CR07, bai.CR07)
FPdata.LNCaP.AR$CR08 <- analyzeFP("AR", AR.bindingSites, "CR08", bam.CR08, bai.CR08)

```

AR - generate FPdata object
```{r}




##
ins <- FPdata.LNCaP.AR[["WT01"]][["insMatrix"]]
out <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\plots\\plot1wt01.svg"
idx <- FPdata.LNCaP.AR[["WT01"]][["idxPvaluePass"]]
ins2 <- ins[idx,]
out2 <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\plots\\plot2wt01.svg"
plotInsertionSiteProbability(ins2, 250, 15, "WT01", out2)

##
ins <- FPdata.LNCaP.AR[["WT02"]][["insMatrix"]]
out <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\plots\\plot1wt02.svg"
plotInsertionSiteProbability(ins, 250, 15, "WT02", out)
idx <- FPdata.LNCaP.AR[["WT02"]][["idxPvaluePass"]]
ins2 <- ins[idx,]
out2 <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\plots\\plot2wt02.svg"
plotInsertionSiteProbability(ins2, 250, 15, "WT02", out2)

##
ins <- FPdata.LNCaP.AR[["CR01"]][["insMatrix"]]
out <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\plots\\plot1cr01.svg"
plotInsertionSiteProbability(ins, 250, 15, "CR01", out)
idx <- FPdata.LNCaP.AR[["CR01"]][["idxPvaluePass"]]
ins2 <- ins[idx,]
out2 <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\plots\\plot2cr01.svg"
plotInsertionSiteProbability(ins2, 250, 15, "CR01", out2)

##
ins <- FPdata.LNCaP.AR[["CR02"]][["insMatrix"]]
out <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\plots\\plot1cr02.svg"
plotInsertionSiteProbability(ins, 250, 15, "CR02", out)
idx <- FPdata.LNCaP.AR[["CR02"]][["idxPvaluePass"]]
ins2 <- ins[idx,]
out2 <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\plots\\plot2cr02.svg"
plotInsertionSiteProbability(ins2, 250, 15, "CR02", out2)

##
ins <- FPdata.LNCaP.AR[["CR04"]][["insMatrix"]]
out <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\plots\\plot1cr04.svg"
plotInsertionSiteProbability(ins, 250, 15, "CR04", out)
idx <- FPdata.LNCaP.AR[["CR04"]][["idxPvaluePass"]]
ins2 <- ins[idx,]
out2 <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\plots\\plot2cr04.svg"
plotInsertionSiteProbability(ins2, 250, 15, "CR04", out2)

##
ins <- FPdata.LNCaP.AR[["CR05"]][["insMatrix"]]
out <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\plots\\plot1cr05.svg"
plotInsertionSiteProbability(ins, 250, 15, "CR05", out)
idx <- FPdata.LNCaP.AR[["CR05"]][["idxPvaluePass"]]
ins2 <- ins[idx,]
out2 <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\plots\\plot2cr05.svg"
plotInsertionSiteProbability(ins2, 250, 15, "CR05", out2)

##
ins <- FPdata.LNCaP.AR[["CR07"]][["insMatrix"]]
out <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\plots\\plot1cr07.svg"
plotInsertionSiteProbability(ins, 250, 15, "CR07", out)
idx <- FPdata.LNCaP.AR[["CR07"]][["idxPvaluePass"]]
ins2 <- ins[idx,]
out2 <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\plots\\plot2cr07.svg"
plotInsertionSiteProbability(ins2, 250, 15, "CR07", out2)

##
ins <- FPdata.LNCaP.AR[["CR08"]][["insMatrix"]]
out <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\plots\\plot1cr08.svg"
plotInsertionSiteProbability(ins, 250, 15, "CR08", out)
idx <- FPdata.LNCaP.AR[["CR08"]][["idxPvaluePass"]]
ins2 <- ins[idx,]
out2 <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\plots\\plot2cr08.svg"
plotInsertionSiteProbability(ins2, 250, 15, "CR08", out2)


#### adjust counts with seqbias
modelpath <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\footprints\\wt01.tn5model.yml"
refFastaPath <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\data\\footprints\\chr1.fa"
model <- seqbias.load(refFastaPath, modelpath)
extSites <- promoters(AR.bindingSites[["bindingSites"]], upstream = 250, downstream = (250 + motifWidth), use.names = TRUE)

seqBiasPrediction <- seqbias.predict(model, extSites)
## Invert the prediction to be able to use as a probability vector
inverseSeqBiasPrediction <- 1 / seqBiasPrediction[[1]]

numSites <- length(ins2[,1])
insCorr <- ins2

for (a in 1:numSites){
  seq <- seqBiasPrediction[[idx[a]]]
  insCorr[a,] <- insCorr[a,] / seq
}

out2 <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\plots\\corr.svg"
plotInsertionSiteProbability(insCorr, 250, 15, "WT01", out2)




#### Generate a motif-aligned signal heatmap ####
plotMotifAlignedHeatmap <- function(insertionMatrix, bindingSites, plotTitle, svgOutPath){
  
  #### Margin controls ####
  # margin(a,b,c,d)
  # a = size of graph from top to bottom, higher value = smaller. default = 0.1
  # b = size of graph from left to right, higher value = smaller. default = 0.005
  # c = flips x axis?
  # d = margin from right side of page, higher = smaller. set at 0.2 so legends dont overlap
  # good settings for ATACseq = c(0.1,0.005,0.05,0.2)
  # bias setting >1 puts more colors at higher values, very useful for dealing with washout of low values
  # upper.extreme controls to heatmap scale
  
  ## Get the matrix data
  numSites <- length(insertionMatrix[,1])
  numBP <- length(insertionMatrix[1,])
  
  ## Create list structure required to run the featureAlignedHeatmap function
  plotData <- list()
  com <- paste0("plotData$", plotTitle, " <- insertionMatrix")
  eval(parse(text = com))
  
  ## Make the plot
  ChIPpeakAnno::featureAlignedHeatmap(
    plotData,
    feature.gr = reCenterPeaks(bindingSites, width = numBP),
    upper.extreme = 1,
    annoMcols = "score",
    sortBy = "score",
    n.tile = numBP,
    margin = c(0.1, 0.005, 0.05, 0.2),
    color = colorRampPalette(c("white", "red"), bias = 0.5)(100),
    gp = gpar(fontsize = 10),
    newpage = TRUE)
  grid.export(svgOutPath)
  dev.off()
}

BiocManager::install("ChIPpeakAnno")
library(ChIPpeakAnno)

out1 <- "C:\\Users\\jsk33\\OneDrive\\R\\ATACseqAnalysis\\LNCaP\\plots\\heat1.svg"
plotMotifAlignedHeatmap(ins2, extSites[idx], "test", out1)

```

Make matrix and annotate
```{r}

## IMPORTANT
## EXTRACT THE CHR VALUE FROM A GRANGES OBJECT
## as.character(seqnames(AR.bindingSites[["bindingSites"]]))

numSites <- FPdata.LNCaP.AR$WT01[["numSites"]]

## Output object for Alessandro
FPdataMatrix.LNCaP.AR <- list()

FPdataMatrix.LNCaP.AR$Gene <- AR.bindingSites[["Gene"]]
FPdataMatrix.LNCaP.AR$PWM <- AR.bindingSites[["PWM"]]
FPdataMatrix.LNCaP.AR$PWMmatchThreshold <- AR.bindingSites[["score"]]
FPdataMatrix.LNCaP.AR$numSites <- AR.bindingSites[["numSites"]]
FPdataMatrix.LNCaP.AR$minPWMscore <- AR.bindingSites[["minPWMscore"]]
FPdataMatrix.LNCaP.AR$maxPWMscore <- AR.bindingSites[["maxPWMscore"]]

## Annotations
AR.bindingSite.Annotations <- data.frame("siteIndex" = 1:numSites)
AR.bindingSite.Annotations$PWMmatchScore <- AR.bindingSites[["bindingSites"]]@elementMetadata@listData[["score"]]
AR.bindingSite.Annotations$chromosome <- as.character(seqnames(AR.bindingSites[["bindingSites"]]))
AR.bindingSite.Annotations$start <- AR.bindingSites[["bindingSites"]]@ranges@start
AR.bindingSite.Annotations$locus <- paste0(AR.bindingSite.Annotations$chromosome, ":", AR.bindingSite.Annotations$start)

##
#if (!requireNamespace("BiocManager", quietly = TRUE))
#	 install.packages("BiocManager")
#BiocManager::install("ChIPseeker")

### Gene annotations
library(ChIPseeker)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(org.Hs.eg.db)

##
annotated_granges <- annotatePeak(
             AR.bindingSites[["bindingSites"]],
             tssRegion = c(-3000, 3000),
             TxDb = TxDb.Hsapiens.UCSC.hg38.knownGene,
             level = "transcript",
             assignGenomicAnnotation = TRUE,
             genomicAnnotationPriority = c("Promoter", "5UTR", "3UTR", "Exon", "Intron", "Downstream", "Intergenic"),
             annoDb =NULL,
             addFlankGeneInfo = FALSE,
             flankDistance = 5000,
             sameStrand = FALSE,
             ignoreOverlap = FALSE,
             ignoreUpstream = FALSE,
             ignoreDownstream = FALSE,
             overlap = "TSS", verbose = TRUE)

##
source("C:\\Users\\Jordan\\OneDrive\\Documents\\Github\\vaxtools\\R\\utils.R")
annotated_granges@anno$symbol <- getGeneSymbolsFromEntrezId(annotated_granges@anno$geneId)

##
AR.bindingSite.Annotations$closestGene <- annotated_granges@anno@elementMetadata@listData[["symbol"]]

FPdataMatrix.LNCaP.AR$bindingSiteAnnotations <- AR.bindingSite.Annotations

##### MATRIX
AR.bindingMatrix <- matrix(data = 0, nrow = 8, ncol = numSites)
rownames(AR.bindingMatrix) <- c("WT01", "WT02", "CR01", "CR02", "CR04", "CR05", "CR07", "CR08")

##
AR.bindingMatrix[1, FPdata.LNCaP.AR[["WT01"]][["idxBFpass"]]] <- 1
AR.bindingMatrix[2, FPdata.LNCaP.AR[["WT02"]][["idxBFpass"]]] <- 1
AR.bindingMatrix[3, FPdata.LNCaP.AR[["CR01"]][["idxBFpass"]]] <- 1
AR.bindingMatrix[4, FPdata.LNCaP.AR[["CR02"]][["idxBFpass"]]] <- 1
AR.bindingMatrix[5, FPdata.LNCaP.AR[["CR04"]][["idxBFpass"]]] <- 1
AR.bindingMatrix[6, FPdata.LNCaP.AR[["CR05"]][["idxBFpass"]]] <- 1
AR.bindingMatrix[7, FPdata.LNCaP.AR[["CR07"]][["idxBFpass"]]] <- 1
AR.bindingMatrix[8, FPdata.LNCaP.AR[["CR08"]][["idxBFpass"]]] <- 1

##
FPdataMatrix.LNCaP.AR$bindingMatrix <- AR.bindingMatrix

#### Pvalue
AR.pvalueMatrix <- matrix(data = 0, nrow = 8, ncol = numSites)
rownames(AR.pvalueMatrix) <- c("WT01", "WT02", "CR01", "CR02", "CR04", "CR05", "CR07", "CR08")

##
AR.pvalueMatrix[1,] <- FPdata.LNCaP.AR[["WT01"]][["pvalue"]]
AR.pvalueMatrix[2,] <- FPdata.LNCaP.AR[["WT02"]][["pvalue"]]
AR.pvalueMatrix[3,] <- FPdata.LNCaP.AR[["CR01"]][["pvalue"]]
AR.pvalueMatrix[4,] <- FPdata.LNCaP.AR[["CR02"]][["pvalue"]]
AR.pvalueMatrix[5,] <- FPdata.LNCaP.AR[["CR04"]][["pvalue"]]
AR.pvalueMatrix[6,] <- FPdata.LNCaP.AR[["CR05"]][["pvalue"]]
AR.pvalueMatrix[7,] <- FPdata.LNCaP.AR[["CR07"]][["pvalue"]]
AR.pvalueMatrix[8,] <- FPdata.LNCaP.AR[["CR08"]][["pvalue"]]

##
FPdataMatrix.LNCaP.AR$pvalueMatrix <- AR.pvalueMatrix


outPath <- "C:\\Users\\Jordan\\OneDrive\\FPanalysis\\FPdataMatrix_LNCaP_AR.RDS"
saveRDS(FPdataMatrix.LNCaP.AR, file = outPath)

```



Plots
```{r}



AR.matrix <- FPdata.AR[["bindingMatrix"]]
x <- AR.matrix[,1:100]
pheatmap(x)

## Subset for only locations where both WT are the same
y <- as.data.frame(AR.matrix)
z <- dplyr::filter(y, WT01 == 1, WT02 == 2)

aa <- which(y[1,] == 1 & y[2,] == 1)
ab <- which(y[1,] == 0 & y[2,] == 0)
ac <- c(aa, ab)
new <- y[1:8,ac]

new2 <- new[,24728:24828]
pheatmap(new2,
         cluster_rows = FALSE,
         cluster_cols = FALSE)


EZH2.matrix <- FPdata.EZH2[["bindingMatrix"]]
x <- EZH2.matrix[,1:100]
pheatmap(x)

FOXM1.matrix <- FPdata.FOXM1[["bindingMatrix"]]
x <- FOXM1.matrix[,1:100]
pheatmap(x)

MYC.matrix <- FPdata.MYC[["bindingMatrix"]]
x <- MYC.matrix[,1:100]
pheatmap(x)

MYCN.matrix <- FPdata.MYCN[["bindingMatrix"]]
x <- MYCN.matrix[,1:100]
pheatmap(x)

SOX2.matrix <- FPdata.SOX2[["bindingMatrix"]]
x <- SOX2.matrix[,1:100]
pheatmap(x)


```

Diff peaks overlap
```{r}

ctrl.wt01.treat.cr01.peaksPath <- "D:\\diffpeaks\\ctrl-wt01.treat-cr01_peaks.narrowPeak"
ctrl.wt01.treat.cr02.peaksPath <- "D:\\diffpeaks\\ctrl-wt01.treat-cr02_peaks.narrowPeak"
ctrl.wt01.treat.cr04.peaksPath <- "D:\\diffpeaks\\ctrl-wt01.treat-cr04_peaks.narrowPeak"
ctrl.wt01.treat.cr05.peaksPath <- "D:\\diffpeaks\\ctrl-wt01.treat-cr05_peaks.narrowPeak"
ctrl.wt01.treat.cr07.peaksPath <- "D:\\diffpeaks\\ctrl-wt01.treat-cr07_peaks.narrowPeak"
ctrl.wt01.treat.cr08.peaksPath <- "D:\\diffpeaks\\ctrl-wt01.treat-cr08_peaks.narrowPeak"

ctrl.wt02.treat.cr01.peaksPath <- "D:\\diffpeaks\\ctrl-wt02.treat-cr01_peaks.narrowPeak"
ctrl.wt02.treat.cr02.peaksPath <- "D:\\diffpeaks\\ctrl-wt02.treat-cr01_peaks.narrowPeak"
ctrl.wt02.treat.cr04.peaksPath <- "D:\\diffpeaks\\ctrl-wt02.treat-cr01_peaks.narrowPeak"
ctrl.wt02.treat.cr05.peaksPath <- "D:\\diffpeaks\\ctrl-wt02.treat-cr01_peaks.narrowPeak"
ctrl.wt02.treat.cr07.peaksPath <- "D:\\diffpeaks\\ctrl-wt02.treat-cr01_peaks.narrowPeak"
ctrl.wt02.treat.cr08.peaksPath <- "D:\\diffpeaks\\ctrl-wt02.treat-cr01_peaks.narrowPeak"

ctrl.wt01.treat.cr01.peaks <- importBED(ctrl.wt01.treat.cr01.peaksPath)
ctrl.wt01.treat.cr02.peaks <- importBED(ctrl.wt01.treat.cr02.peaksPath)
ctrl.wt01.treat.cr04.peaks <- importBED(ctrl.wt01.treat.cr04.peaksPath)
ctrl.wt01.treat.cr05.peaks <- importBED(ctrl.wt01.treat.cr05.peaksPath)
ctrl.wt01.treat.cr07.peaks <- importBED(ctrl.wt01.treat.cr07.peaksPath)
ctrl.wt01.treat.cr08.peaks <- importBED(ctrl.wt01.treat.cr08.peaksPath)

ctrl.wt02.treat.cr01.peaks <- importBED(ctrl.wt02.treat.cr01.peaksPath)
ctrl.wt02.treat.cr02.peaks <- importBED(ctrl.wt02.treat.cr02.peaksPath)
ctrl.wt02.treat.cr04.peaks <- importBED(ctrl.wt02.treat.cr04.peaksPath)
ctrl.wt02.treat.cr05.peaks <- importBED(ctrl.wt02.treat.cr05.peaksPath)
ctrl.wt02.treat.cr07.peaks <- importBED(ctrl.wt02.treat.cr07.peaksPath)
ctrl.wt02.treat.cr08.peaks <- importBED(ctrl.wt02.treat.cr08.peaksPath)

cr01.wt.overlap.peaks <- subsetByOverlaps(ctrl.wt01.treat.cr01.peaks, ctrl.wt02.treat.cr01.peaks)
cr02.wt.overlap.peaks <- subsetByOverlaps(ctrl.wt01.treat.cr02.peaks, ctrl.wt02.treat.cr02.peaks)
cr04.wt.overlap.peaks <- subsetByOverlaps(ctrl.wt01.treat.cr04.peaks, ctrl.wt02.treat.cr04.peaks)
cr05.wt.overlap.peaks <- subsetByOverlaps(ctrl.wt01.treat.cr05.peaks, ctrl.wt02.treat.cr05.peaks)
cr07.wt.overlap.peaks <- subsetByOverlaps(ctrl.wt01.treat.cr07.peaks, ctrl.wt02.treat.cr07.peaks)
cr08.wt.overlap.peaks <- subsetByOverlaps(ctrl.wt01.treat.cr08.peaks, ctrl.wt02.treat.cr08.peaks)

cr01.cr02.overlap.peaks <- subsetByOverlaps(cr01.wt.overlap.peaks, cr02.wt.overlap.peaks)
cr04.cr05.overlap.peaks <- subsetByOverlaps(cr04.wt.overlap.peaks, cr05.wt.overlap.peaks)
cr07.cr08.overlap.peaks <- subsetByOverlaps(cr07.wt.overlap.peaks, cr08.wt.overlap.peaks)

all.clones.overlap.peaks <- c(cr01.cr02.overlap.peaks, cr04.cr05.overlap.peaks, cr07.cr08.overlap.peaks)

wt01idx <- getGRoverlapIndex(WT01.peaks, all.clones.overlap.peaks)
wt02idx <- getGRoverlapIndex(WT02.peaks, all.clones.overlap.peaks)
cr01idx <- getGRoverlapIndex(cr01.wt.overlap.peaks, all.clones.overlap.peaks)
cr02idx <- getGRoverlapIndex(cr02.wt.overlap.peaks, all.clones.overlap.peaks)
cr04idx <- getGRoverlapIndex(cr04.wt.overlap.peaks, all.clones.overlap.peaks)
cr05idx <- getGRoverlapIndex(cr05.wt.overlap.peaks, all.clones.overlap.peaks)
cr07idx <- getGRoverlapIndex(cr07.wt.overlap.peaks, all.clones.overlap.peaks)
cr08idx <- getGRoverlapIndex(cr08.wt.overlap.peaks, all.clones.overlap.peaks)

newmatrix <- matrix(data = 0, nrow = 8, ncol = 173306)
rownames(newmatrix) <- c("WT01", "WT02", "CR01", "CR02", "CR04", "CR05", "CR07", "CR08")

newmatrix[1,wt01idx] <- 1
newmatrix[2,wt02idx] <- 1
newmatrix[3,cr01idx] <- 1
newmatrix[4,cr02idx] <- 1
newmatrix[5,cr04idx] <- 1
newmatrix[6,cr05idx] <- 1
newmatrix[7,cr07idx] <- 1
newmatrix[8,cr08idx] <- 1


pheatmap(newmatrix[,5000:5100],
         cluster_rows = FALSE,
         cluster_cols = FALSE)

```

Peak overlaps
```{r}

# WT01.peaksPath <- "C:\\Users\\jsk33\\Desktop\\peaks\\LNCaP-WT-01-REP1_globalnorm_peaks.narrowPeak"
# WT02.peaksPath <- "C:\\Users\\jsk33\\Desktop\\peaks\\LNCaP-WT-02-REP1_globalnorm_peaks.narrowPeak"
# CR01.peaksPath <- "C:\\Users\\jsk33\\Desktop\\peaks\\LNCaP-CR-01-REP1_globalnorm_peaks.narrowPeak"
# CR02.peaksPath <- "C:\\Users\\jsk33\\Desktop\\peaks\\LNCaP-CR-02-REP1_globalnorm_peaks.narrowPeak"
# CR04.peaksPath <- "C:\\Users\\jsk33\\Desktop\\peaks\\LNCaP-CR-04-REP1_globalnorm_peaks.narrowPeak"
# CR05.peaksPath <- "C:\\Users\\jsk33\\Desktop\\peaks\\LNCaP-CR-05-REP1_globalnorm_peaks.narrowPeak"
# CR07.peaksPath <- "C:\\Users\\jsk33\\Desktop\\peaks\\LNCaP-CR-07-REP1_globalnorm_peaks.narrowPeak"
# CR08.peaksPath <- "C:\\Users\\jsk33\\Desktop\\peaks\\LNCaP-CR-08-REP1_globalnorm_peaks.narrowPeak"

WT01.peaksPath <- "D:\\peaks\\LNCaP-WT-01-REP1_globalnorm_peaks.narrowPeak"
WT02.peaksPath <- "D:\\peaks\\LNCaP-WT-02-REP1_globalnorm_peaks.narrowPeak"
CR01.peaksPath <- "D:\\peaks\\LNCaP-CR-01-REP1_globalnorm_peaks.narrowPeak"
CR02.peaksPath <- "D:\\peaks\\LNCaP-CR-02-REP1_globalnorm_peaks.narrowPeak"
CR04.peaksPath <- "D:\\peaks\\LNCaP-CR-04-REP1_globalnorm_peaks.narrowPeak"
CR05.peaksPath <- "D:\\peaks\\LNCaP-CR-05-REP1_globalnorm_peaks.narrowPeak"
CR07.peaksPath <- "D:\\peaks\\LNCaP-CR-07-REP1_globalnorm_peaks.narrowPeak"
CR08.peaksPath <- "D:\\peaks\\LNCaP-CR-08-REP1_globalnorm_peaks.narrowPeak"

WT01.peaks <- importBED(WT01.peaksPath)
WT02.peaks <- importBED(WT02.peaksPath)
CR01.peaks <- importBED(CR01.peaksPath)
CR02.peaks <- importBED(CR02.peaksPath)
CR04.peaks <- importBED(CR04.peaksPath)
CR05.peaks <- importBED(CR05.peaksPath)
CR07.peaks <- importBED(CR07.peaksPath)
CR08.peaks <- importBED(CR08.peaksPath)

FPdata.AR.WT01 <- FPdata.AR[["WT01"]]
FPdata.AR.WT01.inputSites <- FPdata.AR.WT01[["inputSites"]]
WT01.AR.peak.overlap.index <- findOverlaps(FPdata.AR.WT01.inputSites, WT01.peaks)
AR.WT01.idx <- unique(WT01.AR.peak.overlap.index@from)

FPdata.AR.WT02 <- FPdata.AR[["WT02"]]
FPdata.AR.WT02.inputSites <- FPdata.AR.WT02[["inputSites"]]
WT02.AR.peak.overlap.index <- findOverlaps(FPdata.AR.WT02.inputSites, WT02.peaks)
AR.WT02.idx <- unique(WT02.AR.peak.overlap.index@from)

FPdata.AR.CR01 <- FPdata.AR[["CR01"]]
FPdata.AR.CR01.inputSites <- FPdata.AR.CR01[["inputSites"]]
CR01.AR.peak.overlap.index <- findOverlaps(FPdata.AR.CR01.inputSites, CR01.peaks)
AR.CR01.idx <- unique(CR01.AR.peak.overlap.index@from)

FPdata.AR.CR02 <- FPdata.AR[["CR02"]]
FPdata.AR.CR02.inputSites <- FPdata.AR.CR02[["inputSites"]]
CR02.AR.peak.overlap.index <- findOverlaps(FPdata.AR.CR02.inputSites, CR02.peaks)
AR.CR02.idx <- unique(CR02.AR.peak.overlap.index@from)

FPdata.AR.CR04 <- FPdata.AR[["CR04"]]
FPdata.AR.CR04.inputSites <- FPdata.AR.CR04[["inputSites"]]
CR04.AR.peak.overlap.index <- findOverlaps(FPdata.AR.CR04.inputSites, CR04.peaks)
AR.CR04.idx <- unique(CR04.AR.peak.overlap.index@from)

FPdata.AR.CR05 <- FPdata.AR[["CR05"]]
FPdata.AR.CR05.inputSites <- FPdata.AR.CR05[["inputSites"]]
CR05.AR.peak.overlap.index <- findOverlaps(FPdata.AR.CR05.inputSites, CR05.peaks)
AR.CR05.idx <- unique(CR05.AR.peak.overlap.index@from)

FPdata.AR.CR07 <- FPdata.AR[["CR07"]]
FPdata.AR.CR07.inputSites <- FPdata.AR.CR07[["inputSites"]]
CR07.AR.peak.overlap.index <- findOverlaps(FPdata.AR.CR07.inputSites, CR07.peaks)
AR.CR07.idx <- unique(CR07.AR.peak.overlap.index@from)

FPdata.AR.CR08 <- FPdata.AR[["CR08"]]
FPdata.AR.CR08.inputSites <- FPdata.AR.CR08[["inputSites"]]
CR08.AR.peak.overlap.index <- findOverlaps(FPdata.AR.CR08.inputSites, CR08.peaks)
AR.CR08.idx <- unique(CR08.AR.peak.overlap.index@from)

newmatrix <- matrix(data = 0, nrow = 8, ncol = 62585)
rownames(newmatrix) <- c("WT01", "WT02", "CR01", "CR02", "CR04", "CR05", "CR07", "CR08")
newmatrix[1,AR.WT01.idx] <- 1
newmatrix[2,AR.WT02.idx] <- 1
newmatrix[3,AR.CR01.idx] <- 1
newmatrix[4,AR.CR02.idx] <- 1
newmatrix[5,AR.CR04.idx] <- 1
newmatrix[6,AR.CR05.idx] <- 1
newmatrix[7,AR.CR07.idx] <- 1
newmatrix[8,AR.CR08.idx] <- 1

pheatmap(newmatrix[,1:100],
         cluster_rows = FALSE,
         cluster_cols = FALSE)

AR.matrix <- FPdata.AR[["bindingMatrix"]]
x <- AR.matrix[,AR.WT01.idx]
pheatmap(x[,100:200])

y <- as.data.frame(AR.matrix)
z <- dplyr::filter(y, WT01 == 1, WT02 == 2)

aa <- which(y[1,] == 1 & y[2,] == 1)
ab <- which(y[1,] == 0 & y[2,] == 0)
ac <- c(aa, ab)
new <- y[1:8,ac]

new2 <- new[,24728:24828]
pheatmap(new2,
         cluster_rows = FALSE,
         cluster_cols = FALSE)

```


```{r}

library(chromoMap)
library(Repitools)

chrom <- "C:\\Users\\Jordan\\Desktop\\xsomes.txt"

ctrl.wt01.treat.cr02.peaksPath <- "D:\\diffpeaks\\ctrl-wt01.treat-cr02_peaks.narrowPeak"
ctrl.wt01.treat.cr02.peaks <- readBed(ctrl.wt01.treat.cr02.peaksPath, track.line = FALSE, remove.unusual = FALSE, zero.based = TRUE)
ctrl.wt01.treat.cr02.peaks <- keepStandardChromosomes(ctrl.wt01.treat.cr02.peaks, pruning.mode="coarse")
ctrl.wt01.treat.cr02.peaks <- trim(ctrl.wt01.treat.cr02.peaks)

xx <- Repitools::annoGR2DF(ctrl.wt01.treat.cr02.peaks)

xy <- data.frame(name = 1:100,
                 chr = xx[1:100,1],
                 start = xx[1:100,2],
                 end = xx[1:100,3],
                 data = xx[1:100, 5]/100)


write.table(xy,
            file = "C:\\Users\\Jordan\\Desktop\\test.txt",
            quote = FALSE,
            col.names = FALSE,
            row.names = FALSE,
            append = FALSE,
            sep = "\t"
            )
  


chromoMap("C:\\Users\\Jordan\\Desktop\\xsomes.txt",
          "C:\\Users\\Jordan\\Desktop\\test.txt",
          data_based_color_map = T,
          data_type = numeric
          )


```


```{r}
###################################################
### code chunk number 1: ChromHeatMap.Rnw:57-64
###################################################
library('ALL')
data('ALL')
selSamples <- ALL$mol.biol %in% c('ALL1/AF4', 'E2A/PBX1')
ALLs <- ALL[, selSamples]

library('ChromHeatMap')
chrdata<-makeChrStrandData(exprs(ALLs), lib='hgu95av2')


###################################################
### code chunk number 2: ChromHeatMap.Rnw:87-89
###################################################
groupcol <- ifelse( ALLs$mol.biol == 'ALL1/AF4', 'red', 'green' )
plotChrMap(chrdata, 19, strands='both', RowSideColors=groupcol)


###################################################
### code chunk number 3: ChromHeatMap.Rnw:98-99
###################################################
plotmap<-plotChrMap(chrdata, 1, cytoband='q23', interval=50000, srtCyto=0, cexCyto=1.2)


###################################################
### code chunk number 4: ChromHeatMap.Rnw:135-137 (eval = FALSE)
###################################################
## probes <- grabChrMapProbes( plotmap )
## genes <- unlist(mget(probes, envir=hgu95av2SYMBOL, ifnotfound=NA))


###################################################
### code chunk number 5: ChromHeatMap.Rnw:156-157
###################################################
sessionInfo()


```


---
output: html_document
editor_options: 
  chunk_output_type: console
---

SETUP
```{r}
#### Disable scientific notation in variables
options(scipen = 999)
options(warn = -1)
hg38TotalBP <- 3272116950

suppressMessages(library(GenomicRanges))
suppressMessages(library(Rsamtools))
suppressMessages(library(GenomicAlignments))
suppressMessages(library(stats4))
suppressMessages(library(BiocGenerics))
suppressMessages(library(parallel))
suppressMessages(library(genomation))
suppressMessages(library(seqLogo))
suppressMessages(library(ChIPpeakAnno))
suppressMessages(library(rlist))
suppressMessages(library(TxDb.Hsapiens.UCSC.hg38.knownGene))
suppressMessages(library(pheatmap))
suppressMessages(library(org.Hs.eg.db))

generateNullFP <- function(iterations, inputSignal, analysisWidth, motifWidth){
  # This script will be used to generate indiviudal null models at predicted motif binding sites across the genome when scanning for TF footprinting from ATAC-seq data. To generate these null models, the current model will need to:
  # Consider the total signal (number of insertions) at each specific ~200 bp locus
  # Use the actul underlying reference sequence of that ~200 bp stretch from the hg38 reference genome
  # Use published or experimentally derived models of Tn5 sequence specific insertion bias
  # For each locus, build a probablistic model of insertion site distributions based on the underlying sequence and Tn5 insertion bias
  # Generate the null model graph by weighted random residstribution of the total observed signal at that site
  # Importantly, the null model must be generated separately for the plus and minus strand, it can then be combined and compared to the combined signal from the reference observed signal at that sequence
  # These null models can then be used for a site-by-site comparison of the null model against the observed data to accept or reject the null hypothesis
  # iterations = number of iterations
  # inputSignals = unique values for total signal
  # analysisWidth = total bp in region of interest (flank + background + motif)
  # motifWidth = motif width
  
  # declare vector of size n to store average motif signal values
  averages <- c()
  
  # generate the null models and calculate motif averages
  for (a in 1:iterations){
    
    # declare the null vector
    null <- c(1:(analysisWidth))
    
    # randomly distribute the total signal
    # size = the number of values to distribute
    # prob = probability of each site
    # length = length of the generated vector
    null <- c(as.vector(rmultinom(1, size=inputSignal, prob=rep(1, length(null)))))
    
    ## Calculate the mean signal in motif region
    motifStart <- ((analysisWidth - motifWidth)/2)
    motifEnd <- (motifStart + motifWidth)
    motifAvg <- (sum(null[motifStart:motifEnd])) / motifWidth
    
    ## Store the average values
    averages[a] <- motifAvg
    
  } # end for (a in 1:n)
  return(averages)
} # end generateNullFP function

analyzeFP <- function(geneName, geneSitesPath, sampleName, sampleBamPath){
  
  cat("Input variables:", "\n", geneName, "\n", geneSitesPath, "\n", sampleName, "\n", sampleBamPath, "\n")
  cat("Reading sites RDS", "\n")
  sites <- readRDS(geneSitesPath)
  inputSites <- sites[[1]][["sites"]]
  inputSites@elementType <- "GENE"
  inputSites <- keepStandardChromosomes(inputSites, pruning.mode="coarse")
  inputSites <- trim(inputSites)
  cat("Checking number of sites and motif width", "\n")
  motifWidth <- length(sites[[1]][["PWM"]][1,])
  numSites <- length(inputSites)
  cat("Extending sites", "\n")
  extSites <- promoters(inputSites, upstream = 250, downstream = (250 + motifWidth), use.names=TRUE)
  extSites <- keepStandardChromosomes(extSites, pruning.mode="coarse")
  extSites <- trim(extSites)
  cat("Loading reads from bam file", "\n")
  param <- ScanBamParam(which = extSites)
  bamIn <- readGAlignments(sampleBamPath, param = param)
  grIn <- granges(bamIn)
  grIn <- keepStandardChromosomes(grIn, pruning.mode="coarse")
  grIn <- trim(grIn)
  cat("Resizing reads and shifting +4/-5", "\n")
  grIn2 <- resize(grIn, width = 1)
  plusIdx <- which(strand(grIn2) == "+")
  minusIdx <- which(strand(grIn2) == "-")
  grPlus <- grIn2[plusIdx]
  grMinus <- grIn2[minusIdx]
  grPlusShifted <- shift(grPlus, shift=4L)
  grMinusShifted <- shift(grMinus, shift=-5L)
  grMerged <- c(grPlusShifted, grMinusShifted)
  shiftedInsertions <- grMerged
  cat("Generating insertion matrix", "\n")
  insRLE <- coverage(grMerged)
  insViews <- Views(insRLE, extSites)
  insMatrix <- as.matrix(insViews)
  cat("Calculating normalization factors", "\n")
  libSize <- length(bamIn)
  coverageSize <- sum(as.numeric(width(reduce(grIn, ignore.strand=TRUE))))
  libFactor <- libSize / coverageSize
  cat("Calculating raw sites basic footprint stats", "\n")
  rawSiteBasicStats <- matrix(data = NA, nrow = numSites, ncol = 10)
  colnames(rawSiteBasicStats) <- c("Site index", "Total signal", "Total signal per bp", "Motif signal per bp",
                                 "Flank signal per bp", "Background signal per bp", "Wide flank signal per bp",
                                 "Flank / Background", "Motif / Flank", "Motif / Wide Flank")
  for (b in 1:numSites){
    rawSiteBasicStats[b,1] <- b # Site index
    rawSiteBasicStats[b,2] <- sum(insMatrix[b,]) # Total signal
    rawSiteBasicStats[b,3] <- rawSiteBasicStats[b,2] / (500 + motifWidth) # Total signal per bp
    rawSiteBasicStats[b,4] <- sum(insMatrix[b,(250:(250 + motifWidth))] / motifWidth) # Motif signal per bp
    rawSiteBasicStats[b,5] <- (sum(insMatrix[b,200:250]) + sum(insMatrix[b,(250 + motifWidth):(300 + motifWidth)])) / 100 # Flank signal per bp
    rawSiteBasicStats[b,6] <- (sum(insMatrix[b,1:50]) + sum(insMatrix[b,(500 + motifWidth-50):(500 + motifWidth)])) / 100 # Background signal per bp
    rawSiteBasicStats[b,7] <- (sum(insMatrix[b,1:250]) + sum(insMatrix[b,(250 + motifWidth):(500 + motifWidth)])) / 500 # Wide flank signal per bp
    rawSiteBasicStats[b,8] <- rawSiteBasicStats[b,5] / rawSiteBasicStats[b,6] # Flank / background
    rawSiteBasicStats[b,9] <- rawSiteBasicStats[b,4] / rawSiteBasicStats[b,5] # Motif / flank
    rawSiteBasicStats[b,10] <- rawSiteBasicStats[b,4] / rawSiteBasicStats[b,7] # Motif / wide flank
  } # end for (b in 1:numSites)
  cat("Analyzing insertion probability", "\n")
  rawInsProb <- c()
  for (c in 1:(500 + motifWidth)){
    rawInsProb[c] <- sum(insMatrix[,c])} # end for (c in 1:(500 + motifWidth))
  cat("Finding raw total signal", "\n")
  rawTotalSignal<- sum(rawInsProb)
  rawInsProb <- rawInsProb / rawTotalSignal
  uniqueTotalSignals <- unique(rawSiteBasicStats[,2])
  siteWidth <- 500 + motifWidth
  cat("Generating null models", "\n")
  nullModels <- matrix(data = NA, ncol = 2, nrow = length(uniqueTotalSignals))
  colnames(nullModels) <- c("Input signal", "Avg motif signal in null model")
  for (c in 1:length(uniqueTotalSignals)){
    nullVec <- generateNullFP(1000, uniqueTotalSignals[c], siteWidth, motifWidth)
    nullModels[c,1] <- uniqueTotalSignals[c]
    nullModels[c,2] <- mean(nullVec)
  } # end for (c in 1:length(uniqueTotalSignals))
  cat("Performing hypothesis testing", "\n")
  ttest <- list()
  pvalue <- c()
  tvalue <- c()
  for (d in 1:numSites){
    currentSignal <- c(rawSiteBasicStats[d,2])
    currentNullModel <- nullModels[which(nullModels[,1]==currentSignal),2]
    ttest[[d]] <- t.test(insMatrix[d,250:(250+motifWidth)], mu=currentNullModel, alternative="less", conf.level = 0.95)
    pvalue[d] <- ttest[[d]][["p.value"]]
    tvalue[d] <- ttest[[d]][["statistic"]][["t"]]} # end for (d in 1:numSites)
  cat("Performing test corrections", "\n")
  idxPvaluePass <- which(pvalue < 0.05)
  pvaluePass <- pvalue[idxPvaluePass]
  ppassNumSites <- length(idxPvaluePass)
  idxBFpass <- which(pvalue < (0.05 / numSites))
  BFpvaluePass <- pvalue[idxBFpass]
  BFpassNumSites <- length(idxBFpass)
  BHpvalue <- p.adjust(pvalue, method = "BH")
  idxBHpass <- which(BHpvalue < 0.05)
  BHpvaluePass <- pvalue[idxBHpass]
  BHpassNumSites <- length(idxBHpass)
  cat("Transferring data to storage list", "\n")
  tempList <- list()
  tempList$geneName <- geneName
  tempList$geneSitesPath <- geneSitesPath
  tempList$sampleName <- sampleName
  tempList$sampleBamPath <- sampleBamPath
  tempList$numSites <- numSites
  tempList$motifWidth <- motifWidth
  tempList$PWM <- sites[[1]][["PWM"]]
  tempList$inputSites <- inputSites
  tempList$extSites <- extSites
  tempList$insMatrix <- insMatrix
  tempList$libSize <- libSize
  tempList$coverageSize <- coverageSize
  tempList$libFactor <- libFactor
  tempList$rawSiteBasicStats <- rawSiteBasicStats
  tempList$rawInsProb <- rawInsProb
  tempList$ttest <- ttest
  tempList$pvalue <- pvalue
  tempList$tvalue <- tvalue
  tempList$idxPvaluePass <- idxPvaluePass
  tempList$pvaluePass <- pvaluePass
  tempList$ppassNumSites <- ppassNumSites
  tempList$idxBFpass <- idxBFpass
  tempList$BFpvaluePass <- BFpvaluePass
  tempList$BFpassNumSites <- BFpassNumSites
  tempList$BHpvalue <- BHpvalue
  tempList$idxBHpass <- idxBHpass
  tempList$BHpvaluePass <- BHpvaluePass
  tempList$BHpassNumSites <- BHpassNumSites
  cat("Returning data", "\n")
  return(tempList)
}

importBED <- function(bedPath){
  bedin <- readBed(bedPath, track.line = FALSE, remove.unusual = FALSE, zero.based = TRUE)
  bedin <- keepStandardChromosomes(bedin, pruning.mode="coarse")
  bedin <- trim(bedin)
  return(bedin)
}

getGRoverlapIndex <- function(input1, input2){
  overlap <- findOverlaps(input1, input2)
  index <- unique(overlap@from)
  return(index)
}


#### INPUT FILES ##################################################################################################
bam.CR01 <- "C:\\Users\\jsk33\\Desktop\\lncap\\LNCaP-CR-01-REP1.u.bam"
bai.CR01 <- "C:\\Users\\jsk33\\Desktop\\lncap\\LNCaP-CR-01-REP1.u.bai"
bam.CR02 <- "C:\\Users\\jsk33\\Desktop\\lncap\\LNCaP-CR-02-REP1.u.bam"
bai.CR02 <- "C:\\Users\\jsk33\\Desktop\\lncap\\LNCaP-CR-02-REP1.u.bai"
bam.CR04 <- "C:\\Users\\jsk33\\Desktop\\lncap\\LNCaP-CR-04-REP1.u.bam"
bai.CR04 <- "C:\\Users\\jsk33\\Desktop\\lncap\\LNCaP-CR-04-REP1.u.bai"
bam.CR05 <- "C:\\Users\\jsk33\\Desktop\\lncap\\LNCaP-CR-05-REP1.u.bam"
bai.CR05 <- "C:\\Users\\jsk33\\Desktop\\lncap\\LNCaP-CR-05-REP1.u.bai"
bam.CR07 <- "C:\\Users\\jsk33\\Desktop\\lncap\\LNCaP-CR-07-REP1.u.bam"
bai.CR07 <- "C:\\Users\\jsk33\\Desktop\\lncap\\LNCaP-CR-07-REP1.u.bai"
bam.CR08 <- "C:\\Users\\jsk33\\Desktop\\lncap\\LNCaP-CR-08-REP1.u.bam"
bai.CR08 <- "C:\\Users\\jsk33\\Desktop\\lncap\\LNCaP-CR-08-REP1.u.bai"
bam.WT01 <- "C:\\Users\\jsk33\\Desktop\\lncap\\LNCaP-WT-01-REP1.u.bam"
bai.WT01 <- "C:\\Users\\jsk33\\Desktop\\lncap\\LNCaP-WT-01-REP1.u.bai"
bam.WT02 <- "C:\\Users\\jsk33\\Desktop\\lncap\\LNCaP-WT-02-REP1.u.bam"
bai.WT02 <- "C:\\Users\\jsk33\\Desktop\\lncap\\LNCaP-WT-02-REP1.u.bai"
##
AR.sites.path <- "C:\\Users\\jsk33\\Documents\\git\\miscLabScripts\\analysisWithAlessandroLNCaP\\AR.bindingSites.RDS"
EZH2.sites.path <- "C:\\Users\\jsk33\\Documents\\git\\miscLabScripts\\analysisWithAlessandroLNCaP\\EZH2.bindingSites.RDS"
FOXM1.sites.path <- "C:\\Users\\jsk33\\Documents\\git\\miscLabScripts\\analysisWithAlessandroLNCaP\\FOXM1.bindingSites.RDS"
MYC.sites.path <- "C:\\Users\\jsk33\\Documents\\git\\miscLabScripts\\analysisWithAlessandroLNCaP\\MYC.bindingSites.RDS"
MYCN.sites.path <- "C:\\Users\\jsk33\\Documents\\git\\miscLabScripts\\analysisWithAlessandroLNCaP\\MYCN.bindingSites.RDS"
SOX2.sites.path <- "C:\\Users\\jsk33\\Documents\\git\\miscLabScripts\\analysisWithAlessandroLNCaP\\SOX2.bindingSites.RDS"
#### INPUT FILES ##################################################################################################
```

AR
```{r}

FPdata.AR <- list()
FPdata.AR$WT01 <- analyzeFP("AR", AR.sites.path, "WT01", bam.WT01)
FPdata.AR$WT02 <- analyzeFP("AR", AR.sites.path, "WT02", bam.WT02)
FPdata.AR$CR01 <- analyzeFP("AR", AR.sites.path, "CR01", bam.CR01)
FPdata.AR$CR02 <- analyzeFP("AR", AR.sites.path, "CR02", bam.CR01)
FPdata.AR$CR04 <- analyzeFP("AR", AR.sites.path, "CR04", bam.CR01)
FPdata.AR$CR05 <- analyzeFP("AR", AR.sites.path, "CR05", bam.CR01)
FPdata.AR$CR07 <- analyzeFP("AR", AR.sites.path, "CR07", bam.CR01)
FPdata.AR$CR08 <- analyzeFP("AR", AR.sites.path, "CR08", bam.CR01)

numSites <- AR.FP.data.CR01[["numSites"]]

AR.bindingMatrix <- matrix(data = 0, nrow = 8, ncol = numSites)
rownames(AR.bindingMatrix) <- c("WT01", "WT02", "CR01", "CR02", "CR04", "CR05", "CR07", "CR08")

AR.bindingMatrix[1,AR.FP.data.WT01[["idxBFpass"]]] <- 1
AR.bindingMatrix[2,AR.FP.data.WT02[["idxBFpass"]]] <- 1
AR.bindingMatrix[3,AR.FP.data.CR01[["idxBFpass"]]] <- 1
AR.bindingMatrix[4,AR.FP.data.CR02[["idxBFpass"]]] <- 1
AR.bindingMatrix[5,AR.FP.data.CR04[["idxBFpass"]]] <- 1
AR.bindingMatrix[6,AR.FP.data.CR05[["idxBFpass"]]] <- 1
AR.bindingMatrix[7,AR.FP.data.CR07[["idxBFpass"]]] <- 1
AR.bindingMatrix[8,AR.FP.data.CR08[["idxBFpass"]]] <- 1
FPdata.AR$bindingMatrix <- AR.bindingMatrix

outPath <- "C:\\Users\\jsk33\\Documents\\git\\miscLabScripts\\analysisWithAlessandroLNCaP\\FPdata.AR.RDS"
saveRDS(FPdata.AR, file = outPath)

```

SOX2
```{r}

SOX2.FP.data.WT01 <- analyzeFP("SOX2", SOX2.sites.path, "WT01", bam.WT01)
SOX2.FP.data.WT02 <- analyzeFP("SOX2", SOX2.sites.path, "WT02", bam.WT02)
SOX2.FP.data.CR01 <- analyzeFP("SOX2", SOX2.sites.path, "CR01", bam.CR01)
SOX2.FP.data.CR02 <- analyzeFP("SOX2", SOX2.sites.path, "CR02", bam.CR01)
SOX2.FP.data.CR04 <- analyzeFP("SOX2", SOX2.sites.path, "CR04", bam.CR01)
SOX2.FP.data.CR05 <- analyzeFP("SOX2", SOX2.sites.path, "CR05", bam.CR01)
SOX2.FP.data.CR07 <- analyzeFP("SOX2", SOX2.sites.path, "CR07", bam.CR01)
SOX2.FP.data.CR08 <- analyzeFP("SOX2", SOX2.sites.path, "CR08", bam.CR01)

FPdata.SOX2 <- list()
FPdata.SOX2$WT01 <- SOX2.FP.data.WT01
FPdata.SOX2$WT02 <- SOX2.FP.data.WT02
FPdata.SOX2$CR01 <- SOX2.FP.data.CR01
FPdata.SOX2$CR02 <- SOX2.FP.data.CR02
FPdata.SOX2$CR04 <- SOX2.FP.data.CR04
FPdata.SOX2$CR05 <- SOX2.FP.data.CR05
FPdata.SOX2$CR07 <- SOX2.FP.data.CR07
FPdata.SOX2$CR08 <- SOX2.FP.data.CR08

numSites <- SOX2.FP.data.CR01[["numSites"]]
SOX2.bindingMatrix <- matrix(data = 0, nrow = 8, ncol = numSites)
rownames(SOX2.bindingMatrix) <- c("WT01", "WT02", "CR01", "CR02", "CR04", "CR05", "CR07", "CR08")

SOX2.bindingMatrix[1,SOX2.FP.data.WT01[["idxBFpass"]]] <- 1
SOX2.bindingMatrix[2,SOX2.FP.data.WT02[["idxBFpass"]]] <- 1
SOX2.bindingMatrix[3,SOX2.FP.data.CR01[["idxBFpass"]]] <- 1
SOX2.bindingMatrix[4,SOX2.FP.data.CR02[["idxBFpass"]]] <- 1
SOX2.bindingMatrix[5,SOX2.FP.data.CR04[["idxBFpass"]]] <- 1
SOX2.bindingMatrix[6,SOX2.FP.data.CR05[["idxBFpass"]]] <- 1
SOX2.bindingMatrix[7,SOX2.FP.data.CR07[["idxBFpass"]]] <- 1
SOX2.bindingMatrix[8,SOX2.FP.data.CR08[["idxBFpass"]]] <- 1
FPdata.SOX2$bindingMatrix <- SOX2.bindingMatrix

outPath <- "C:\\Users\\jsk33\\Documents\\git\\miscLabScripts\\analysisWithAlessandroLNCaP\\FPdata.SOX2.RDS"
saveRDS(FPdata.SOX2, file = outPath)

```

EZH2
```{r}

EZH2.FP.data.WT01 <- analyzeFP("EZH2", EZH2.sites.path, "WT01", bam.WT01)
EZH2.FP.data.WT02 <- analyzeFP("EZH2", EZH2.sites.path, "WT02", bam.WT02)
EZH2.FP.data.CR01 <- analyzeFP("EZH2", EZH2.sites.path, "CR01", bam.CR01)
EZH2.FP.data.CR02 <- analyzeFP("EZH2", EZH2.sites.path, "CR02", bam.CR01)
EZH2.FP.data.CR04 <- analyzeFP("EZH2", EZH2.sites.path, "CR04", bam.CR01)
EZH2.FP.data.CR05 <- analyzeFP("EZH2", EZH2.sites.path, "CR05", bam.CR01)
EZH2.FP.data.CR07 <- analyzeFP("EZH2", EZH2.sites.path, "CR07", bam.CR01)
EZH2.FP.data.CR08 <- analyzeFP("EZH2", EZH2.sites.path, "CR08", bam.CR01)

FPdata.EZH2 <- list()
FPdata.EZH2$WT01 <- EZH2.FP.data.WT01
FPdata.EZH2$WT02 <- EZH2.FP.data.WT02
FPdata.EZH2$CR01 <- EZH2.FP.data.CR01
FPdata.EZH2$CR02 <- EZH2.FP.data.CR02
FPdata.EZH2$CR04 <- EZH2.FP.data.CR04
FPdata.EZH2$CR05 <- EZH2.FP.data.CR05
FPdata.EZH2$CR07 <- EZH2.FP.data.CR07
FPdata.EZH2$CR08 <- EZH2.FP.data.CR08

numSites <- EZH2.FP.data.CR01[["numSites"]]
EZH2.bindingMatrix <- matrix(data = 0, nrow = 8, ncol = numSites)
rownames(EZH2.bindingMatrix) <- c("WT01", "WT02", "CR01", "CR02", "CR04", "CR05", "CR07", "CR08")

EZH2.bindingMatrix[1,EZH2.FP.data.WT01[["idxBFpass"]]] <- 1
EZH2.bindingMatrix[2,EZH2.FP.data.WT02[["idxBFpass"]]] <- 1
EZH2.bindingMatrix[3,EZH2.FP.data.CR01[["idxBFpass"]]] <- 1
EZH2.bindingMatrix[4,EZH2.FP.data.CR02[["idxBFpass"]]] <- 1
EZH2.bindingMatrix[5,EZH2.FP.data.CR04[["idxBFpass"]]] <- 1
EZH2.bindingMatrix[6,EZH2.FP.data.CR05[["idxBFpass"]]] <- 1
EZH2.bindingMatrix[7,EZH2.FP.data.CR07[["idxBFpass"]]] <- 1
EZH2.bindingMatrix[8,EZH2.FP.data.CR08[["idxBFpass"]]] <- 1
FPdata.EZH2$bindingMatrix <- EZH2.bindingMatrix

outPath <- "C:\\Users\\jsk33\\Documents\\git\\miscLabScripts\\analysisWithAlessandroLNCaP\\FPdata.EZH2.RDS"
saveRDS(FPdata.EZH2, file = outPath)

```

MYCN
```{r}

MYCN.FP.data.WT01 <- analyzeFP("MYCN", MYCN.sites.path, "WT01", bam.WT01)
MYCN.FP.data.WT02 <- analyzeFP("MYCN", MYCN.sites.path, "WT02", bam.WT02)
MYCN.FP.data.CR01 <- analyzeFP("MYCN", MYCN.sites.path, "CR01", bam.CR01)
MYCN.FP.data.CR02 <- analyzeFP("MYCN", MYCN.sites.path, "CR02", bam.CR01)
MYCN.FP.data.CR04 <- analyzeFP("MYCN", MYCN.sites.path, "CR04", bam.CR01)
MYCN.FP.data.CR05 <- analyzeFP("MYCN", MYCN.sites.path, "CR05", bam.CR01)
MYCN.FP.data.CR07 <- analyzeFP("MYCN", MYCN.sites.path, "CR07", bam.CR01)
MYCN.FP.data.CR08 <- analyzeFP("MYCN", MYCN.sites.path, "CR08", bam.CR01)

FPdata.MYCN <- list()
FPdata.MYCN$WT01 <- MYCN.FP.data.WT01
FPdata.MYCN$WT02 <- MYCN.FP.data.WT02
FPdata.MYCN$CR01 <- MYCN.FP.data.CR01
FPdata.MYCN$CR02 <- MYCN.FP.data.CR02
FPdata.MYCN$CR04 <- MYCN.FP.data.CR04
FPdata.MYCN$CR05 <- MYCN.FP.data.CR05
FPdata.MYCN$CR07 <- MYCN.FP.data.CR07
FPdata.MYCN$CR08 <- MYCN.FP.data.CR08

numSites <- MYCN.FP.data.CR01[["numSites"]]
MYCN.bindingMatrix <- matrix(data = 0, nrow = 8, ncol = numSites)
rownames(MYCN.bindingMatrix) <- c("WT01", "WT02", "CR01", "CR02", "CR04", "CR05", "CR07", "CR08")

MYCN.bindingMatrix[1,MYCN.FP.data.WT01[["idxBFpass"]]] <- 1
MYCN.bindingMatrix[2,MYCN.FP.data.WT02[["idxBFpass"]]] <- 1
MYCN.bindingMatrix[3,MYCN.FP.data.CR01[["idxBFpass"]]] <- 1
MYCN.bindingMatrix[4,MYCN.FP.data.CR02[["idxBFpass"]]] <- 1
MYCN.bindingMatrix[5,MYCN.FP.data.CR04[["idxBFpass"]]] <- 1
MYCN.bindingMatrix[6,MYCN.FP.data.CR05[["idxBFpass"]]] <- 1
MYCN.bindingMatrix[7,MYCN.FP.data.CR07[["idxBFpass"]]] <- 1
MYCN.bindingMatrix[8,MYCN.FP.data.CR08[["idxBFpass"]]] <- 1
FPdata.MYCN$bindingMatrix <- MYCN.bindingMatrix

outPath <- "C:\\Users\\jsk33\\Documents\\git\\miscLabScripts\\analysisWithAlessandroLNCaP\\FPdata.MYCN.RDS"
saveRDS(FPdata.MYCN, file = outPath)

```

MYC
```{r}

MYC.FP.data.WT01 <- analyzeFP("MYC", MYC.sites.path, "WT01", bam.WT01)
MYC.FP.data.WT02 <- analyzeFP("MYC", MYC.sites.path, "WT02", bam.WT02)
MYC.FP.data.CR01 <- analyzeFP("MYC", MYC.sites.path, "CR01", bam.CR01)
MYC.FP.data.CR02 <- analyzeFP("MYC", MYC.sites.path, "CR02", bam.CR01)
MYC.FP.data.CR04 <- analyzeFP("MYC", MYC.sites.path, "CR04", bam.CR01)
MYC.FP.data.CR05 <- analyzeFP("MYC", MYC.sites.path, "CR05", bam.CR01)
MYC.FP.data.CR07 <- analyzeFP("MYC", MYC.sites.path, "CR07", bam.CR01)
MYC.FP.data.CR08 <- analyzeFP("MYC", MYC.sites.path, "CR08", bam.CR01)

FPdata.MYC <- list()
FPdata.MYC$WT01 <- MYC.FP.data.WT01
FPdata.MYC$WT02 <- MYC.FP.data.WT02
FPdata.MYC$CR01 <- MYC.FP.data.CR01
FPdata.MYC$CR02 <- MYC.FP.data.CR02
FPdata.MYC$CR04 <- MYC.FP.data.CR04
FPdata.MYC$CR05 <- MYC.FP.data.CR05
FPdata.MYC$CR07 <- MYC.FP.data.CR07
FPdata.MYC$CR08 <- MYC.FP.data.CR08

numSites <- MYC.FP.data.CR01[["numSites"]]
MYC.bindingMatrix <- matrix(data = 0, nrow = 8, ncol = numSites)
rownames(MYC.bindingMatrix) <- c("WT01", "WT02", "CR01", "CR02", "CR04", "CR05", "CR07", "CR08")

MYC.bindingMatrix[1,MYC.FP.data.WT01[["idxBFpass"]]] <- 1
MYC.bindingMatrix[2,MYC.FP.data.WT02[["idxBFpass"]]] <- 1
MYC.bindingMatrix[3,MYC.FP.data.CR01[["idxBFpass"]]] <- 1
MYC.bindingMatrix[4,MYC.FP.data.CR02[["idxBFpass"]]] <- 1
MYC.bindingMatrix[5,MYC.FP.data.CR04[["idxBFpass"]]] <- 1
MYC.bindingMatrix[6,MYC.FP.data.CR05[["idxBFpass"]]] <- 1
MYC.bindingMatrix[7,MYC.FP.data.CR07[["idxBFpass"]]] <- 1
MYC.bindingMatrix[8,MYC.FP.data.CR08[["idxBFpass"]]] <- 1
FPdata.MYC$bindingMatrix <- MYC.bindingMatrix

outPath <- "C:\\Users\\jsk33\\Documents\\git\\miscLabScripts\\analysisWithAlessandroLNCaP\\FPdata.MYC.RDS"
saveRDS(FPdata.MYC, file = outPath)

```

FOXM1
```{r}

FOXM1.FP.data.WT01 <- analyzeFP("FOXM1", FOXM1.sites.path, "WT01", bam.WT01)
FOXM1.FP.data.WT02 <- analyzeFP("FOXM1", FOXM1.sites.path, "WT02", bam.WT02)
FOXM1.FP.data.CR01 <- analyzeFP("FOXM1", FOXM1.sites.path, "CR01", bam.CR01)
FOXM1.FP.data.CR02 <- analyzeFP("FOXM1", FOXM1.sites.path, "CR02", bam.CR01)
FOXM1.FP.data.CR04 <- analyzeFP("FOXM1", FOXM1.sites.path, "CR04", bam.CR01)
FOXM1.FP.data.CR05 <- analyzeFP("FOXM1", FOXM1.sites.path, "CR05", bam.CR01)
FOXM1.FP.data.CR07 <- analyzeFP("FOXM1", FOXM1.sites.path, "CR07", bam.CR01)
FOXM1.FP.data.CR08 <- analyzeFP("FOXM1", FOXM1.sites.path, "CR08", bam.CR01)

FPdata.FOXM1 <- list()
FPdata.FOXM1$WT01 <- FOXM1.FP.data.WT01
FPdata.FOXM1$WT02 <- FOXM1.FP.data.WT02
FPdata.FOXM1$CR01 <- FOXM1.FP.data.CR01
FPdata.FOXM1$CR02 <- FOXM1.FP.data.CR02
FPdata.FOXM1$CR04 <- FOXM1.FP.data.CR04
FPdata.FOXM1$CR05 <- FOXM1.FP.data.CR05
FPdata.FOXM1$CR07 <- FOXM1.FP.data.CR07
FPdata.FOXM1$CR08 <- FOXM1.FP.data.CR08

numSites <- FOXM1.FP.data.CR01[["numSites"]]
FOXM1.bindingMatrix <- matrix(data = 0, nrow = 8, ncol = numSites)
rownames(FOXM1.bindingMatrix) <- c("WT01", "WT02", "CR01", "CR02", "CR04", "CR05", "CR07", "CR08")

FOXM1.bindingMatrix[1,MYC.FP.data.WT01[["idxBFpass"]]] <- 1
FOXM1.bindingMatrix[2,MYC.FP.data.WT02[["idxBFpass"]]] <- 1
FOXM1.bindingMatrix[3,MYC.FP.data.CR01[["idxBFpass"]]] <- 1
FOXM1.bindingMatrix[4,MYC.FP.data.CR02[["idxBFpass"]]] <- 1
FOXM1.bindingMatrix[5,MYC.FP.data.CR04[["idxBFpass"]]] <- 1
FOXM1.bindingMatrix[6,MYC.FP.data.CR05[["idxBFpass"]]] <- 1
FOXM1.bindingMatrix[7,MYC.FP.data.CR07[["idxBFpass"]]] <- 1
FOXM1.bindingMatrix[8,MYC.FP.data.CR08[["idxBFpass"]]] <- 1
FPdata.FOXM1$bindingMatrix <- FOXM1.bindingMatrix

outPath <- "C:\\Users\\jsk33\\Documents\\git\\miscLabScripts\\analysisWithAlessandroLNCaP\\FPdata.FOXM1.RDS"
saveRDS(FPdata.FOXM1, file = outPath)

```

Load FP data
```{r}

FPdata.AR <- readRDS("C:\\Users\\jsk33\\Documents\\git\\miscLabScripts\\analysisWithAlessandroLNCaP\\FPdata.AR.RDS")
FPdata.SOX2 <- readRDS("C:\\Users\\jsk33\\Documents\\git\\miscLabScripts\\analysisWithAlessandroLNCaP\\FPdata.SOX2.RDS")
FPdata.EZH2 <- readRDS("C:\\Users\\jsk33\\Documents\\git\\miscLabScripts\\analysisWithAlessandroLNCaP\\FPdata.EZH2.RDS")
FPdata.MYCN <- readRDS("C:\\Users\\jsk33\\Documents\\git\\miscLabScripts\\analysisWithAlessandroLNCaP\\FPdata.MYCN.RDS")
FPdata.MYC <- readRDS("C:\\Users\\jsk33\\Documents\\git\\miscLabScripts\\analysisWithAlessandroLNCaP\\FPdata.MYC.RDS")
FPdata.FOXM1 <- readRDS("C:\\Users\\jsk33\\Documents\\git\\miscLabScripts\\analysisWithAlessandroLNCaP\\FPdata.FOXM1.RDS")

```

Plots
```{r}
install.packages("pheatmap")
library(pheatmap)

library(dplyr)

AR.matrix <- FPdata.AR[["bindingMatrix"]]
x <- AR.matrix[,1:100]
pheatmap(x)

## Subset for only locations where both WT are the same
y <- as.data.frame(AR.matrix)
z <- dplyr::filter(y, WT01 == 1, WT02 == 2)

aa <- which(y[1,] == 1 & y[2,] == 1)
ab <- which(y[1,] == 0 & y[2,] == 0)
ac <- c(aa, ab)
new <- y[1:8,ac]

new2 <- new[,24728:24828]
pheatmap(new2,
         cluster_rows = FALSE,
         cluster_cols = FALSE)


EZH2.matrix <- FPdata.EZH2[["bindingMatrix"]]
x <- EZH2.matrix[,1:100]
pheatmap(x)

FOXM1.matrix <- FPdata.FOXM1[["bindingMatrix"]]
x <- FOXM1.matrix[,1:100]
pheatmap(x)

MYC.matrix <- FPdata.MYC[["bindingMatrix"]]
x <- MYC.matrix[,1:100]
pheatmap(x)

MYCN.matrix <- FPdata.MYCN[["bindingMatrix"]]
x <- MYCN.matrix[,1:100]
pheatmap(x)

SOX2.matrix <- FPdata.SOX2[["bindingMatrix"]]
x <- SOX2.matrix[,1:100]
pheatmap(x)


```

Diff peaks overlap
```{r}

ctrl.wt01.treat.cr01.peaksPath <- "D:\\diffpeaks\\ctrl-wt01.treat-cr01_peaks.narrowPeak"
ctrl.wt01.treat.cr02.peaksPath <- "D:\\diffpeaks\\ctrl-wt01.treat-cr02_peaks.narrowPeak"
ctrl.wt01.treat.cr04.peaksPath <- "D:\\diffpeaks\\ctrl-wt01.treat-cr04_peaks.narrowPeak"
ctrl.wt01.treat.cr05.peaksPath <- "D:\\diffpeaks\\ctrl-wt01.treat-cr05_peaks.narrowPeak"
ctrl.wt01.treat.cr07.peaksPath <- "D:\\diffpeaks\\ctrl-wt01.treat-cr07_peaks.narrowPeak"
ctrl.wt01.treat.cr08.peaksPath <- "D:\\diffpeaks\\ctrl-wt01.treat-cr08_peaks.narrowPeak"

ctrl.wt02.treat.cr01.peaksPath <- "D:\\diffpeaks\\ctrl-wt02.treat-cr01_peaks.narrowPeak"
ctrl.wt02.treat.cr02.peaksPath <- "D:\\diffpeaks\\ctrl-wt02.treat-cr01_peaks.narrowPeak"
ctrl.wt02.treat.cr04.peaksPath <- "D:\\diffpeaks\\ctrl-wt02.treat-cr01_peaks.narrowPeak"
ctrl.wt02.treat.cr05.peaksPath <- "D:\\diffpeaks\\ctrl-wt02.treat-cr01_peaks.narrowPeak"
ctrl.wt02.treat.cr07.peaksPath <- "D:\\diffpeaks\\ctrl-wt02.treat-cr01_peaks.narrowPeak"
ctrl.wt02.treat.cr08.peaksPath <- "D:\\diffpeaks\\ctrl-wt02.treat-cr01_peaks.narrowPeak"

ctrl.wt01.treat.cr01.peaks <- importBED(ctrl.wt01.treat.cr01.peaksPath)
ctrl.wt01.treat.cr02.peaks <- importBED(ctrl.wt01.treat.cr02.peaksPath)
ctrl.wt01.treat.cr04.peaks <- importBED(ctrl.wt01.treat.cr04.peaksPath)
ctrl.wt01.treat.cr05.peaks <- importBED(ctrl.wt01.treat.cr05.peaksPath)
ctrl.wt01.treat.cr07.peaks <- importBED(ctrl.wt01.treat.cr07.peaksPath)
ctrl.wt01.treat.cr08.peaks <- importBED(ctrl.wt01.treat.cr08.peaksPath)

ctrl.wt02.treat.cr01.peaks <- importBED(ctrl.wt02.treat.cr01.peaksPath)
ctrl.wt02.treat.cr02.peaks <- importBED(ctrl.wt02.treat.cr02.peaksPath)
ctrl.wt02.treat.cr04.peaks <- importBED(ctrl.wt02.treat.cr04.peaksPath)
ctrl.wt02.treat.cr05.peaks <- importBED(ctrl.wt02.treat.cr05.peaksPath)
ctrl.wt02.treat.cr07.peaks <- importBED(ctrl.wt02.treat.cr07.peaksPath)
ctrl.wt02.treat.cr08.peaks <- importBED(ctrl.wt02.treat.cr08.peaksPath)

cr01.wt.overlap.peaks <- subsetByOverlaps(ctrl.wt01.treat.cr01.peaks, ctrl.wt02.treat.cr01.peaks)
cr02.wt.overlap.peaks <- subsetByOverlaps(ctrl.wt01.treat.cr02.peaks, ctrl.wt02.treat.cr02.peaks)
cr04.wt.overlap.peaks <- subsetByOverlaps(ctrl.wt01.treat.cr04.peaks, ctrl.wt02.treat.cr04.peaks)
cr05.wt.overlap.peaks <- subsetByOverlaps(ctrl.wt01.treat.cr05.peaks, ctrl.wt02.treat.cr05.peaks)
cr07.wt.overlap.peaks <- subsetByOverlaps(ctrl.wt01.treat.cr07.peaks, ctrl.wt02.treat.cr07.peaks)
cr08.wt.overlap.peaks <- subsetByOverlaps(ctrl.wt01.treat.cr08.peaks, ctrl.wt02.treat.cr08.peaks)

cr01.cr02.overlap.peaks <- subsetByOverlaps(cr01.wt.overlap.peaks, cr02.wt.overlap.peaks)
cr04.cr05.overlap.peaks <- subsetByOverlaps(cr04.wt.overlap.peaks, cr05.wt.overlap.peaks)
cr07.cr08.overlap.peaks <- subsetByOverlaps(cr07.wt.overlap.peaks, cr08.wt.overlap.peaks)

all.clones.overlap.peaks <- c(cr01.cr02.overlap.peaks, cr04.cr05.overlap.peaks, cr07.cr08.overlap.peaks)

wt01idx <- getGRoverlapIndex(WT01.peaks, all.clones.overlap.peaks)
wt02idx <- getGRoverlapIndex(WT02.peaks, all.clones.overlap.peaks)
cr01idx <- getGRoverlapIndex(cr01.wt.overlap.peaks, all.clones.overlap.peaks)
cr02idx <- getGRoverlapIndex(cr02.wt.overlap.peaks, all.clones.overlap.peaks)
cr04idx <- getGRoverlapIndex(cr04.wt.overlap.peaks, all.clones.overlap.peaks)
cr05idx <- getGRoverlapIndex(cr05.wt.overlap.peaks, all.clones.overlap.peaks)
cr07idx <- getGRoverlapIndex(cr07.wt.overlap.peaks, all.clones.overlap.peaks)
cr08idx <- getGRoverlapIndex(cr08.wt.overlap.peaks, all.clones.overlap.peaks)

newmatrix <- matrix(data = 0, nrow = 8, ncol = 173306)
rownames(newmatrix) <- c("WT01", "WT02", "CR01", "CR02", "CR04", "CR05", "CR07", "CR08")

newmatrix[1,wt01idx] <- 1
newmatrix[2,wt02idx] <- 1
newmatrix[3,cr01idx] <- 1
newmatrix[4,cr02idx] <- 1
newmatrix[5,cr04idx] <- 1
newmatrix[6,cr05idx] <- 1
newmatrix[7,cr07idx] <- 1
newmatrix[8,cr08idx] <- 1


pheatmap(newmatrix[,5000:5100],
         cluster_rows = FALSE,
         cluster_cols = FALSE)

```

Peak overlaps
```{r}

# WT01.peaksPath <- "C:\\Users\\jsk33\\Desktop\\peaks\\LNCaP-WT-01-REP1_globalnorm_peaks.narrowPeak"
# WT02.peaksPath <- "C:\\Users\\jsk33\\Desktop\\peaks\\LNCaP-WT-02-REP1_globalnorm_peaks.narrowPeak"
# CR01.peaksPath <- "C:\\Users\\jsk33\\Desktop\\peaks\\LNCaP-CR-01-REP1_globalnorm_peaks.narrowPeak"
# CR02.peaksPath <- "C:\\Users\\jsk33\\Desktop\\peaks\\LNCaP-CR-02-REP1_globalnorm_peaks.narrowPeak"
# CR04.peaksPath <- "C:\\Users\\jsk33\\Desktop\\peaks\\LNCaP-CR-04-REP1_globalnorm_peaks.narrowPeak"
# CR05.peaksPath <- "C:\\Users\\jsk33\\Desktop\\peaks\\LNCaP-CR-05-REP1_globalnorm_peaks.narrowPeak"
# CR07.peaksPath <- "C:\\Users\\jsk33\\Desktop\\peaks\\LNCaP-CR-07-REP1_globalnorm_peaks.narrowPeak"
# CR08.peaksPath <- "C:\\Users\\jsk33\\Desktop\\peaks\\LNCaP-CR-08-REP1_globalnorm_peaks.narrowPeak"

WT01.peaksPath <- "D:\\peaks\\LNCaP-WT-01-REP1_globalnorm_peaks.narrowPeak"
WT02.peaksPath <- "D:\\peaks\\LNCaP-WT-02-REP1_globalnorm_peaks.narrowPeak"
CR01.peaksPath <- "D:\\peaks\\LNCaP-CR-01-REP1_globalnorm_peaks.narrowPeak"
CR02.peaksPath <- "D:\\peaks\\LNCaP-CR-02-REP1_globalnorm_peaks.narrowPeak"
CR04.peaksPath <- "D:\\peaks\\LNCaP-CR-04-REP1_globalnorm_peaks.narrowPeak"
CR05.peaksPath <- "D:\\peaks\\LNCaP-CR-05-REP1_globalnorm_peaks.narrowPeak"
CR07.peaksPath <- "D:\\peaks\\LNCaP-CR-07-REP1_globalnorm_peaks.narrowPeak"
CR08.peaksPath <- "D:\\peaks\\LNCaP-CR-08-REP1_globalnorm_peaks.narrowPeak"

WT01.peaks <- importBED(WT01.peaksPath)
WT02.peaks <- importBED(WT02.peaksPath)
CR01.peaks <- importBED(CR01.peaksPath)
CR02.peaks <- importBED(CR02.peaksPath)
CR04.peaks <- importBED(CR04.peaksPath)
CR05.peaks <- importBED(CR05.peaksPath)
CR07.peaks <- importBED(CR07.peaksPath)
CR08.peaks <- importBED(CR08.peaksPath)

FPdata.AR.WT01 <- FPdata.AR[["WT01"]]
FPdata.AR.WT01.inputSites <- FPdata.AR.WT01[["inputSites"]]
WT01.AR.peak.overlap.index <- findOverlaps(FPdata.AR.WT01.inputSites, WT01.peaks)
AR.WT01.idx <- unique(WT01.AR.peak.overlap.index@from)

FPdata.AR.WT02 <- FPdata.AR[["WT02"]]
FPdata.AR.WT02.inputSites <- FPdata.AR.WT02[["inputSites"]]
WT02.AR.peak.overlap.index <- findOverlaps(FPdata.AR.WT02.inputSites, WT02.peaks)
AR.WT02.idx <- unique(WT02.AR.peak.overlap.index@from)

FPdata.AR.CR01 <- FPdata.AR[["CR01"]]
FPdata.AR.CR01.inputSites <- FPdata.AR.CR01[["inputSites"]]
CR01.AR.peak.overlap.index <- findOverlaps(FPdata.AR.CR01.inputSites, CR01.peaks)
AR.CR01.idx <- unique(CR01.AR.peak.overlap.index@from)

FPdata.AR.CR02 <- FPdata.AR[["CR02"]]
FPdata.AR.CR02.inputSites <- FPdata.AR.CR02[["inputSites"]]
CR02.AR.peak.overlap.index <- findOverlaps(FPdata.AR.CR02.inputSites, CR02.peaks)
AR.CR02.idx <- unique(CR02.AR.peak.overlap.index@from)

FPdata.AR.CR04 <- FPdata.AR[["CR04"]]
FPdata.AR.CR04.inputSites <- FPdata.AR.CR04[["inputSites"]]
CR04.AR.peak.overlap.index <- findOverlaps(FPdata.AR.CR04.inputSites, CR04.peaks)
AR.CR04.idx <- unique(CR04.AR.peak.overlap.index@from)

FPdata.AR.CR05 <- FPdata.AR[["CR05"]]
FPdata.AR.CR05.inputSites <- FPdata.AR.CR05[["inputSites"]]
CR05.AR.peak.overlap.index <- findOverlaps(FPdata.AR.CR05.inputSites, CR05.peaks)
AR.CR05.idx <- unique(CR05.AR.peak.overlap.index@from)

FPdata.AR.CR07 <- FPdata.AR[["CR07"]]
FPdata.AR.CR07.inputSites <- FPdata.AR.CR07[["inputSites"]]
CR07.AR.peak.overlap.index <- findOverlaps(FPdata.AR.CR07.inputSites, CR07.peaks)
AR.CR07.idx <- unique(CR07.AR.peak.overlap.index@from)

FPdata.AR.CR08 <- FPdata.AR[["CR08"]]
FPdata.AR.CR08.inputSites <- FPdata.AR.CR08[["inputSites"]]
CR08.AR.peak.overlap.index <- findOverlaps(FPdata.AR.CR08.inputSites, CR08.peaks)
AR.CR08.idx <- unique(CR08.AR.peak.overlap.index@from)

newmatrix <- matrix(data = 0, nrow = 8, ncol = 62585)
rownames(newmatrix) <- c("WT01", "WT02", "CR01", "CR02", "CR04", "CR05", "CR07", "CR08")
newmatrix[1,AR.WT01.idx] <- 1
newmatrix[2,AR.WT02.idx] <- 1
newmatrix[3,AR.CR01.idx] <- 1
newmatrix[4,AR.CR02.idx] <- 1
newmatrix[5,AR.CR04.idx] <- 1
newmatrix[6,AR.CR05.idx] <- 1
newmatrix[7,AR.CR07.idx] <- 1
newmatrix[8,AR.CR08.idx] <- 1

pheatmap(newmatrix[,1:100],
         cluster_rows = FALSE,
         cluster_cols = FALSE)

AR.matrix <- FPdata.AR[["bindingMatrix"]]
x <- AR.matrix[,AR.WT01.idx]
pheatmap(x[,100:200])

y <- as.data.frame(AR.matrix)
z <- dplyr::filter(y, WT01 == 1, WT02 == 2)

aa <- which(y[1,] == 1 & y[2,] == 1)
ab <- which(y[1,] == 0 & y[2,] == 0)
ac <- c(aa, ab)
new <- y[1:8,ac]

new2 <- new[,24728:24828]
pheatmap(new2,
         cluster_rows = FALSE,
         cluster_cols = FALSE)

```


```{r}

library(chromoMap)
library(Repitools)

chrom <- "C:\\Users\\Jordan\\Desktop\\xsomes.txt"

ctrl.wt01.treat.cr02.peaksPath <- "D:\\diffpeaks\\ctrl-wt01.treat-cr02_peaks.narrowPeak"
ctrl.wt01.treat.cr02.peaks <- readBed(ctrl.wt01.treat.cr02.peaksPath, track.line = FALSE, remove.unusual = FALSE, zero.based = TRUE)
ctrl.wt01.treat.cr02.peaks <- keepStandardChromosomes(ctrl.wt01.treat.cr02.peaks, pruning.mode="coarse")
ctrl.wt01.treat.cr02.peaks <- trim(ctrl.wt01.treat.cr02.peaks)

xx <- Repitools::annoGR2DF(ctrl.wt01.treat.cr02.peaks)

xy <- data.frame(name = 1:100,
                 chr = xx[1:100,1],
                 start = xx[1:100,2],
                 end = xx[1:100,3],
                 data = xx[1:100, 5]/100)


write.table(xy,
            file = "C:\\Users\\Jordan\\Desktop\\test.txt",
            quote = FALSE,
            col.names = FALSE,
            row.names = FALSE,
            append = FALSE,
            sep = "\t"
            )
  


chromoMap("C:\\Users\\Jordan\\Desktop\\xsomes.txt",
          "C:\\Users\\Jordan\\Desktop\\test.txt",
          data_based_color_map = T,
          data_type = numeric
          )


```


```{r}
###################################################
### code chunk number 1: ChromHeatMap.Rnw:57-64
###################################################
library('ALL')
data('ALL')
selSamples <- ALL$mol.biol %in% c('ALL1/AF4', 'E2A/PBX1')
ALLs <- ALL[, selSamples]

library('ChromHeatMap')
chrdata<-makeChrStrandData(exprs(ALLs), lib='hgu95av2')


###################################################
### code chunk number 2: ChromHeatMap.Rnw:87-89
###################################################
groupcol <- ifelse( ALLs$mol.biol == 'ALL1/AF4', 'red', 'green' )
plotChrMap(chrdata, 19, strands='both', RowSideColors=groupcol)


###################################################
### code chunk number 3: ChromHeatMap.Rnw:98-99
###################################################
plotmap<-plotChrMap(chrdata, 1, cytoband='q23', interval=50000, srtCyto=0, cexCyto=1.2)


###################################################
### code chunk number 4: ChromHeatMap.Rnw:135-137 (eval = FALSE)
###################################################
## probes <- grabChrMapProbes( plotmap )
## genes <- unlist(mget(probes, envir=hgu95av2SYMBOL, ifnotfound=NA))


###################################################
### code chunk number 5: ChromHeatMap.Rnw:156-157
###################################################
sessionInfo()


```












